
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
        <meta name="description" content="Extend DeepVariant automatically for species without GIAB-quality benchmark truth sets.">
      
      
        <meta name="author" content="Jenna Kalleberg">
      
      
      
        <link rel="prev" href="../configuration/">
      
      
        <link rel="next" href="../../user-guide/usage_guide/">
      
      <link rel="icon" href="../../assets/images/favicon.png">
      <meta name="generator" content="mkdocs-1.4.3, mkdocs-material-9.1.14">
    
    
      
        <title>Human GIAB Tutorial - DeepVariant-TrioTrain</title>
      
    
    
      <link rel="stylesheet" href="../../assets/stylesheets/main.85bb2934.min.css">
      
      

    
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
      
    
    
    <script>__md_scope=new URL("../..",location),__md_hash=e=>[...e].reduce((e,_)=>(e<<5)-e+_.charCodeAt(0),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
    
  </head>
  
  
    <body dir="ltr">
  
    
    
      <script>var palette=__md_get("__palette");if(palette&&"object"==typeof palette.color)for(var key of Object.keys(palette.color))document.body.setAttribute("data-md-color-"+key,palette.color[key])</script>
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#human-giab-tutorial" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

  

<header class="md-header md-header--shadow md-header--lifted" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href="../.." title="DeepVariant-TrioTrain" class="md-header__button md-logo" aria-label="DeepVariant-TrioTrain" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54Z"/></svg>

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3V6m0 5h18v2H3v-2m0 5h18v2H3v-2Z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            DeepVariant-TrioTrain
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              Human GIAB Tutorial
            
          </span>
        </div>
      </div>
    </div>
    
    
    
      <label class="md-header__button md-icon" for="__search">
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg>
      </label>
      <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12Z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
        <button type="reset" class="md-search__icon md-icon" title="Clear" aria-label="Clear" tabindex="-1">
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12 19 6.41Z"/></svg>
        </button>
      </nav>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list" role="presentation"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
    
    
      <div class="md-header__source">
        <a href="https://github.com/jkalleberg/DV-TrioTrain" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.4.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2023 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    jkalleberg/DV-TrioTrain
  </div>
</a>
      </div>
    
  </nav>
  
    
      
<nav class="md-tabs" aria-label="Tabs" data-md-component="tabs">
  <div class="md-grid">
    <ul class="md-tabs__list">
      
        
  
  


  <li class="md-tabs__item">
    <a href="../.." class="md-tabs__link">
      Home
    </a>
  </li>

      
        
  
  
    
  


  
  
  
    <li class="md-tabs__item">
      <a href="../getting-started/" class="md-tabs__link md-tabs__link--active">
        Getting Started
      </a>
    </li>
  

      
        
  
  


  
  
  
    <li class="md-tabs__item">
      <a href="../../user-guide/usage_guide/" class="md-tabs__link">
        User Guide
      </a>
    </li>
  

      
        
  
  


  
  
  
    <li class="md-tabs__item">
      <a href="../../about/about/" class="md-tabs__link">
        About
      </a>
    </li>
  

      
    </ul>
  </div>
</nav>
    
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

  


  

<nav class="md-nav md-nav--primary md-nav--lifted md-nav--integrated" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href="../.." title="DeepVariant-TrioTrain" class="md-nav__button md-logo" aria-label="DeepVariant-TrioTrain" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54Z"/></svg>

    </a>
    DeepVariant-TrioTrain
  </label>
  
    <div class="md-nav__source">
      <a href="https://github.com/jkalleberg/DV-TrioTrain" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.4.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2023 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    jkalleberg/DV-TrioTrain
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="../.." class="md-nav__link">
        Home
      </a>
    </li>
  

    
      
      
      

  
  
    
  
  
    
    <li class="md-nav__item md-nav__item--active md-nav__item--nested">
      
      
      
      
      <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_2" checked>
      
      
      
        <label class="md-nav__link" for="__nav_2" id="__nav_2_label" tabindex="0">
          Getting Started
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_2_label" aria-expanded="true">
        <label class="md-nav__title" for="__nav_2">
          <span class="md-nav__icon md-icon"></span>
          Getting Started
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../getting-started/" class="md-nav__link">
        Start Here
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../installation/" class="md-nav__link">
        Installation Guide
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../configuration/" class="md-nav__link">
        Configuration Guide
      </a>
    </li>
  

            
          
            
              
  
  
    
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" type="checkbox" id="__toc">
      
      
        
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          Human GIAB Tutorial
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="./" class="md-nav__link md-nav__link--active">
        Human GIAB Tutorial
      </a>
      
        

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#1-confirm-successful-configuration" class="md-nav__link">
    1. Confirm Successful Configuration
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#2-activate-environment" class="md-nav__link">
    2. Activate Environment
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#4-download-pre-trained-models" class="md-nav__link">
    4. Download pre-trained models
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#5-download-raw-data" class="md-nav__link">
    5. Download Raw Data
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#6-process-raw-data" class="md-nav__link">
    6. Process Raw Data
  </a>
  
    <nav class="md-nav" aria-label="6. Process Raw Data">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#a-merge-the-popvcf" class="md-nav__link">
    a. Merge the PopVCF
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#b-download-sequence-data" class="md-nav__link">
    b. Download Sequence Data
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#7-create-triotrain-inputs" class="md-nav__link">
    7. Create TrioTrain Inputs
  </a>
  
    <nav class="md-nav" aria-label="7. Create TrioTrain Inputs">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#a-reference-dictionary-file-dict" class="md-nav__link">
    a. Reference Dictionary File (.dict)
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#b-metadata-file-csv" class="md-nav__link">
    b. Metadata file (.csv)
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#c-slurm-resource-config-file-json" class="md-nav__link">
    c. SLURM Resource Config File (.json)
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#8-run-shuffling-demo" class="md-nav__link">
    8. Run Shuffling Demo
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#9-run-triotrain-with-a-human-trio" class="md-nav__link">
    9. Run TrioTrain with a Human Trio
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#merge-results-per-iteration" class="md-nav__link">
    Merge Results | Per-Iteration
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#clean-up-directories-per-iteration" class="md-nav__link">
    Clean Up Directories | Per-Iteration
  </a>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  

    
      
      
      

  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
      
      
      <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_3" >
      
      
      
        <label class="md-nav__link" for="__nav_3" id="__nav_3_label" tabindex="0">
          User Guide
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_3_label" aria-expanded="false">
        <label class="md-nav__title" for="__nav_3">
          <span class="md-nav__icon md-icon"></span>
          User Guide
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../user-guide/usage_guide/" class="md-nav__link">
        Overview
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../user-guide/existing_models/" class="md-nav__link">
        Models
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../user-guide/get-help/" class="md-nav__link">
        Getting help
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../user-guide/citation/" class="md-nav__link">
        How to cite
      </a>
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  

    
      
      
      

  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
      
      
      <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4" >
      
      
      
        <label class="md-nav__link" for="__nav_4" id="__nav_4_label" tabindex="0">
          About
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_4_label" aria-expanded="false">
        <label class="md-nav__title" for="__nav_4">
          <span class="md-nav__icon md-icon"></span>
          About
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../about/about/" class="md-nav__link">
        Learn more
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../about/license/" class="md-nav__link">
        License
      </a>
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
          
          
            <div class="md-content" data-md-component="content">
              <article class="md-content__inner md-typeset">
                
                  

  
  


<h1 id="human-giab-tutorial">Human GIAB Tutorial</h1>
<p><font size= "4"> 
<strong>This tutorial completes one round of re-training with a Human trio from GIAB.</strong>
</font></p>
<div class="admonition warning">
<p class="admonition-title">Warning | Storage Needs</p>
<p>Completing this tutorial will produce ~2T of intermediate and output data. Ensure you  have sufficient space before proceeding!</p>
</div>
<h2 id="1-confirm-successful-configuration">1. Confirm Successful Configuration</h2>
<details class="success">
<summary>Check | Installation</summary>
<div class="highlight"><span class="filename">Run the following at the command line:</span><pre><span></span><code><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a><span class="nb">cd</span><span class="w"> </span>DV-TrioTrain
<a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a>ls
</code></pre></div>
<div class="highlight"><span class="filename">Expected outputs:</span><pre><span></span><code><a id="__codelineno-1-1" name="__codelineno-1-1" href="#__codelineno-1-1"></a>deepvariant_1.4.0-gpu.sif<span class="w">  </span>docs<span class="w">    </span>hap.py_v0.3.12.sif<span class="w">  </span>miniconda_envs<span class="w">  </span>README.md<span class="w">  </span>triotrain
<a id="__codelineno-1-2" name="__codelineno-1-2" href="#__codelineno-1-2"></a>deepvariant_1.4.0.sif<span class="w">      </span>errors<span class="w">  </span>LICENSE<span class="w">             </span>mkdocs.yml<span class="w">      </span>scripts
</code></pre></div>
<ul>
<li>
<p><code>triotrain/</code> directory contains Python modules for the DV-TrioTrain package</p>
</li>
<li>
<p><code>scripts/</code> directory contains Bash helper scripts and functions that can be used as templates</p>
</li>
</ul>
</details>
<h2 id="2-activate-environment">2. Activate Environment</h2>
<p>Repeat the first two steps of Configuration:</p>
<ul>
<li><a href="../configuration/#interactive">Step 1: Interactive Session</a></li>
<li><a href="../configuration/#modules">Step 2: Load Modules</a></li>
</ul>
<h2 id="4-download-pre-trained-models">4. Download pre-trained models</h2>
<div class="admonition warning">
<p class="admonition-title">Warning | Download Size</p>
</div>
<p>Running a local copy of a container requires us to create a local copy of the <code>model.ckpt</code> files from v1.4.0 of DeepVariant. These checkpoints are the human-trained models produced by Google Genomics Health Group. <a href="../../user-guide/existing_models/">Details about published models compatible with TrioTrain can be found here</a>.</p>
<p>We need to download (2) two model checkpoints:</p>
<ul>
<li>the default human model</li>
<li>the WGS.AF human model</li>
</ul>
<div class="highlight"><span class="filename">Run the following at the command line:</span><pre><span></span><code><a id="__codelineno-2-1" name="__codelineno-2-1" href="#__codelineno-2-1"></a>bash<span class="w"> </span>scripts/setup/download_models.sh
</code></pre></div>
<details class="example">
<summary>Example | <code>download_models.sh</code></summary>
<div class="highlight"><pre><span></span><code><a id="__codelineno-3-1" name="__codelineno-3-1" href="#__codelineno-3-1"></a>#!/bin/bash
<a id="__codelineno-3-2" name="__codelineno-3-2" href="#__codelineno-3-2"></a># scripts/setup/download_models.sh
<a id="__codelineno-3-3" name="__codelineno-3-3" href="#__codelineno-3-3"></a>&quot;&quot;&quot;
<a id="__codelineno-3-4" name="__codelineno-3-4" href="#__codelineno-3-4"></a>NOTES: These model ckpts are only required for warm-starting re-training, and are not used with &#39;run_deepvariant&#39;
<a id="__codelineno-3-5" name="__codelineno-3-5" href="#__codelineno-3-5"></a>
<a id="__codelineno-3-6" name="__codelineno-3-6" href="#__codelineno-3-6"></a>&quot;&quot;&quot;
<a id="__codelineno-3-7" name="__codelineno-3-7" href="#__codelineno-3-7"></a>echo -e &quot;=== scripts/setup/download_models.sh &gt; start $(date)&quot;
<a id="__codelineno-3-8" name="__codelineno-3-8" href="#__codelineno-3-8"></a>
<a id="__codelineno-3-9" name="__codelineno-3-9" href="#__codelineno-3-9"></a>##======= Downloading default WGS model (without AF channel) ===================##
<a id="__codelineno-3-10" name="__codelineno-3-10" href="#__codelineno-3-10"></a>#   This model allows for ONLY 7 layers in the example images 
<a id="__codelineno-3-11" name="__codelineno-3-11" href="#__codelineno-3-11"></a>#   Meaning, it is NOT compatible examples built with the allele frequency channel!
<a id="__codelineno-3-12" name="__codelineno-3-12" href="#__codelineno-3-12"></a>
<a id="__codelineno-3-13" name="__codelineno-3-13" href="#__codelineno-3-13"></a>#   To view in web-browser, execute this to obtain a valid link: 
<a id="__codelineno-3-14" name="__codelineno-3-14" href="#__codelineno-3-14"></a>#       echo &quot;https://console.cloud.google.com/storage/browser/deepvariant/models/DeepVariant/${BIN_VERSION_DV}/DeepVariant-inception_v3-${BIN_VERSION_DV}+data-wgs_standard&quot;
<a id="__codelineno-3-15" name="__codelineno-3-15" href="#__codelineno-3-15"></a>
<a id="__codelineno-3-16" name="__codelineno-3-16" href="#__codelineno-3-16"></a>echo &quot;$(date &#39;+%Y-%m-%d %H:%M:%S&#39;) INFO: Downloading 7-channel WGS model checkpoint (with InsertSize, but without AlleleFreq) - DeepVariant v${BIN_VERSION_DV}&quot;
<a id="__codelineno-3-17" name="__codelineno-3-17" href="#__codelineno-3-17"></a>NO_AF_MODEL_BUCKET=&quot;https://storage.googleapis.com/deepvariant/models/DeepVariant/${BIN_VERSION_DV}/DeepVariant-inception_v3-${BIN_VERSION_DV}+data-wgs_standard&quot;
<a id="__codelineno-3-18" name="__codelineno-3-18" href="#__codelineno-3-18"></a>
<a id="__codelineno-3-19" name="__codelineno-3-19" href="#__codelineno-3-19"></a># Use the same file naming convention and enable restarting of any interrupted downloads
<a id="__codelineno-3-20" name="__codelineno-3-20" href="#__codelineno-3-20"></a>curl --create-dirs --continue-at - &quot;${NO_AF_MODEL_BUCKET}/model.ckpt.{data-00000-of-00001,index,example_info.json,meta}&quot; --output &quot;triotrain/model_training/pretrained_models/v${BIN_VERSION_DV}_withIS_noAF/model.ckpt.#1&quot;
<a id="__codelineno-3-21" name="__codelineno-3-21" href="#__codelineno-3-21"></a>echo &quot;$(date &#39;+%Y-%m-%d %H:%M:%S&#39;) INFO: Done - Downloading 7-channel WGS model checkpoint (with InsertSize, but without AlleleFreq)&quot;
<a id="__codelineno-3-22" name="__codelineno-3-22" href="#__codelineno-3-22"></a>##========================================================================##
<a id="__codelineno-3-23" name="__codelineno-3-23" href="#__codelineno-3-23"></a>
<a id="__codelineno-3-24" name="__codelineno-3-24" href="#__codelineno-3-24"></a>##======= Downloading custom WGS model (with Allele Frequency channel) ======================##
<a id="__codelineno-3-25" name="__codelineno-3-25" href="#__codelineno-3-25"></a>#   This model allows for 8 layers in the example images 
<a id="__codelineno-3-26" name="__codelineno-3-26" href="#__codelineno-3-26"></a>#   Meaning, it IS compatible compatible examples built with the allele frequency channel!
<a id="__codelineno-3-27" name="__codelineno-3-27" href="#__codelineno-3-27"></a>
<a id="__codelineno-3-28" name="__codelineno-3-28" href="#__codelineno-3-28"></a>#   To view in web-browser, execute this to obtain a valid link:  
<a id="__codelineno-3-29" name="__codelineno-3-29" href="#__codelineno-3-29"></a>#       https://console.cloud.google.com/storage/browser/brain-genomics-public/research/allele_frequency/pretrained_model_WGS/1.4.0;tab=objects?pageState=(%22StorageObjectListTable%22:(%22f%22:%22%255B%255D%22))&amp;prefix=&amp;forceOnObjectsSortingFiltering=false
<a id="__codelineno-3-30" name="__codelineno-3-30" href="#__codelineno-3-30"></a>
<a id="__codelineno-3-31" name="__codelineno-3-31" href="#__codelineno-3-31"></a>echo &quot;$(date &#39;+%Y-%m-%d %H:%M:%S&#39;) INFO: Downloading 8-channel WGS model checkpoint (with InsertSize, with AlleleFreq) - DeepVariant v${BIN_VERSION_DV}&quot;
<a id="__codelineno-3-32" name="__codelineno-3-32" href="#__codelineno-3-32"></a>
<a id="__codelineno-3-33" name="__codelineno-3-33" href="#__codelineno-3-33"></a>AF_MODEL_BUCKET=&quot;https://storage.googleapis.com/brain-genomics-public/research/allele_frequency/pretrained_model_WGS/${BIN_VERSION_DV}&quot;
<a id="__codelineno-3-34" name="__codelineno-3-34" href="#__codelineno-3-34"></a>
<a id="__codelineno-3-35" name="__codelineno-3-35" href="#__codelineno-3-35"></a># Use the same file naming convention and enable restarting of any interrupted downloads
<a id="__codelineno-3-36" name="__codelineno-3-36" href="#__codelineno-3-36"></a>curl --create-dirs --continue-at - &quot;${AF_MODEL_BUCKET}/model.ckpt.{data-00000-of-00001,index,meta,example_info.json}&quot; --output &quot;triotrain/model_training/pretrained_models/v${BIN_VERSION_DV}_withIS_withAF/wgs_af.model.ckpt.#1&quot;
<a id="__codelineno-3-37" name="__codelineno-3-37" href="#__codelineno-3-37"></a>echo &quot;$(date &#39;+%Y-%m-%d %H:%M:%S&#39;) INFO: Done - Downloading 8-channel WGS model checkpoint (with InsertSize, with AlleleFreq)&quot;
<a id="__codelineno-3-38" name="__codelineno-3-38" href="#__codelineno-3-38"></a>
<a id="__codelineno-3-39" name="__codelineno-3-39" href="#__codelineno-3-39"></a>##========================================================================##
<a id="__codelineno-3-40" name="__codelineno-3-40" href="#__codelineno-3-40"></a>
<a id="__codelineno-3-41" name="__codelineno-3-41" href="#__codelineno-3-41"></a>echo -e &quot;=== scripts/setup/download_models.sh &gt; end $(date)&quot;
</code></pre></div>
</details>
<details class="success">
<summary>Check | Default human model</summary>
<div class="highlight"><span class="filename">Run the following at the command line:</span><pre><span></span><code><a id="__codelineno-4-1" name="__codelineno-4-1" href="#__codelineno-4-1"></a>ls<span class="w"> </span>triotrain/model_training/pretrained_models/v1.4.0_withIS_noAF/
</code></pre></div>
<div class="highlight"><span class="filename">Expected outputs:</span><pre><span></span><code><a id="__codelineno-5-1" name="__codelineno-5-1" href="#__codelineno-5-1"></a>model.ckpt.data-00000-of-00001<span class="w">  </span>model.ckpt.example_info.json<span class="w">  </span>model.ckpt.index<span class="w">  </span>model.ckpt.meta
</code></pre></div>
</details>
<details class="success">
<summary>Check | WGS.AF human model</summary>
<div class="highlight"><span class="filename">Run the following at the command line:</span><pre><span></span><code><a id="__codelineno-6-1" name="__codelineno-6-1" href="#__codelineno-6-1"></a>ls<span class="w"> </span>triotrain/model_training/pretrained_models/v1.4.0_withIS_withAF/
</code></pre></div>
<div class="highlight"><span class="filename">Expected outputs:</span><pre><span></span><code><a id="__codelineno-7-1" name="__codelineno-7-1" href="#__codelineno-7-1"></a>wgs_af.model.ckpt.data-00000-of-00001<span class="w">  </span>wgs_af.model.ckpt.index
<a id="__codelineno-7-2" name="__codelineno-7-2" href="#__codelineno-7-2"></a>wgs_af.model.ckpt.example_info.json<span class="w">    </span>wgs_af.model.ckpt.meta
</code></pre></div>
</details>
<h2 id="5-download-raw-data">5. Download Raw Data</h2>
<div class="admonition warning">
<p class="admonition-title">Warning | Download Size</p>
</div>
<p>We are using two Human trios from the v4.2.1 GIAB benchmarking data.</p>
<table>
<thead>
<tr>
<th>TrioNumber</th>
<th>TrioName</th>
<th>CommonID</th>
<th>SampleID</th>
<th>Relationship</th>
</tr>
</thead>
<tbody>
<tr>
<td>1</td>
<td>AshkenaziJew</td>
<td>HG002</td>
<td>NA24385</td>
<td>Son</td>
</tr>
<tr>
<td></td>
<td></td>
<td>HG003</td>
<td>NA24149</td>
<td>Father</td>
</tr>
<tr>
<td></td>
<td></td>
<td>HG004</td>
<td>NA24143</td>
<td>Mother</td>
</tr>
<tr>
<td>2</td>
<td>HanChinese</td>
<td>HG005</td>
<td>NA24631</td>
<td>Son</td>
</tr>
<tr>
<td></td>
<td></td>
<td>HG006</td>
<td>NA24694</td>
<td>Father</td>
</tr>
<tr>
<td></td>
<td></td>
<td>HG007</td>
<td>NA24695</td>
<td>Mother</td>
</tr>
</tbody>
</table>
<p>We need (5) types of raw data:</p>
<table>
<thead>
<tr>
<th>Number</th>
<th>Description</th>
<th>Extension</th>
</tr>
</thead>
<tbody>
<tr>
<td>1.</td>
<td>the GRCh38 reference genome</td>
<td><code>.fasta</code>, <code>.fai</code></td>
</tr>
<tr>
<td>2.</td>
<td>1kGP Population Allele Frequency, with index file</td>
<td><code>.vcf.gz</code>, <code>vcf.gz.tbi</code></td>
</tr>
<tr>
<td>3.</td>
<td>benchmarking files</td>
<td></td>
</tr>
<tr>
<td></td>
<td>per-genome truth callsets, with index files</td>
<td><code>.vcf.gz</code>, <code>vcf.gz.tbi</code></td>
</tr>
<tr>
<td></td>
<td>per genome truth regions file</td>
<td><code>.bed</code></td>
</tr>
<tr>
<td>4.</td>
<td>sample metadata</td>
<td></td>
</tr>
<tr>
<td></td>
<td>checksums file</td>
<td><code>.md5</code></td>
</tr>
<tr>
<td></td>
<td>index files</td>
<td><code>.txt</code></td>
</tr>
<tr>
<td>5.</td>
<td>the aligned reads files, with index file</td>
<td><code>.bam</code>, <code>.bai</code></td>
</tr>
</tbody>
</table>
<div class="highlight"><span class="filename">Run the following at the command line:</span><pre><span></span><code><a id="__codelineno-8-1" name="__codelineno-8-1" href="#__codelineno-8-1"></a>bash<span class="w"> </span>scripts/setup/download_GIAB.sh
</code></pre></div>
<details class="example">
<summary>Example | <code>download_GIAB.sh</code></summary>
<div class="highlight"><pre><span></span><code><a id="__codelineno-9-1" name="__codelineno-9-1" href="#__codelineno-9-1"></a>#!/bin/bash
<a id="__codelineno-9-2" name="__codelineno-9-2" href="#__codelineno-9-2"></a># scripts/setup/download_GIAB.sh
<a id="__codelineno-9-3" name="__codelineno-9-3" href="#__codelineno-9-3"></a>echo -e &quot;=== scripts/setup/download_GIAB.sh &gt; start $(date)&quot;
<a id="__codelineno-9-4" name="__codelineno-9-4" href="#__codelineno-9-4"></a>
<a id="__codelineno-9-5" name="__codelineno-9-5" href="#__codelineno-9-5"></a># NOTES --------------------------------------------------------------#
<a id="__codelineno-9-6" name="__codelineno-9-6" href="#__codelineno-9-6"></a># The DV developers use the following data in their walk-through docs:
<a id="__codelineno-9-7" name="__codelineno-9-7" href="#__codelineno-9-7"></a># 1) Reference Genome Version: GRCh38_no_alt_analysis_set.fasta
<a id="__codelineno-9-8" name="__codelineno-9-8" href="#__codelineno-9-8"></a># 2) Benchmark Sample: HGOO3 [AshkenazimTrio-Father], typically only Chr20.
<a id="__codelineno-9-9" name="__codelineno-9-9" href="#__codelineno-9-9"></a># 3) PopVCF (without genotypes): v3_missing2ref
<a id="__codelineno-9-10" name="__codelineno-9-10" href="#__codelineno-9-10"></a>
<a id="__codelineno-9-11" name="__codelineno-9-11" href="#__codelineno-9-11"></a># However, for a direct comparision between our cattle model and previous models, we need to calculate Mendelian Inheritance Errors (MIE). Therefore, we will be downloading the parents in this trio. We&#39;ll also need to download the reference PopVCF used to build the humanWGS_AF model.
<a id="__codelineno-9-12" name="__codelineno-9-12" href="#__codelineno-9-12"></a>
<a id="__codelineno-9-13" name="__codelineno-9-13" href="#__codelineno-9-13"></a>#-------------------------------------------------------------------#
<a id="__codelineno-9-14" name="__codelineno-9-14" href="#__codelineno-9-14"></a>#                            GIAB Trio1                             #
<a id="__codelineno-9-15" name="__codelineno-9-15" href="#__codelineno-9-15"></a>#-------------------------------------------------------------------#
<a id="__codelineno-9-16" name="__codelineno-9-16" href="#__codelineno-9-16"></a># Known as the Ashkenazi Jew Trio
<a id="__codelineno-9-17" name="__codelineno-9-17" href="#__codelineno-9-17"></a># From Personal Genome Project
<a id="__codelineno-9-18" name="__codelineno-9-18" href="#__codelineno-9-18"></a># AJ Son = HG002_NA24385_son
<a id="__codelineno-9-19" name="__codelineno-9-19" href="#__codelineno-9-19"></a># AJ Father = HG003_NA24149_father
<a id="__codelineno-9-20" name="__codelineno-9-20" href="#__codelineno-9-20"></a># AJ Mother = HG004_NA24143_mother
<a id="__codelineno-9-21" name="__codelineno-9-21" href="#__codelineno-9-21"></a>
<a id="__codelineno-9-22" name="__codelineno-9-22" href="#__codelineno-9-22"></a>#-------------------------------------------------------------------#
<a id="__codelineno-9-23" name="__codelineno-9-23" href="#__codelineno-9-23"></a>#                            GIAB Trio2                             #
<a id="__codelineno-9-24" name="__codelineno-9-24" href="#__codelineno-9-24"></a>#-------------------------------------------------------------------#
<a id="__codelineno-9-25" name="__codelineno-9-25" href="#__codelineno-9-25"></a># Known as the Han Chinese Trio
<a id="__codelineno-9-26" name="__codelineno-9-26" href="#__codelineno-9-26"></a># From Personal Genome Project
<a id="__codelineno-9-27" name="__codelineno-9-27" href="#__codelineno-9-27"></a># Chinese Son = HG005_NA24631_son
<a id="__codelineno-9-28" name="__codelineno-9-28" href="#__codelineno-9-28"></a># Chinese Father = HG006_NA24694_father
<a id="__codelineno-9-29" name="__codelineno-9-29" href="#__codelineno-9-29"></a># Chinese Mother = HG007_NA24695_mother
<a id="__codelineno-9-30" name="__codelineno-9-30" href="#__codelineno-9-30"></a>
<a id="__codelineno-9-31" name="__codelineno-9-31" href="#__codelineno-9-31"></a># A preprint describing these calls is at https://doi.org/10.1101/2020.07.24.212712.  The paper(s) above can be cited for use of the benchmark, and please cite http://www.nature.com/articles/sdata201625 (doi:10.1038/sdata.2016.25) when using the corresponding sequencing data.
<a id="__codelineno-9-32" name="__codelineno-9-32" href="#__codelineno-9-32"></a>
<a id="__codelineno-9-33" name="__codelineno-9-33" href="#__codelineno-9-33"></a>cd triotrain/variant_calling/data/GIAB
<a id="__codelineno-9-34" name="__codelineno-9-34" href="#__codelineno-9-34"></a>
<a id="__codelineno-9-35" name="__codelineno-9-35" href="#__codelineno-9-35"></a>##===================================================================
<a id="__codelineno-9-36" name="__codelineno-9-36" href="#__codelineno-9-36"></a>##                         General Files                             
<a id="__codelineno-9-37" name="__codelineno-9-37" href="#__codelineno-9-37"></a>##===================================================================
<a id="__codelineno-9-38" name="__codelineno-9-38" href="#__codelineno-9-38"></a>
<a id="__codelineno-9-39" name="__codelineno-9-39" href="#__codelineno-9-39"></a>## --------------- Download Reference Genome ------------------------
<a id="__codelineno-9-40" name="__codelineno-9-40" href="#__codelineno-9-40"></a>## NOTES ----
<a id="__codelineno-9-41" name="__codelineno-9-41" href="#__codelineno-9-41"></a>## Files Downloaded Include:
<a id="__codelineno-9-42" name="__codelineno-9-42" href="#__codelineno-9-42"></a>## 1. GRCh38 reference genome [FASTA]
<a id="__codelineno-9-43" name="__codelineno-9-43" href="#__codelineno-9-43"></a>## 2. GRCh38 index [FAI]
<a id="__codelineno-9-44" name="__codelineno-9-44" href="#__codelineno-9-44"></a>
<a id="__codelineno-9-45" name="__codelineno-9-45" href="#__codelineno-9-45"></a># Create the output dir
<a id="__codelineno-9-46" name="__codelineno-9-46" href="#__codelineno-9-46"></a>install --directory --verbose reference
<a id="__codelineno-9-47" name="__codelineno-9-47" href="#__codelineno-9-47"></a>
<a id="__codelineno-9-48" name="__codelineno-9-48" href="#__codelineno-9-48"></a>REFDIR=ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/000/001/405/GCA_000001405.15_GRCh38/seqs_for_alignment_pipelines.ucsc_ids/
<a id="__codelineno-9-49" name="__codelineno-9-49" href="#__codelineno-9-49"></a>REFFILE=GCA_000001405.15_GRCh38_no_alt_analysis_set
<a id="__codelineno-9-50" name="__codelineno-9-50" href="#__codelineno-9-50"></a>
<a id="__codelineno-9-51" name="__codelineno-9-51" href="#__codelineno-9-51"></a>if [ ! -f ./reference/md5checksums.txt ]; then
<a id="__codelineno-9-52" name="__codelineno-9-52" href="#__codelineno-9-52"></a>    echo &quot;$(date &#39;+%Y-%m-%d %H:%M:%S&#39;) INFO: downloading GRCh38 checksum now...&quot;
<a id="__codelineno-9-53" name="__codelineno-9-53" href="#__codelineno-9-53"></a>    curl -s --continue-at - ${REFDIR}/md5checksums.txt -o ./reference/md5checksums.txt
<a id="__codelineno-9-54" name="__codelineno-9-54" href="#__codelineno-9-54"></a>else
<a id="__codelineno-9-55" name="__codelineno-9-55" href="#__codelineno-9-55"></a>    echo &quot;$(date &#39;+%Y-%m-%d %H:%M:%S&#39;) INFO: file found | &#39;./reference/md5checksums.txt&#39;&quot;
<a id="__codelineno-9-56" name="__codelineno-9-56" href="#__codelineno-9-56"></a>fi
<a id="__codelineno-9-57" name="__codelineno-9-57" href="#__codelineno-9-57"></a>
<a id="__codelineno-9-58" name="__codelineno-9-58" href="#__codelineno-9-58"></a># Define the file extensions to be downloaded:
<a id="__codelineno-9-59" name="__codelineno-9-59" href="#__codelineno-9-59"></a>declare -a Ext=(&quot;.fna.gz&quot; &quot;.fna.fai&quot;)
<a id="__codelineno-9-60" name="__codelineno-9-60" href="#__codelineno-9-60"></a>
<a id="__codelineno-9-61" name="__codelineno-9-61" href="#__codelineno-9-61"></a>if [ ! -f &quot;./reference/GRCh38_no_alt_analysis_set.fasta&quot; ]; then 
<a id="__codelineno-9-62" name="__codelineno-9-62" href="#__codelineno-9-62"></a>    for e in ${Ext[@]}; do
<a id="__codelineno-9-63" name="__codelineno-9-63" href="#__codelineno-9-63"></a>        if [ -f ./reference/md5checksums.txt ]; then
<a id="__codelineno-9-64" name="__codelineno-9-64" href="#__codelineno-9-64"></a>            echo &quot;$(date &#39;+%Y-%m-%d %H:%M:%S&#39;) INFO: downloading ${REFFILE}${e} now...&quot;
<a id="__codelineno-9-65" name="__codelineno-9-65" href="#__codelineno-9-65"></a>            curl -s --continue-at - &quot;${REFDIR}/${REFFILE}${e}&quot; -o &quot;./reference/${REFFILE}${e}&quot; 
<a id="__codelineno-9-66" name="__codelineno-9-66" href="#__codelineno-9-66"></a>            echo &quot;$(date &#39;+%Y-%m-%d %H:%M:%S&#39;) INFO: checking ${REFFILE}${e} for corruption...&quot;
<a id="__codelineno-9-67" name="__codelineno-9-67" href="#__codelineno-9-67"></a>            check_sum=$(cat ./reference/md5checksums.txt | grep &quot;${REFFILE}${e}&quot;)
<a id="__codelineno-9-68" name="__codelineno-9-68" href="#__codelineno-9-68"></a>            old_path=&quot;./&quot;
<a id="__codelineno-9-69" name="__codelineno-9-69" href="#__codelineno-9-69"></a>            new_path=&quot;./reference/&quot;
<a id="__codelineno-9-70" name="__codelineno-9-70" href="#__codelineno-9-70"></a>            valid_check_sum=&quot;${check_sum/$old_path/$new_path}&quot;
<a id="__codelineno-9-71" name="__codelineno-9-71" href="#__codelineno-9-71"></a>            echo &quot;$(date &#39;+%Y-%m-%d %H:%M:%S&#39;) INFO: $(echo $valid_check_sum | md5sum -c)&quot;
<a id="__codelineno-9-72" name="__codelineno-9-72" href="#__codelineno-9-72"></a>        else
<a id="__codelineno-9-73" name="__codelineno-9-73" href="#__codelineno-9-73"></a>            echo &quot;$(date &#39;+%Y-%m-%d %H:%M:%S&#39;) INFO: file found | &#39;${REFFILE}${e}&#39;&quot;
<a id="__codelineno-9-74" name="__codelineno-9-74" href="#__codelineno-9-74"></a>        fi
<a id="__codelineno-9-75" name="__codelineno-9-75" href="#__codelineno-9-75"></a>    done
<a id="__codelineno-9-76" name="__codelineno-9-76" href="#__codelineno-9-76"></a>fi
<a id="__codelineno-9-77" name="__codelineno-9-77" href="#__codelineno-9-77"></a>
<a id="__codelineno-9-78" name="__codelineno-9-78" href="#__codelineno-9-78"></a>if [ -f &quot;./reference/${REFFILE}.fna.gz&quot; ]; then
<a id="__codelineno-9-79" name="__codelineno-9-79" href="#__codelineno-9-79"></a>    echo &quot;$(date &#39;+%Y-%m-%d %H:%M:%S&#39;) INFO: Unzipping GRCh38 and re-naming reference files now...&quot;
<a id="__codelineno-9-80" name="__codelineno-9-80" href="#__codelineno-9-80"></a>    gunzip -c &quot;./reference/${REFFILE}.fna.gz&quot; &gt; &quot;./reference/GRCh38_no_alt_analysis_set.fasta&quot;
<a id="__codelineno-9-81" name="__codelineno-9-81" href="#__codelineno-9-81"></a>    rm &quot;./reference/${REFFILE}.fna.gz&quot;
<a id="__codelineno-9-82" name="__codelineno-9-82" href="#__codelineno-9-82"></a>    mv &quot;./reference/${REFFILE}.fna.fai&quot; ./reference/GRCh38_no_alt_analysis_set.fasta.fai
<a id="__codelineno-9-83" name="__codelineno-9-83" href="#__codelineno-9-83"></a>else
<a id="__codelineno-9-84" name="__codelineno-9-84" href="#__codelineno-9-84"></a>    echo &quot;$(date &#39;+%Y-%m-%d %H:%M:%S&#39;) INFO: reference re-named already... SKIPPING AHEAD&quot;
<a id="__codelineno-9-85" name="__codelineno-9-85" href="#__codelineno-9-85"></a>fi
<a id="__codelineno-9-86" name="__codelineno-9-86" href="#__codelineno-9-86"></a>
<a id="__codelineno-9-87" name="__codelineno-9-87" href="#__codelineno-9-87"></a>## --------------- Download Population VCF --------------------------
<a id="__codelineno-9-88" name="__codelineno-9-88" href="#__codelineno-9-88"></a>## NOTES ----
<a id="__codelineno-9-89" name="__codelineno-9-89" href="#__codelineno-9-89"></a>## Files Downloaded Include:
<a id="__codelineno-9-90" name="__codelineno-9-90" href="#__codelineno-9-90"></a>## 1. One-thousand genomes (1kGP) allele frequencies [VCF]
<a id="__codelineno-9-91" name="__codelineno-9-91" href="#__codelineno-9-91"></a>## 2. Allele-frequencies index [TBI]
<a id="__codelineno-9-92" name="__codelineno-9-92" href="#__codelineno-9-92"></a>## These data are ~940GiB to download completly.
<a id="__codelineno-9-93" name="__codelineno-9-93" href="#__codelineno-9-93"></a>## THERE ARE TWO PAGES OF FILES! (v)
<a id="__codelineno-9-94" name="__codelineno-9-94" href="#__codelineno-9-94"></a>## The newest version of WGS_AF model can be viewed on GCP here: https://console.cloud.google.com/storage/browser/brain-genomics-public/research/cohort/1KGP/cohort_dv_glnexus_opt/v3_missing2ref?pageState=(%22StorageObjectListTable%22:(%22f%22:%22%255B%255D%22))&amp;prefix=&amp;forceOnObjectsSortingFiltering=false
<a id="__codelineno-9-95" name="__codelineno-9-95" href="#__codelineno-9-95"></a>
<a id="__codelineno-9-96" name="__codelineno-9-96" href="#__codelineno-9-96"></a># Create the output dir
<a id="__codelineno-9-97" name="__codelineno-9-97" href="#__codelineno-9-97"></a>mkdir -p allele_freq
<a id="__codelineno-9-98" name="__codelineno-9-98" href="#__codelineno-9-98"></a>
<a id="__codelineno-9-99" name="__codelineno-9-99" href="#__codelineno-9-99"></a># Define where to get the AF data
<a id="__codelineno-9-100" name="__codelineno-9-100" href="#__codelineno-9-100"></a>AF_DIR=&quot;https://storage.googleapis.com/brain-genomics-public/research/cohort/1KGP/cohort_dv_glnexus_opt/v3_missing2ref&quot;
<a id="__codelineno-9-101" name="__codelineno-9-101" href="#__codelineno-9-101"></a>
<a id="__codelineno-9-102" name="__codelineno-9-102" href="#__codelineno-9-102"></a>for i in {{1..22},X,Y}
<a id="__codelineno-9-103" name="__codelineno-9-103" href="#__codelineno-9-103"></a>do  
<a id="__codelineno-9-104" name="__codelineno-9-104" href="#__codelineno-9-104"></a>    if [ ! -f ./allele_freq/cohort-chr${i}.release_missing2ref.no_calls.vcf.gz ]; then
<a id="__codelineno-9-105" name="__codelineno-9-105" href="#__codelineno-9-105"></a>        echo &quot;$(date &#39;+%Y-%m-%d %H:%M:%S&#39;) INFO: downloading chr${i} PopVCF...&quot;
<a id="__codelineno-9-106" name="__codelineno-9-106" href="#__codelineno-9-106"></a>        curl -s --continue-at - ${AF_DIR}/cohort-chr${i}.release_missing2ref.no_calls.vcf.gz -o ./allele_freq/cohort-chr${i}.release_missing2ref.no_calls.vcf.gz 
<a id="__codelineno-9-107" name="__codelineno-9-107" href="#__codelineno-9-107"></a>    else
<a id="__codelineno-9-108" name="__codelineno-9-108" href="#__codelineno-9-108"></a>        echo &quot;$(date &#39;+%Y-%m-%d %H:%M:%S&#39;) INFO: file found | &#39;chr${i} PopVCF&#39;&quot;
<a id="__codelineno-9-109" name="__codelineno-9-109" href="#__codelineno-9-109"></a>    fi
<a id="__codelineno-9-110" name="__codelineno-9-110" href="#__codelineno-9-110"></a>
<a id="__codelineno-9-111" name="__codelineno-9-111" href="#__codelineno-9-111"></a>    if [ ! -f ./allele_freq/cohort-chr${i}.release_missing2ref.no_calls.vcf.gz.tbi ]; then
<a id="__codelineno-9-112" name="__codelineno-9-112" href="#__codelineno-9-112"></a>        echo &quot;$(date &#39;+%Y-%m-%d %H:%M:%S&#39;) INFO: downloading chr${i} index ...&quot; 
<a id="__codelineno-9-113" name="__codelineno-9-113" href="#__codelineno-9-113"></a>        curl -s --continue-at - ${AF_DIR}/cohort-chr${i}.release_missing2ref.no_calls.vcf.gz.tbi -o ./allele_freq/cohort-chr${i}.release_missing2ref.no_calls.vcf.gz.tbi 
<a id="__codelineno-9-114" name="__codelineno-9-114" href="#__codelineno-9-114"></a>    else
<a id="__codelineno-9-115" name="__codelineno-9-115" href="#__codelineno-9-115"></a>        echo &quot;$(date &#39;+%Y-%m-%d %H:%M:%S&#39;) INFO: file found | &#39;chr${i} index&#39;&quot;
<a id="__codelineno-9-116" name="__codelineno-9-116" href="#__codelineno-9-116"></a>    fi
<a id="__codelineno-9-117" name="__codelineno-9-117" href="#__codelineno-9-117"></a>done
<a id="__codelineno-9-118" name="__codelineno-9-118" href="#__codelineno-9-118"></a>
<a id="__codelineno-9-119" name="__codelineno-9-119" href="#__codelineno-9-119"></a># Merge the chr AF into a genome-wide AF
<a id="__codelineno-9-120" name="__codelineno-9-120" href="#__codelineno-9-120"></a>if [ ! -f &quot;./allele_freq/PopVCF.merge.list&quot; ]; then 
<a id="__codelineno-9-121" name="__codelineno-9-121" href="#__codelineno-9-121"></a>    for i in {{1..22},X,Y}
<a id="__codelineno-9-122" name="__codelineno-9-122" href="#__codelineno-9-122"></a>    do
<a id="__codelineno-9-123" name="__codelineno-9-123" href="#__codelineno-9-123"></a>        echo &quot;././triotrain/variant_calling/data/GIAB/allele_freq/cohort-chr$i.release_missing2ref.no_calls.vcf.gz&quot; &gt;&gt; ./allele_freq/PopVCF.merge.list
<a id="__codelineno-9-124" name="__codelineno-9-124" href="#__codelineno-9-124"></a>    done
<a id="__codelineno-9-125" name="__codelineno-9-125" href="#__codelineno-9-125"></a>fi
<a id="__codelineno-9-126" name="__codelineno-9-126" href="#__codelineno-9-126"></a>
<a id="__codelineno-9-127" name="__codelineno-9-127" href="#__codelineno-9-127"></a>if [ ! -f &quot;./allele_freq/concat_PopVCFs.sh&quot; ]; then 
<a id="__codelineno-9-128" name="__codelineno-9-128" href="#__codelineno-9-128"></a>echo -e &quot;source ./scripts/setup/modules.sh
<a id="__codelineno-9-129" name="__codelineno-9-129" href="#__codelineno-9-129"></a>bcftools concat --file-list ./triotrain/variant_calling/data/GIAB/allele_freq/PopVCF.merge.list -Oz -o ./triotrain/variant_calling/data/GIAB/allele_freq/cohort.release_missing2ref.no_calls.vcf.gz
<a id="__codelineno-9-130" name="__codelineno-9-130" href="#__codelineno-9-130"></a>bcftools index ./triotrain/variant_calling/data/GIAB/allele_freq/cohort.release_missing2ref.no_calls.vcf.gz&quot; &gt; ./allele_freq/concat_PopVCFs.sh
<a id="__codelineno-9-131" name="__codelineno-9-131" href="#__codelineno-9-131"></a>fi
<a id="__codelineno-9-132" name="__codelineno-9-132" href="#__codelineno-9-132"></a>
<a id="__codelineno-9-133" name="__codelineno-9-133" href="#__codelineno-9-133"></a>##===================================================================
<a id="__codelineno-9-134" name="__codelineno-9-134" href="#__codelineno-9-134"></a>##                       Trio-Specific Files                             
<a id="__codelineno-9-135" name="__codelineno-9-135" href="#__codelineno-9-135"></a>##===================================================================
<a id="__codelineno-9-136" name="__codelineno-9-136" href="#__codelineno-9-136"></a>
<a id="__codelineno-9-137" name="__codelineno-9-137" href="#__codelineno-9-137"></a>## --------------- Download GIAB Truth Files ------------------------
<a id="__codelineno-9-138" name="__codelineno-9-138" href="#__codelineno-9-138"></a>## NOTES ---- 
<a id="__codelineno-9-139" name="__codelineno-9-139" href="#__codelineno-9-139"></a>## Currently using v4.2.1
<a id="__codelineno-9-140" name="__codelineno-9-140" href="#__codelineno-9-140"></a>## Files Downloaded Include:
<a id="__codelineno-9-141" name="__codelineno-9-141" href="#__codelineno-9-141"></a>## 1. benchmarking region files [BED]
<a id="__codelineno-9-142" name="__codelineno-9-142" href="#__codelineno-9-142"></a>## 2. benchmarking genotypes [VCF]
<a id="__codelineno-9-143" name="__codelineno-9-143" href="#__codelineno-9-143"></a>## 3. benchmarking genotype index [TBI]
<a id="__codelineno-9-144" name="__codelineno-9-144" href="#__codelineno-9-144"></a>## The benchmarking files from NIST can be found here: https://ftp-trace.ncbi.nlm.nih.gov/ReferenceSamples/giab/release/
<a id="__codelineno-9-145" name="__codelineno-9-145" href="#__codelineno-9-145"></a>
<a id="__codelineno-9-146" name="__codelineno-9-146" href="#__codelineno-9-146"></a># Create output dir
<a id="__codelineno-9-147" name="__codelineno-9-147" href="#__codelineno-9-147"></a>mkdir -p benchmark
<a id="__codelineno-9-148" name="__codelineno-9-148" href="#__codelineno-9-148"></a>
<a id="__codelineno-9-149" name="__codelineno-9-149" href="#__codelineno-9-149"></a># Define where to get the truth data
<a id="__codelineno-9-150" name="__codelineno-9-150" href="#__codelineno-9-150"></a>TRUTHDIR=https://ftp-trace.ncbi.nlm.nih.gov/ReferenceSamples/giab/release
<a id="__codelineno-9-151" name="__codelineno-9-151" href="#__codelineno-9-151"></a>benchmark_version=&quot;_GRCh38_1_22_v4.2.1_benchmark&quot;
<a id="__codelineno-9-152" name="__codelineno-9-152" href="#__codelineno-9-152"></a>bench_ext=(&quot;_noinconsistent.bed&quot; &quot;.vcf.gz&quot; &quot;.vcf.gz.tbi&quot;)
<a id="__codelineno-9-153" name="__codelineno-9-153" href="#__codelineno-9-153"></a>
<a id="__codelineno-9-154" name="__codelineno-9-154" href="#__codelineno-9-154"></a>#-------------------------------------------------------------------#
<a id="__codelineno-9-155" name="__codelineno-9-155" href="#__codelineno-9-155"></a>#                            GIAB Trio1                             #
<a id="__codelineno-9-156" name="__codelineno-9-156" href="#__codelineno-9-156"></a>#-------------------------------------------------------------------#
<a id="__codelineno-9-157" name="__codelineno-9-157" href="#__codelineno-9-157"></a>declare -A Trio1=([&quot;HG002&quot;]=&quot;HG002_NA24385_son&quot; [&quot;HG003&quot;]=&quot;HG003_NA24149_father&quot; [&quot;HG004&quot;]=&quot;HG004_NA24143_mother&quot;)
<a id="__codelineno-9-158" name="__codelineno-9-158" href="#__codelineno-9-158"></a>for t in ${!Trio1[@]}; do
<a id="__codelineno-9-159" name="__codelineno-9-159" href="#__codelineno-9-159"></a>    echo &quot;$(date &#39;+%Y-%m-%d %H:%M:%S&#39;) INFO: downloading ${t}=${Trio1[${t}]} benchmarking files now...&quot;
<a id="__codelineno-9-160" name="__codelineno-9-160" href="#__codelineno-9-160"></a>    for e in ${!bench_ext[@]}; do
<a id="__codelineno-9-161" name="__codelineno-9-161" href="#__codelineno-9-161"></a>        if [ ! -f &quot;./benchmark/${t}${benchmark_version}${bench_ext[${e}]}&quot; ]; then
<a id="__codelineno-9-162" name="__codelineno-9-162" href="#__codelineno-9-162"></a>            echo &quot;$(date &#39;+%Y-%m-%d %H:%M:%S&#39;) INFO: downloading ${t}${benchmark_version}${bench_ext[${e}]}...&quot;
<a id="__codelineno-9-163" name="__codelineno-9-163" href="#__codelineno-9-163"></a>            curl --continue-at - ${TRUTHDIR}/AshkenazimTrio/${Trio1[${t}]}/NISTv4.2.1/GRCh38/${t}${benchmark_version}${bench_ext[${e}]} -o ./benchmark/${t}${benchmark_version}${bench_ext[${e}]}
<a id="__codelineno-9-164" name="__codelineno-9-164" href="#__codelineno-9-164"></a>            # No MD5 listed on NIST FTP SITE!
<a id="__codelineno-9-165" name="__codelineno-9-165" href="#__codelineno-9-165"></a>        else
<a id="__codelineno-9-166" name="__codelineno-9-166" href="#__codelineno-9-166"></a>            echo &quot;$(date &#39;+%Y-%m-%d %H:%M:%S&#39;) INFO: file found | &#39;${t}${benchmark_version}${bench_ext[${e}]}&#39;&quot;
<a id="__codelineno-9-167" name="__codelineno-9-167" href="#__codelineno-9-167"></a>        fi
<a id="__codelineno-9-168" name="__codelineno-9-168" href="#__codelineno-9-168"></a>    done
<a id="__codelineno-9-169" name="__codelineno-9-169" href="#__codelineno-9-169"></a>
<a id="__codelineno-9-170" name="__codelineno-9-170" href="#__codelineno-9-170"></a>    # Download the README
<a id="__codelineno-9-171" name="__codelineno-9-171" href="#__codelineno-9-171"></a>    if [ ! -f &quot;./benchmark/${t}_README_v4.2.1.txt&quot; ]; then
<a id="__codelineno-9-172" name="__codelineno-9-172" href="#__codelineno-9-172"></a>        echo &quot;$(date &#39;+%Y-%m-%d %H:%M:%S&#39;) INFO: downloading ${t} readme...&quot;
<a id="__codelineno-9-173" name="__codelineno-9-173" href="#__codelineno-9-173"></a>        curl -s --continue-at - ${TRUTHDIR}/AshkenazimTrio/${Trio1[${t}]}/NISTv4.2.1/README_v4.2.1.txt -o ./benchmark/${t}_README_v4.2.1.txt 
<a id="__codelineno-9-174" name="__codelineno-9-174" href="#__codelineno-9-174"></a>    else
<a id="__codelineno-9-175" name="__codelineno-9-175" href="#__codelineno-9-175"></a>        echo &quot;$(date &#39;+%Y-%m-%d %H:%M:%S&#39;) INFO: file found | &#39;${t}_README_v4.2.1.txt&#39;&quot;
<a id="__codelineno-9-176" name="__codelineno-9-176" href="#__codelineno-9-176"></a>    fi
<a id="__codelineno-9-177" name="__codelineno-9-177" href="#__codelineno-9-177"></a>done
<a id="__codelineno-9-178" name="__codelineno-9-178" href="#__codelineno-9-178"></a>
<a id="__codelineno-9-179" name="__codelineno-9-179" href="#__codelineno-9-179"></a>#-------------------------------------------------------------------#
<a id="__codelineno-9-180" name="__codelineno-9-180" href="#__codelineno-9-180"></a>#                            GIAB Trio2                             #
<a id="__codelineno-9-181" name="__codelineno-9-181" href="#__codelineno-9-181"></a>#-------------------------------------------------------------------#
<a id="__codelineno-9-182" name="__codelineno-9-182" href="#__codelineno-9-182"></a>bench_ext=(&quot;.bed&quot; &quot;.vcf.gz&quot; &quot;.vcf.gz.tbi&quot;)
<a id="__codelineno-9-183" name="__codelineno-9-183" href="#__codelineno-9-183"></a>declare -A Trio2=([&quot;HG005&quot;]=&quot;HG005_NA24631_son&quot; [&quot;HG006&quot;]=&quot;HG006_NA24694_father&quot; [&quot;HG007&quot;]=&quot;HG007_NA24695_mother&quot;)
<a id="__codelineno-9-184" name="__codelineno-9-184" href="#__codelineno-9-184"></a>
<a id="__codelineno-9-185" name="__codelineno-9-185" href="#__codelineno-9-185"></a>for t in ${!Trio2[@]}; do
<a id="__codelineno-9-186" name="__codelineno-9-186" href="#__codelineno-9-186"></a>    echo &quot;$(date &#39;+%Y-%m-%d %H:%M:%S&#39;) INFO: downloading ${t}=${Trio2[${t}]} benchmarking files now...&quot;
<a id="__codelineno-9-187" name="__codelineno-9-187" href="#__codelineno-9-187"></a>
<a id="__codelineno-9-188" name="__codelineno-9-188" href="#__codelineno-9-188"></a>    # Download the MD5
<a id="__codelineno-9-189" name="__codelineno-9-189" href="#__codelineno-9-189"></a>    if [ ! -f &quot;./benchmark/${t}_benchmark.md5&quot; ]; then
<a id="__codelineno-9-190" name="__codelineno-9-190" href="#__codelineno-9-190"></a>        echo &quot;$(date &#39;+%Y-%m-%d %H:%M:%S&#39;) INFO: downloading ${t} MD5...&quot;
<a id="__codelineno-9-191" name="__codelineno-9-191" href="#__codelineno-9-191"></a>        curl -s --continue-at - ${TRUTHDIR}/ChineseTrio/${Trio2[${t}]}/NISTv4.2.1/md5.in -o ./benchmark/${t}_benchmark.md5 
<a id="__codelineno-9-192" name="__codelineno-9-192" href="#__codelineno-9-192"></a>    else
<a id="__codelineno-9-193" name="__codelineno-9-193" href="#__codelineno-9-193"></a>        echo &quot;$(date &#39;+%Y-%m-%d %H:%M:%S&#39;) INFO: file found | &#39;${t}_benchmark.md5&#39;&quot;
<a id="__codelineno-9-194" name="__codelineno-9-194" href="#__codelineno-9-194"></a>    fi
<a id="__codelineno-9-195" name="__codelineno-9-195" href="#__codelineno-9-195"></a>
<a id="__codelineno-9-196" name="__codelineno-9-196" href="#__codelineno-9-196"></a>    for e in ${!bench_ext[@]}; do
<a id="__codelineno-9-197" name="__codelineno-9-197" href="#__codelineno-9-197"></a>        if [ ! -f &quot;./benchmark/${t}${benchmark_version}${bench_ext[${e}]}&quot; ]; then
<a id="__codelineno-9-198" name="__codelineno-9-198" href="#__codelineno-9-198"></a>            echo &quot;$(date &#39;+%Y-%m-%d %H:%M:%S&#39;) INFO: downloading ${t}${benchmark_version}${bench_ext[${e}]}...&quot;
<a id="__codelineno-9-199" name="__codelineno-9-199" href="#__codelineno-9-199"></a>            curl --continue-at - ${TRUTHDIR}/ChineseTrio/${Trio2[${t}]}/NISTv4.2.1/GRCh38/${t}${benchmark_version}${bench_ext[${e}]} -o ./benchmark/${t}${benchmark_version}${bench_ext[${e}]}
<a id="__codelineno-9-200" name="__codelineno-9-200" href="#__codelineno-9-200"></a>
<a id="__codelineno-9-201" name="__codelineno-9-201" href="#__codelineno-9-201"></a>            echo &quot;$(date &#39;+%Y-%m-%d %H:%M:%S&#39;) INFO: checking ./benchmark/${t}${benchmark_version}${bench_ext[${e}]} for corruption...&quot;
<a id="__codelineno-9-202" name="__codelineno-9-202" href="#__codelineno-9-202"></a>            check_sum=$(cat ./benchmark/${t}_benchmark.md5 | grep &quot;${t}${benchmark_version}${bench_ext[${e}]}&quot; | head -n1)
<a id="__codelineno-9-203" name="__codelineno-9-203" href="#__codelineno-9-203"></a>            old_path=&quot;GRCh38/&quot;
<a id="__codelineno-9-204" name="__codelineno-9-204" href="#__codelineno-9-204"></a>            new_path=&quot;./benchmark/&quot;
<a id="__codelineno-9-205" name="__codelineno-9-205" href="#__codelineno-9-205"></a>            valid_check_sum=&quot;${check_sum/$old_path/$new_path}&quot;
<a id="__codelineno-9-206" name="__codelineno-9-206" href="#__codelineno-9-206"></a>            msg=$(echo $valid_check_sum | md5sum -c -)
<a id="__codelineno-9-207" name="__codelineno-9-207" href="#__codelineno-9-207"></a>            if [ $? -eq 0 ]; then 
<a id="__codelineno-9-208" name="__codelineno-9-208" href="#__codelineno-9-208"></a>                echo $(date &#39;+%Y-%m-%d %H:%M:%S&#39;) INFO: $msg || error_exit $msg tracker_file.txt
<a id="__codelineno-9-209" name="__codelineno-9-209" href="#__codelineno-9-209"></a>            fi
<a id="__codelineno-9-210" name="__codelineno-9-210" href="#__codelineno-9-210"></a>        else
<a id="__codelineno-9-211" name="__codelineno-9-211" href="#__codelineno-9-211"></a>            echo &quot;$(date &#39;+%Y-%m-%d %H:%M:%S&#39;) INFO: file found | &#39;${t}${benchmark_version}${bench_ext[${e}]}&#39;&quot;
<a id="__codelineno-9-212" name="__codelineno-9-212" href="#__codelineno-9-212"></a>        fi
<a id="__codelineno-9-213" name="__codelineno-9-213" href="#__codelineno-9-213"></a>    done
<a id="__codelineno-9-214" name="__codelineno-9-214" href="#__codelineno-9-214"></a>
<a id="__codelineno-9-215" name="__codelineno-9-215" href="#__codelineno-9-215"></a>    # Download the README
<a id="__codelineno-9-216" name="__codelineno-9-216" href="#__codelineno-9-216"></a>    if [ ! -f &quot;./benchmark/${t}_README_v4.2.1.txt&quot; ]; then
<a id="__codelineno-9-217" name="__codelineno-9-217" href="#__codelineno-9-217"></a>        echo &quot;$(date &#39;+%Y-%m-%d %H:%M:%S&#39;) INFO: downloading ${t} readme...&quot;
<a id="__codelineno-9-218" name="__codelineno-9-218" href="#__codelineno-9-218"></a>        curl -s --continue-at - ${TRUTHDIR}/ChineseTrio/${Trio2[${t}]}/NISTv4.2.1/README_v4.2.1.txt -o ./benchmark/${t}_README_v4.2.1.txt 
<a id="__codelineno-9-219" name="__codelineno-9-219" href="#__codelineno-9-219"></a>    else
<a id="__codelineno-9-220" name="__codelineno-9-220" href="#__codelineno-9-220"></a>        echo &quot;$(date &#39;+%Y-%m-%d %H:%M:%S&#39;) INFO: file found | &#39;${t}_README_v4.2.1.txt&#39;&quot;
<a id="__codelineno-9-221" name="__codelineno-9-221" href="#__codelineno-9-221"></a>    fi
<a id="__codelineno-9-222" name="__codelineno-9-222" href="#__codelineno-9-222"></a>
<a id="__codelineno-9-223" name="__codelineno-9-223" href="#__codelineno-9-223"></a>done
<a id="__codelineno-9-224" name="__codelineno-9-224" href="#__codelineno-9-224"></a>
<a id="__codelineno-9-225" name="__codelineno-9-225" href="#__codelineno-9-225"></a>## --------------- Download GIAB Aligned Reads ----------------------
<a id="__codelineno-9-226" name="__codelineno-9-226" href="#__codelineno-9-226"></a>## NOTES ---- 
<a id="__codelineno-9-227" name="__codelineno-9-227" href="#__codelineno-9-227"></a>## Files Downloaded Include:
<a id="__codelineno-9-228" name="__codelineno-9-228" href="#__codelineno-9-228"></a>## 1. NIST Illumina 2x250bp sequence reads alligned to GRCh38 [BAM]
<a id="__codelineno-9-229" name="__codelineno-9-229" href="#__codelineno-9-229"></a>## 2. Sequence reads index [BAI]
<a id="__codelineno-9-230" name="__codelineno-9-230" href="#__codelineno-9-230"></a>## 3. MD5 checksum [MD5]
<a id="__codelineno-9-231" name="__codelineno-9-231" href="#__codelineno-9-231"></a>## Indexes for various sequencing data and aligned files can be found here: https://github.com/genome-in-a-bottle/giab_data_indexes
<a id="__codelineno-9-232" name="__codelineno-9-232" href="#__codelineno-9-232"></a>
<a id="__codelineno-9-233" name="__codelineno-9-233" href="#__codelineno-9-233"></a>## Function for downloading BAM + BAI files -----------
<a id="__codelineno-9-234" name="__codelineno-9-234" href="#__codelineno-9-234"></a>download () {
<a id="__codelineno-9-235" name="__codelineno-9-235" href="#__codelineno-9-235"></a>    file_input=$1
<a id="__codelineno-9-236" name="__codelineno-9-236" href="#__codelineno-9-236"></a>
<a id="__codelineno-9-237" name="__codelineno-9-237" href="#__codelineno-9-237"></a>    # split the file_path into an array
<a id="__codelineno-9-238" name="__codelineno-9-238" href="#__codelineno-9-238"></a>    IFS=&#39;_&#39; read -r -a result &lt;&lt;&lt; &quot;$file_input&quot;
<a id="__codelineno-9-239" name="__codelineno-9-239" href="#__codelineno-9-239"></a>    trio_name=${result[0]}
<a id="__codelineno-9-240" name="__codelineno-9-240" href="#__codelineno-9-240"></a>
<a id="__codelineno-9-241" name="__codelineno-9-241" href="#__codelineno-9-241"></a>    if [ ! -f &quot;./bam/${trio_name}.download&quot; ]; then 
<a id="__codelineno-9-242" name="__codelineno-9-242" href="#__codelineno-9-242"></a>        # echo &quot;source ./scripts/setup/modules.sh&quot; &gt; ./bam/$trio_name.download
<a id="__codelineno-9-243" name="__codelineno-9-243" href="#__codelineno-9-243"></a>        # Command Translation:
<a id="__codelineno-9-244" name="__codelineno-9-244" href="#__codelineno-9-244"></a>
<a id="__codelineno-9-245" name="__codelineno-9-245" href="#__codelineno-9-245"></a>        # For each line after the header row:
<a id="__codelineno-9-246" name="__codelineno-9-246" href="#__codelineno-9-246"></a>        # First, use the first column in each row,
<a id="__codelineno-9-247" name="__codelineno-9-247" href="#__codelineno-9-247"></a>        # from the .txt file input, and
<a id="__codelineno-9-248" name="__codelineno-9-248" href="#__codelineno-9-248"></a>        # extract the BAM file path end, called &quot;filename&quot;, using
<a id="__codelineno-9-249" name="__codelineno-9-249" href="#__codelineno-9-249"></a>        # the shell cmd called &quot;basename&quot;, then
<a id="__codelineno-9-250" name="__codelineno-9-250" href="#__codelineno-9-250"></a>        # close that shell command
<a id="__codelineno-9-251" name="__codelineno-9-251" href="#__codelineno-9-251"></a>
<a id="__codelineno-9-252" name="__codelineno-9-252" href="#__codelineno-9-252"></a>        # Next, if &quot;filename&quot; containes the correct reference genome name,
<a id="__codelineno-9-253" name="__codelineno-9-253" href="#__codelineno-9-253"></a>        # print a command to download that file
<a id="__codelineno-9-254" name="__codelineno-9-254" href="#__codelineno-9-254"></a>        # with timestamp wrappers before/after
<a id="__codelineno-9-255" name="__codelineno-9-255" href="#__codelineno-9-255"></a>
<a id="__codelineno-9-256" name="__codelineno-9-256" href="#__codelineno-9-256"></a>        # Next, process the MD5 flexibly...
<a id="__codelineno-9-257" name="__codelineno-9-257" href="#__codelineno-9-257"></a>        # For HG002, use the corrected MD5 sum (output)
<a id="__codelineno-9-258" name="__codelineno-9-258" href="#__codelineno-9-258"></a>        # For all others, use the original MD5 sum ($2)
<a id="__codelineno-9-259" name="__codelineno-9-259" href="#__codelineno-9-259"></a>
<a id="__codelineno-9-260" name="__codelineno-9-260" href="#__codelineno-9-260"></a>        # Next, write the correct MD5 + filepath to a new file, and
<a id="__codelineno-9-261" name="__codelineno-9-261" href="#__codelineno-9-261"></a>        # print the command for checking the MD5 sum 
<a id="__codelineno-9-262" name="__codelineno-9-262" href="#__codelineno-9-262"></a>
<a id="__codelineno-9-263" name="__codelineno-9-263" href="#__codelineno-9-263"></a>        # Next, use the 3rd column in each row, 
<a id="__codelineno-9-264" name="__codelineno-9-264" href="#__codelineno-9-264"></a>        # from the .txt file input, and
<a id="__codelineno-9-265" name="__codelineno-9-265" href="#__codelineno-9-265"></a>        # extract the BAI file path end, called &quot;secondfile&quot;, using
<a id="__codelineno-9-266" name="__codelineno-9-266" href="#__codelineno-9-266"></a>        # the shell cmd called &quot;basename&quot;, then
<a id="__codelineno-9-267" name="__codelineno-9-267" href="#__codelineno-9-267"></a>        # close that shell command
<a id="__codelineno-9-268" name="__codelineno-9-268" href="#__codelineno-9-268"></a>
<a id="__codelineno-9-269" name="__codelineno-9-269" href="#__codelineno-9-269"></a>        # Next, if &quot;filename&quot; containes the correct reference genome name,
<a id="__codelineno-9-270" name="__codelineno-9-270" href="#__codelineno-9-270"></a>        # print a command to download that file
<a id="__codelineno-9-271" name="__codelineno-9-271" href="#__codelineno-9-271"></a>        # with timestamp wrappers before/after
<a id="__codelineno-9-272" name="__codelineno-9-272" href="#__codelineno-9-272"></a>
<a id="__codelineno-9-273" name="__codelineno-9-273" href="#__codelineno-9-273"></a>        # Next, process the MD5 flexibly...
<a id="__codelineno-9-274" name="__codelineno-9-274" href="#__codelineno-9-274"></a>        # For HG002, use the corrected MD5 sum (output)
<a id="__codelineno-9-275" name="__codelineno-9-275" href="#__codelineno-9-275"></a>        # For all others, use the original MD5 sum ($2)
<a id="__codelineno-9-276" name="__codelineno-9-276" href="#__codelineno-9-276"></a>
<a id="__codelineno-9-277" name="__codelineno-9-277" href="#__codelineno-9-277"></a>        # Next, write the correct MD5 + filepath to a new file, and
<a id="__codelineno-9-278" name="__codelineno-9-278" href="#__codelineno-9-278"></a>        # print the command for checking the MD5 sum.
<a id="__codelineno-9-279" name="__codelineno-9-279" href="#__codelineno-9-279"></a>
<a id="__codelineno-9-280" name="__codelineno-9-280" href="#__codelineno-9-280"></a>        awk -F &#39;\t&#39; &#39;NR!=1 { 
<a id="__codelineno-9-281" name="__codelineno-9-281" href="#__codelineno-9-281"></a>            cmd = &quot;basename &quot; $1
<a id="__codelineno-9-282" name="__codelineno-9-282" href="#__codelineno-9-282"></a>            cmd | getline filename
<a id="__codelineno-9-283" name="__codelineno-9-283" href="#__codelineno-9-283"></a>            close(cmd)
<a id="__codelineno-9-284" name="__codelineno-9-284" href="#__codelineno-9-284"></a>            }
<a id="__codelineno-9-285" name="__codelineno-9-285" href="#__codelineno-9-285"></a>
<a id="__codelineno-9-286" name="__codelineno-9-286" href="#__codelineno-9-286"></a>            index(filename, &quot;GRCh38&quot;) {
<a id="__codelineno-9-287" name="__codelineno-9-287" href="#__codelineno-9-287"></a>                print &quot;echo $(date \&quot;+%Y-%m-%d %H:%M:%S\&quot;) INFO: downloading [&quot;filename&quot;] now...&quot; 
<a id="__codelineno-9-288" name="__codelineno-9-288" href="#__codelineno-9-288"></a>                print &quot;curl -o ./triotrain/variant_calling/data/GIAB/bam/&quot;filename&quot; -C - &quot;$1&quot; --keepalive-time 300&quot;
<a id="__codelineno-9-289" name="__codelineno-9-289" href="#__codelineno-9-289"></a>                print &quot;echo $(date \&quot;+%Y-%m-%d %H:%M:%S\&quot;) INFO: downloading [&quot;filename&quot;]... done&quot;
<a id="__codelineno-9-290" name="__codelineno-9-290" href="#__codelineno-9-290"></a>                }
<a id="__codelineno-9-291" name="__codelineno-9-291" href="#__codelineno-9-291"></a>
<a id="__codelineno-9-292" name="__codelineno-9-292" href="#__codelineno-9-292"></a>                filename ~ /HG002.GRCh38/ {
<a id="__codelineno-9-293" name="__codelineno-9-293" href="#__codelineno-9-293"></a>                    cmd = &quot;grep -Fw &quot;filename&quot; ./bam/HG002_corrected_md5sums.feb19upload.txt&quot; 
<a id="__codelineno-9-294" name="__codelineno-9-294" href="#__codelineno-9-294"></a>                    cmd | getline result
<a id="__codelineno-9-295" name="__codelineno-9-295" href="#__codelineno-9-295"></a>                    close(cmd)
<a id="__codelineno-9-296" name="__codelineno-9-296" href="#__codelineno-9-296"></a>                    split(result, output, &quot; &quot;)
<a id="__codelineno-9-297" name="__codelineno-9-297" href="#__codelineno-9-297"></a>                    print &quot;echo &quot;output[1]&quot;\t./triotrain/variant_calling/data/GIAB/bam/&quot;filename&quot; &gt; ./triotrain/variant_calling/data/GIAB/bam/&quot;filename&quot;.md5&quot;
<a id="__codelineno-9-298" name="__codelineno-9-298" href="#__codelineno-9-298"></a>                }
<a id="__codelineno-9-299" name="__codelineno-9-299" href="#__codelineno-9-299"></a>
<a id="__codelineno-9-300" name="__codelineno-9-300" href="#__codelineno-9-300"></a>                filename !~ /HG002/ &amp;&amp; filename ~ /GRCh38/ {
<a id="__codelineno-9-301" name="__codelineno-9-301" href="#__codelineno-9-301"></a>                    print &quot;echo &quot;$2&quot;\t./triotrain/variant_calling/data/GIAB/bam/&quot;filename&quot; &gt; ./triotrain/variant_calling/data/GIAB/bam/&quot;filename&quot;.md5&quot;
<a id="__codelineno-9-302" name="__codelineno-9-302" href="#__codelineno-9-302"></a>                }
<a id="__codelineno-9-303" name="__codelineno-9-303" href="#__codelineno-9-303"></a>
<a id="__codelineno-9-304" name="__codelineno-9-304" href="#__codelineno-9-304"></a>            index(filename, &quot;GRCh38&quot;) {
<a id="__codelineno-9-305" name="__codelineno-9-305" href="#__codelineno-9-305"></a>                print &quot;echo $(date \&quot;+%Y-%m-%d %H:%M:%S\&quot;) INFO: checking [&quot;filename&quot;] for corruption...&quot;
<a id="__codelineno-9-306" name="__codelineno-9-306" href="#__codelineno-9-306"></a>                print &quot;md5sum -c ./triotrain/variant_calling/data/GIAB/bam/&quot;filename&quot;.md5&quot;
<a id="__codelineno-9-307" name="__codelineno-9-307" href="#__codelineno-9-307"></a>                print &quot;echo $(date \&quot;+%Y-%m-%d %H:%M:%S\&quot;) INFO: checking [&quot;filename&quot;] for corruption... done&quot;
<a id="__codelineno-9-308" name="__codelineno-9-308" href="#__codelineno-9-308"></a>
<a id="__codelineno-9-309" name="__codelineno-9-309" href="#__codelineno-9-309"></a>                print &quot;echo ----------------------------------------------------&quot;
<a id="__codelineno-9-310" name="__codelineno-9-310" href="#__codelineno-9-310"></a>            }
<a id="__codelineno-9-311" name="__codelineno-9-311" href="#__codelineno-9-311"></a>
<a id="__codelineno-9-312" name="__codelineno-9-312" href="#__codelineno-9-312"></a>            {
<a id="__codelineno-9-313" name="__codelineno-9-313" href="#__codelineno-9-313"></a>                cmd = &quot;basename &quot; $3
<a id="__codelineno-9-314" name="__codelineno-9-314" href="#__codelineno-9-314"></a>                cmd | getline secondfile
<a id="__codelineno-9-315" name="__codelineno-9-315" href="#__codelineno-9-315"></a>                close(cmd)
<a id="__codelineno-9-316" name="__codelineno-9-316" href="#__codelineno-9-316"></a>            }
<a id="__codelineno-9-317" name="__codelineno-9-317" href="#__codelineno-9-317"></a>
<a id="__codelineno-9-318" name="__codelineno-9-318" href="#__codelineno-9-318"></a>            index(secondfile, &quot;GRCh38&quot;) {
<a id="__codelineno-9-319" name="__codelineno-9-319" href="#__codelineno-9-319"></a>                print &quot;echo $(date \&quot;+%Y-%m-%d %H:%M:%S\&quot;) INFO: downloading [&quot;secondfile&quot;] now...&quot;  
<a id="__codelineno-9-320" name="__codelineno-9-320" href="#__codelineno-9-320"></a>                print &quot;curl -o ./triotrain/variant_calling/data/GIAB/bam/&quot;secondfile&quot; -C - &quot;$3
<a id="__codelineno-9-321" name="__codelineno-9-321" href="#__codelineno-9-321"></a>                print &quot;echo $(date \&quot;+%Y-%m-%d %H:%M:%S\&quot;) INFO: downloading [&quot;secondfile&quot;] ... done&quot;
<a id="__codelineno-9-322" name="__codelineno-9-322" href="#__codelineno-9-322"></a>                }
<a id="__codelineno-9-323" name="__codelineno-9-323" href="#__codelineno-9-323"></a>
<a id="__codelineno-9-324" name="__codelineno-9-324" href="#__codelineno-9-324"></a>                secondfile ~ /HG002.GRCh38/ {
<a id="__codelineno-9-325" name="__codelineno-9-325" href="#__codelineno-9-325"></a>                    cmd = &quot;grep -Fw &quot;secondfile&quot; ./bam/HG002_corrected_md5sums.feb19upload.txt&quot; 
<a id="__codelineno-9-326" name="__codelineno-9-326" href="#__codelineno-9-326"></a>                    cmd | getline result
<a id="__codelineno-9-327" name="__codelineno-9-327" href="#__codelineno-9-327"></a>                    close(cmd)
<a id="__codelineno-9-328" name="__codelineno-9-328" href="#__codelineno-9-328"></a>                    split(result, output, &quot; &quot;)
<a id="__codelineno-9-329" name="__codelineno-9-329" href="#__codelineno-9-329"></a>                    print &quot;echo &quot;output[1]&quot;\t./triotrain/variant_calling/data/GIAB/bam/&quot;secondfile&quot; &gt; ./triotrain/variant_calling/data/GIAB/bam/&quot;secondfile&quot;.md5&quot;
<a id="__codelineno-9-330" name="__codelineno-9-330" href="#__codelineno-9-330"></a>                }
<a id="__codelineno-9-331" name="__codelineno-9-331" href="#__codelineno-9-331"></a>
<a id="__codelineno-9-332" name="__codelineno-9-332" href="#__codelineno-9-332"></a>                secondfile !~ /HG002/ &amp;&amp; secondfile ~ /GRCh38/ {
<a id="__codelineno-9-333" name="__codelineno-9-333" href="#__codelineno-9-333"></a>                    print &quot;echo &quot;$4&quot;\t./triotrain/variant_calling/data/GIAB/bam/&quot;secondfile&quot; &gt; ./triotrain/variant_calling/data/GIAB/bam/&quot;secondfile&quot;.md5&quot;
<a id="__codelineno-9-334" name="__codelineno-9-334" href="#__codelineno-9-334"></a>                }
<a id="__codelineno-9-335" name="__codelineno-9-335" href="#__codelineno-9-335"></a>
<a id="__codelineno-9-336" name="__codelineno-9-336" href="#__codelineno-9-336"></a>            index(secondfile, &quot;GRCh38&quot;) {
<a id="__codelineno-9-337" name="__codelineno-9-337" href="#__codelineno-9-337"></a>                print &quot;echo $(date \&quot;+%Y-%m-%d %H:%M:%S\&quot;) INFO: checking [&quot;secondfile&quot;] for corruption...&quot;
<a id="__codelineno-9-338" name="__codelineno-9-338" href="#__codelineno-9-338"></a>                print &quot;md5sum -c ./triotrain/variant_calling/data/GIAB/bam/&quot;secondfile&quot;.md5&quot;
<a id="__codelineno-9-339" name="__codelineno-9-339" href="#__codelineno-9-339"></a>                print &quot;echo $(date \&quot;+%Y-%m-%d %H:%M:%S\&quot;) INFO: checking [&quot;secondfile&quot;] for corruption... done&quot;
<a id="__codelineno-9-340" name="__codelineno-9-340" href="#__codelineno-9-340"></a>
<a id="__codelineno-9-341" name="__codelineno-9-341" href="#__codelineno-9-341"></a>                print &quot;echo ====================================================&quot;
<a id="__codelineno-9-342" name="__codelineno-9-342" href="#__codelineno-9-342"></a>            }&#39; &quot;./bam/${file_input}&quot;  &gt;&gt; &quot;./bam/${trio_name}.download&quot;
<a id="__codelineno-9-343" name="__codelineno-9-343" href="#__codelineno-9-343"></a>    else
<a id="__codelineno-9-344" name="__codelineno-9-344" href="#__codelineno-9-344"></a>        echo &quot;$(date &#39;+%Y-%m-%d %H:%M:%S&#39;) INFO: file found | &#39;./triotrain/variant_calling/data/GIAB/bam/${trio_name}.download&#39;&quot;
<a id="__codelineno-9-345" name="__codelineno-9-345" href="#__codelineno-9-345"></a>    fi
<a id="__codelineno-9-346" name="__codelineno-9-346" href="#__codelineno-9-346"></a>    }
<a id="__codelineno-9-347" name="__codelineno-9-347" href="#__codelineno-9-347"></a>
<a id="__codelineno-9-348" name="__codelineno-9-348" href="#__codelineno-9-348"></a>## Function for calculating AVG COVERAGE -----------
<a id="__codelineno-9-349" name="__codelineno-9-349" href="#__codelineno-9-349"></a>calc_cov () {
<a id="__codelineno-9-350" name="__codelineno-9-350" href="#__codelineno-9-350"></a>    file_input=$1
<a id="__codelineno-9-351" name="__codelineno-9-351" href="#__codelineno-9-351"></a>
<a id="__codelineno-9-352" name="__codelineno-9-352" href="#__codelineno-9-352"></a>    # split the file_path into an array
<a id="__codelineno-9-353" name="__codelineno-9-353" href="#__codelineno-9-353"></a>    IFS=&#39;_&#39; read -r -a result &lt;&lt;&lt; &quot;$file_input&quot;
<a id="__codelineno-9-354" name="__codelineno-9-354" href="#__codelineno-9-354"></a>    trio_name=${result[0]}
<a id="__codelineno-9-355" name="__codelineno-9-355" href="#__codelineno-9-355"></a>
<a id="__codelineno-9-356" name="__codelineno-9-356" href="#__codelineno-9-356"></a>    if [ ! -f &quot;./bam/${trio_name}.run&quot; ]; then
<a id="__codelineno-9-357" name="__codelineno-9-357" href="#__codelineno-9-357"></a>        # echo &quot;source ./scripts/setup/modules.sh&quot; &gt; ./bam/${trio_name}.run
<a id="__codelineno-9-358" name="__codelineno-9-358" href="#__codelineno-9-358"></a>
<a id="__codelineno-9-359" name="__codelineno-9-359" href="#__codelineno-9-359"></a>        # Command Translation:
<a id="__codelineno-9-360" name="__codelineno-9-360" href="#__codelineno-9-360"></a>        # For each line after the header row:
<a id="__codelineno-9-361" name="__codelineno-9-361" href="#__codelineno-9-361"></a>        # First, use the first column in each row,
<a id="__codelineno-9-362" name="__codelineno-9-362" href="#__codelineno-9-362"></a>        # from the .txt file input, and
<a id="__codelineno-9-363" name="__codelineno-9-363" href="#__codelineno-9-363"></a>        # extract the BAM file path end, called &quot;filename&quot;, using
<a id="__codelineno-9-364" name="__codelineno-9-364" href="#__codelineno-9-364"></a>        # the shell cmd called &quot;basename&quot;, then
<a id="__codelineno-9-365" name="__codelineno-9-365" href="#__codelineno-9-365"></a>        # close that shell command
<a id="__codelineno-9-366" name="__codelineno-9-366" href="#__codelineno-9-366"></a>
<a id="__codelineno-9-367" name="__codelineno-9-367" href="#__codelineno-9-367"></a>        # Next, if &quot;filename&quot; containes the correct reference genome name,
<a id="__codelineno-9-368" name="__codelineno-9-368" href="#__codelineno-9-368"></a>        # print a if/then statement to test if job file to run coverage 
<a id="__codelineno-9-369" name="__codelineno-9-369" href="#__codelineno-9-369"></a>        # calculations exists, and create it if it does not.
<a id="__codelineno-9-370" name="__codelineno-9-370" href="#__codelineno-9-370"></a>
<a id="__codelineno-9-371" name="__codelineno-9-371" href="#__codelineno-9-371"></a>        # print a if/then statement to test if output from avg_coverage 
<a id="__codelineno-9-372" name="__codelineno-9-372" href="#__codelineno-9-372"></a>        # calculations exists, and create it if it does not.
<a id="__codelineno-9-373" name="__codelineno-9-373" href="#__codelineno-9-373"></a>
<a id="__codelineno-9-374" name="__codelineno-9-374" href="#__codelineno-9-374"></a>        awk -F &#39;\t&#39; &#39;NR!=1 { 
<a id="__codelineno-9-375" name="__codelineno-9-375" href="#__codelineno-9-375"></a>            cmd = &quot;basename &quot; $1 
<a id="__codelineno-9-376" name="__codelineno-9-376" href="#__codelineno-9-376"></a>            cmd | getline filename
<a id="__codelineno-9-377" name="__codelineno-9-377" href="#__codelineno-9-377"></a>            close(cmd)
<a id="__codelineno-9-378" name="__codelineno-9-378" href="#__codelineno-9-378"></a>            }
<a id="__codelineno-9-379" name="__codelineno-9-379" href="#__codelineno-9-379"></a>
<a id="__codelineno-9-380" name="__codelineno-9-380" href="#__codelineno-9-380"></a>            index(filename, &quot;GRCh38&quot;) {
<a id="__codelineno-9-381" name="__codelineno-9-381" href="#__codelineno-9-381"></a>                cmd = &quot;basename &quot; filename &quot; .bam&quot;
<a id="__codelineno-9-382" name="__codelineno-9-382" href="#__codelineno-9-382"></a>                cmd | getline label
<a id="__codelineno-9-383" name="__codelineno-9-383" href="#__codelineno-9-383"></a>                close(cmd)
<a id="__codelineno-9-384" name="__codelineno-9-384" href="#__codelineno-9-384"></a>
<a id="__codelineno-9-385" name="__codelineno-9-385" href="#__codelineno-9-385"></a>                print &quot;if [ ! -f ./triotrain/variant_calling/data/GIAB/bam/&quot;label&quot;.coverage.out ]; then&quot;
<a id="__codelineno-9-386" name="__codelineno-9-386" href="#__codelineno-9-386"></a>                print &quot;    echo \&quot;$(date \&quot;+%Y-%m-%d %H:%M:%S\&quot;)  INFO: Calculating Coverage for [&quot;label&quot;] now...\&quot;&quot; 
<a id="__codelineno-9-387" name="__codelineno-9-387" href="#__codelineno-9-387"></a>                print &quot;    samtools coverage ./triotrain/variant_calling/data/GIAB/bam/&quot;filename&quot; --output ./triotrain/variant_calling/data/GIAB/bam/&quot;label&quot;.coverage.out&quot;
<a id="__codelineno-9-388" name="__codelineno-9-388" href="#__codelineno-9-388"></a>                print &quot;    echo \&quot;$(date \&quot;+%Y-%m-%d %H:%M:%S\&quot;)  INFO: Calculating Coverage for [&quot;label&quot;] now... done\&quot;&quot;
<a id="__codelineno-9-389" name="__codelineno-9-389" href="#__codelineno-9-389"></a>                print &quot;else&quot;
<a id="__codelineno-9-390" name="__codelineno-9-390" href="#__codelineno-9-390"></a>                print &quot;    echo \&quot;$(date \&quot;+%Y-%m-%d %H:%M:%S\&quot;)  INFO: bash file found | ./triotrain/variant_calling/data/GIAB/bam/&quot;label&quot;.coverage.out\&quot;&quot;
<a id="__codelineno-9-391" name="__codelineno-9-391" href="#__codelineno-9-391"></a>                print &quot;fi&quot;
<a id="__codelineno-9-392" name="__codelineno-9-392" href="#__codelineno-9-392"></a>                print &quot;echo ----------------------------------------------------&quot;
<a id="__codelineno-9-393" name="__codelineno-9-393" href="#__codelineno-9-393"></a>                split(label, output, &quot;.&quot;)
<a id="__codelineno-9-394" name="__codelineno-9-394" href="#__codelineno-9-394"></a>                print &quot;if [ ! -f ./triotrain/variant_calling/data/GIAB/bam/&quot;output[1]&quot;.avg_coverage.out ]; then&quot;
<a id="__codelineno-9-395" name="__codelineno-9-395" href="#__codelineno-9-395"></a>                print &quot;    echo \&quot;$(date \&quot;+%Y-%m-%d %H:%M:%S\&quot;)  INFO: Calculating Average Coverage for [&quot;output[1]&quot;] now...\&quot;&quot; 
<a id="__codelineno-9-396" name="__codelineno-9-396" href="#__codelineno-9-396"></a>                print &quot;    awk &#39;\&#39;&#39;{ total += $6; count++ } END { print \&quot;&quot;label&quot; AVERAGE COVERAGE = \&quot; total/count }&#39;\&#39;&#39; ./triotrain/variant_calling/data/GIAB/bam/&quot;label&quot;.coverage.out &gt; ./triotrain/variant_calling/data/GIAB/bam/&quot;output[1]&quot;.avg_coverage.out&quot;
<a id="__codelineno-9-397" name="__codelineno-9-397" href="#__codelineno-9-397"></a>                print &quot;    echo \&quot;$(date \&quot;+%Y-%m-%d %H:%M:%S\&quot;)  INFO: Calculating Average Coverage for [&quot;output[1]&quot;] now... done\&quot;&quot;
<a id="__codelineno-9-398" name="__codelineno-9-398" href="#__codelineno-9-398"></a>                print &quot;else&quot;
<a id="__codelineno-9-399" name="__codelineno-9-399" href="#__codelineno-9-399"></a>                print &quot;    echo \&quot;$(date \&quot;+%Y-%m-%d %H:%M:%S\&quot;)  INFO: output file found | ./triotrain/variant_calling/data/GIAB/bam/&quot;output[1]&quot;.avg_coverage.out\&quot;&quot;
<a id="__codelineno-9-400" name="__codelineno-9-400" href="#__codelineno-9-400"></a>                print &quot;fi&quot; 
<a id="__codelineno-9-401" name="__codelineno-9-401" href="#__codelineno-9-401"></a>
<a id="__codelineno-9-402" name="__codelineno-9-402" href="#__codelineno-9-402"></a>                print &quot;echo ====================================================&quot;
<a id="__codelineno-9-403" name="__codelineno-9-403" href="#__codelineno-9-403"></a>                } &#39; &quot;./bam/${file_input}&quot;  &gt;&gt; &quot;./bam/${trio_name}.run&quot;
<a id="__codelineno-9-404" name="__codelineno-9-404" href="#__codelineno-9-404"></a>
<a id="__codelineno-9-405" name="__codelineno-9-405" href="#__codelineno-9-405"></a>        # if running on an interactive session with lots of memory, 
<a id="__codelineno-9-406" name="__codelineno-9-406" href="#__codelineno-9-406"></a>        # uncomment the line below
<a id="__codelineno-9-407" name="__codelineno-9-407" href="#__codelineno-9-407"></a>        # . ./bam/$1.run
<a id="__codelineno-9-408" name="__codelineno-9-408" href="#__codelineno-9-408"></a>    else
<a id="__codelineno-9-409" name="__codelineno-9-409" href="#__codelineno-9-409"></a>        echo &quot;$(date &#39;+%Y-%m-%d %H:%M:%S&#39;) INFO: file found | &#39;./triotrain/variant_calling/data/GIAB/bam/${trio_name}.run&#39;&quot;
<a id="__codelineno-9-410" name="__codelineno-9-410" href="#__codelineno-9-410"></a>    fi
<a id="__codelineno-9-411" name="__codelineno-9-411" href="#__codelineno-9-411"></a>    }
<a id="__codelineno-9-412" name="__codelineno-9-412" href="#__codelineno-9-412"></a>
<a id="__codelineno-9-413" name="__codelineno-9-413" href="#__codelineno-9-413"></a># output dir
<a id="__codelineno-9-414" name="__codelineno-9-414" href="#__codelineno-9-414"></a>mkdir -p bam
<a id="__codelineno-9-415" name="__codelineno-9-415" href="#__codelineno-9-415"></a>
<a id="__codelineno-9-416" name="__codelineno-9-416" href="#__codelineno-9-416"></a>#-------------------------------------------------------------------#
<a id="__codelineno-9-417" name="__codelineno-9-417" href="#__codelineno-9-417"></a>#                            GIAB Trio1                             #
<a id="__codelineno-9-418" name="__codelineno-9-418" href="#__codelineno-9-418"></a>#-------------------------------------------------------------------#
<a id="__codelineno-9-419" name="__codelineno-9-419" href="#__codelineno-9-419"></a># BAM and BAI with MD5 checksums can be found here: https://github.com/genome-in-a-bottle/giab_data_indexes/blob/c4d3b95c2ebf14c175151e4723f82e8980722e90/AshkenazimTrio/alignment.index.AJtrio_Illumina_2x250bps_novoalign_GRCh37_GRCh38_NHGRI_06062016
<a id="__codelineno-9-420" name="__codelineno-9-420" href="#__codelineno-9-420"></a>
<a id="__codelineno-9-421" name="__codelineno-9-421" href="#__codelineno-9-421"></a>## CHECKSUMS -----------
<a id="__codelineno-9-422" name="__codelineno-9-422" href="#__codelineno-9-422"></a>echo &quot;$(date &#39;+%Y-%m-%d %H:%M:%S&#39;) INFO: downloading AJ Trio checksum now...&quot;
<a id="__codelineno-9-423" name="__codelineno-9-423" href="#__codelineno-9-423"></a>curl -s -C - https://raw.githubusercontent.com/genome-in-a-bottle/giab_data_indexes/master/AshkenazimTrio/alignment.index.AJtrio_Illumina_2x250bps_novoalign_GRCh37_GRCh38_NHGRI_06062016 -o ./bam/AJtrio_Illumina_2x250bps_novoaligns_GRCh37_GRCh38.txt 
<a id="__codelineno-9-424" name="__codelineno-9-424" href="#__codelineno-9-424"></a>
<a id="__codelineno-9-425" name="__codelineno-9-425" href="#__codelineno-9-425"></a># NOTE: 07 JUNE 2023:
<a id="__codelineno-9-426" name="__codelineno-9-426" href="#__codelineno-9-426"></a># The checksum for HG002 does not match the ^ above MD5,
<a id="__codelineno-9-427" name="__codelineno-9-427" href="#__codelineno-9-427"></a># therefore, download the correct one. 
<a id="__codelineno-9-428" name="__codelineno-9-428" href="#__codelineno-9-428"></a>
<a id="__codelineno-9-429" name="__codelineno-9-429" href="#__codelineno-9-429"></a>curl -s -C - https://ftp.ncbi.nlm.nih.gov/ReferenceSamples/giab/data/AshkenazimTrio/HG002_NA24385_son/NIST_Illumina_2x250bps/novoalign_bams/HG002_corrected_md5sums.feb19upload.txt -o ./bam/HG002_corrected_md5sums.feb19upload.txt
<a id="__codelineno-9-430" name="__codelineno-9-430" href="#__codelineno-9-430"></a>echo &quot;$(date &#39;+%Y-%m-%d %H:%M:%S&#39;) INFO: downloading AJ Trio checksum now... done&quot;
<a id="__codelineno-9-431" name="__codelineno-9-431" href="#__codelineno-9-431"></a>
<a id="__codelineno-9-432" name="__codelineno-9-432" href="#__codelineno-9-432"></a>download &quot;AJtrio_Illumina_2x250bps_novoaligns_GRCh37_GRCh38.txt&quot;
<a id="__codelineno-9-433" name="__codelineno-9-433" href="#__codelineno-9-433"></a># calc_cov &quot;AJtrio_Illumina_2x250bps_novoaligns_GRCh37_GRCh38.txt&quot; 
<a id="__codelineno-9-434" name="__codelineno-9-434" href="#__codelineno-9-434"></a>
<a id="__codelineno-9-435" name="__codelineno-9-435" href="#__codelineno-9-435"></a>#-------------------------------------------------------------------#
<a id="__codelineno-9-436" name="__codelineno-9-436" href="#__codelineno-9-436"></a>#                            GIAB Trio2                             #
<a id="__codelineno-9-437" name="__codelineno-9-437" href="#__codelineno-9-437"></a>#-------------------------------------------------------------------#
<a id="__codelineno-9-438" name="__codelineno-9-438" href="#__codelineno-9-438"></a># BAM and BAI with MD5 checksums can be found here: https://github.com/genome-in-a-bottle/giab_data_indexes/blob/master/ChineseTrio/alignment.index.ChineseTrio_Illumina300X100X_wgs_novoalign_GRCh37_GRCh38_NHGRI_04062016 
<a id="__codelineno-9-439" name="__codelineno-9-439" href="#__codelineno-9-439"></a>
<a id="__codelineno-9-440" name="__codelineno-9-440" href="#__codelineno-9-440"></a>## CHECKSUMS -----------
<a id="__codelineno-9-441" name="__codelineno-9-441" href="#__codelineno-9-441"></a>echo &quot;$(date &#39;+%Y-%m-%d %H:%M:%S&#39;) INFO: downloading HC Trio checksum now...&quot;
<a id="__codelineno-9-442" name="__codelineno-9-442" href="#__codelineno-9-442"></a>curl -s -C - https://raw.githubusercontent.com/genome-in-a-bottle/giab_data_indexes/master/ChineseTrio/alignment.index.ChineseTrio_Illumina300X100X_wgs_novoalign_GRCh37_GRCh38_NHGRI_04062016 -o ./bam/HCtrio_Illumina300X100X_wgs_novoalign_GRCh37_GRCh38.txt 
<a id="__codelineno-9-443" name="__codelineno-9-443" href="#__codelineno-9-443"></a>echo &quot;$(date &#39;+%Y-%m-%d %H:%M:%S&#39;) INFO: downloading HC Trio checksum now... done&quot;
<a id="__codelineno-9-444" name="__codelineno-9-444" href="#__codelineno-9-444"></a>
<a id="__codelineno-9-445" name="__codelineno-9-445" href="#__codelineno-9-445"></a>download &quot;HCtrio_Illumina300X100X_wgs_novoalign_GRCh37_GRCh38.txt&quot;
<a id="__codelineno-9-446" name="__codelineno-9-446" href="#__codelineno-9-446"></a># calc_cov &quot;HCtrio_Illumina300X100X_wgs_novoalign_GRCh37_GRCh38.txt&quot;
<a id="__codelineno-9-447" name="__codelineno-9-447" href="#__codelineno-9-447"></a>
<a id="__codelineno-9-448" name="__codelineno-9-448" href="#__codelineno-9-448"></a>echo -e &quot;=== scripts/setup/download_humanGIABdata.sh &gt; end $(date)&quot;
</code></pre></div>
</details>
<details class="success">
<summary>Check | Data Directories</summary>
<div class="highlight"><span class="filename">Run the following at the command line:</span><pre><span></span><code><a id="__codelineno-10-1" name="__codelineno-10-1" href="#__codelineno-10-1"></a>ls<span class="w"> </span>triotrain/variant_calling/data/GIAB
</code></pre></div>
<div class="highlight"><span class="filename">Expected new directories:</span><pre><span></span><code><a id="__codelineno-11-1" name="__codelineno-11-1" href="#__codelineno-11-1"></a>allele_freq<span class="w">  </span>bam<span class="w">  </span>benchmark<span class="w">  </span>reference
</code></pre></div>
</details>
<details class="success">
<summary>Check | <code>allele_freq/</code></summary>
<div class="highlight"><span class="filename">Run the following at the command line:</span><pre><span></span><code><a id="__codelineno-12-1" name="__codelineno-12-1" href="#__codelineno-12-1"></a>ls<span class="w"> </span>triotrain/variant_calling/data/GIAB/allele_freq/
</code></pre></div>
<div class="highlight"><span class="filename">Expected outputs:</span><pre><span></span><code><a id="__codelineno-13-1" name="__codelineno-13-1" href="#__codelineno-13-1"></a>cohort-chr10.release_missing2ref.no_calls.vcf.gz<span class="w">      </span>cohort-chr21.release_missing2ref.no_calls.vcf.gz.tbi
<a id="__codelineno-13-2" name="__codelineno-13-2" href="#__codelineno-13-2"></a>cohort-chr10.release_missing2ref.no_calls.vcf.gz.tbi<span class="w">  </span>cohort-chr22.release_missing2ref.no_calls.vcf.gz
<a id="__codelineno-13-3" name="__codelineno-13-3" href="#__codelineno-13-3"></a>cohort-chr11.release_missing2ref.no_calls.vcf.gz<span class="w">      </span>cohort-chr22.release_missing2ref.no_calls.vcf.gz.tbi
<a id="__codelineno-13-4" name="__codelineno-13-4" href="#__codelineno-13-4"></a>cohort-chr11.release_missing2ref.no_calls.vcf.gz.tbi<span class="w">  </span>cohort-chr2.release_missing2ref.no_calls.vcf.gz
<a id="__codelineno-13-5" name="__codelineno-13-5" href="#__codelineno-13-5"></a>cohort-chr12.release_missing2ref.no_calls.vcf.gz<span class="w">      </span>cohort-chr2.release_missing2ref.no_calls.vcf.gz.tbi
<a id="__codelineno-13-6" name="__codelineno-13-6" href="#__codelineno-13-6"></a>cohort-chr12.release_missing2ref.no_calls.vcf.gz.tbi<span class="w">  </span>cohort-chr3.release_missing2ref.no_calls.vcf.gz
<a id="__codelineno-13-7" name="__codelineno-13-7" href="#__codelineno-13-7"></a>cohort-chr13.release_missing2ref.no_calls.vcf.gz<span class="w">      </span>cohort-chr3.release_missing2ref.no_calls.vcf.gz.tbi
<a id="__codelineno-13-8" name="__codelineno-13-8" href="#__codelineno-13-8"></a>cohort-chr13.release_missing2ref.no_calls.vcf.gz.tbi<span class="w">  </span>cohort-chr4.release_missing2ref.no_calls.vcf.gz
<a id="__codelineno-13-9" name="__codelineno-13-9" href="#__codelineno-13-9"></a>cohort-chr14.release_missing2ref.no_calls.vcf.gz<span class="w">      </span>cohort-chr4.release_missing2ref.no_calls.vcf.gz.tbi
<a id="__codelineno-13-10" name="__codelineno-13-10" href="#__codelineno-13-10"></a>cohort-chr14.release_missing2ref.no_calls.vcf.gz.tbi<span class="w">  </span>cohort-chr5.release_missing2ref.no_calls.vcf.gz
<a id="__codelineno-13-11" name="__codelineno-13-11" href="#__codelineno-13-11"></a>cohort-chr15.release_missing2ref.no_calls.vcf.gz<span class="w">      </span>cohort-chr5.release_missing2ref.no_calls.vcf.gz.tbi
<a id="__codelineno-13-12" name="__codelineno-13-12" href="#__codelineno-13-12"></a>cohort-chr15.release_missing2ref.no_calls.vcf.gz.tbi<span class="w">  </span>cohort-chr6.release_missing2ref.no_calls.vcf.gz
<a id="__codelineno-13-13" name="__codelineno-13-13" href="#__codelineno-13-13"></a>cohort-chr16.release_missing2ref.no_calls.vcf.gz<span class="w">      </span>cohort-chr6.release_missing2ref.no_calls.vcf.gz.tbi
<a id="__codelineno-13-14" name="__codelineno-13-14" href="#__codelineno-13-14"></a>cohort-chr16.release_missing2ref.no_calls.vcf.gz.tbi<span class="w">  </span>cohort-chr7.release_missing2ref.no_calls.vcf.gz
<a id="__codelineno-13-15" name="__codelineno-13-15" href="#__codelineno-13-15"></a>cohort-chr17.release_missing2ref.no_calls.vcf.gz<span class="w">      </span>cohort-chr7.release_missing2ref.no_calls.vcf.gz.tbi
<a id="__codelineno-13-16" name="__codelineno-13-16" href="#__codelineno-13-16"></a>cohort-chr17.release_missing2ref.no_calls.vcf.gz.tbi<span class="w">  </span>cohort-chr8.release_missing2ref.no_calls.vcf.gz
<a id="__codelineno-13-17" name="__codelineno-13-17" href="#__codelineno-13-17"></a>cohort-chr18.release_missing2ref.no_calls.vcf.gz<span class="w">      </span>cohort-chr8.release_missing2ref.no_calls.vcf.gz.tbi
<a id="__codelineno-13-18" name="__codelineno-13-18" href="#__codelineno-13-18"></a>cohort-chr18.release_missing2ref.no_calls.vcf.gz.tbi<span class="w">  </span>cohort-chr9.release_missing2ref.no_calls.vcf.gz
<a id="__codelineno-13-19" name="__codelineno-13-19" href="#__codelineno-13-19"></a>cohort-chr19.release_missing2ref.no_calls.vcf.gz<span class="w">      </span>cohort-chr9.release_missing2ref.no_calls.vcf.gz.tbi
<a id="__codelineno-13-20" name="__codelineno-13-20" href="#__codelineno-13-20"></a>cohort-chr19.release_missing2ref.no_calls.vcf.gz.tbi<span class="w">  </span>cohort-chrX.release_missing2ref.no_calls.vcf.gz
<a id="__codelineno-13-21" name="__codelineno-13-21" href="#__codelineno-13-21"></a>cohort-chr1.release_missing2ref.no_calls.vcf.gz<span class="w">       </span>cohort-chrX.release_missing2ref.no_calls.vcf.gz.tbi
<a id="__codelineno-13-22" name="__codelineno-13-22" href="#__codelineno-13-22"></a>cohort-chr1.release_missing2ref.no_calls.vcf.gz.tbi<span class="w">   </span>cohort-chrY.release_missing2ref.no_calls.vcf.gz
<a id="__codelineno-13-23" name="__codelineno-13-23" href="#__codelineno-13-23"></a>cohort-chr20.release_missing2ref.no_calls.vcf.gz<span class="w">      </span>cohort-chrY.release_missing2ref.no_calls.vcf.gz.tbi
<a id="__codelineno-13-24" name="__codelineno-13-24" href="#__codelineno-13-24"></a>cohort-chr20.release_missing2ref.no_calls.vcf.gz.tbi<span class="w">  </span>concat_PopVCFs.sh
<a id="__codelineno-13-25" name="__codelineno-13-25" href="#__codelineno-13-25"></a>cohort-chr21.release_missing2ref.no_calls.vcf.gz<span class="w">      </span>PopVCF.merge.list
</code></pre></div>
</details>
<details class="success">
<summary>Check | <code>bam/</code></summary>
<div class="highlight"><span class="filename">Run the following at the command line:</span><pre><span></span><code><a id="__codelineno-14-1" name="__codelineno-14-1" href="#__codelineno-14-1"></a>ls<span class="w"> </span>triotrain/variant_calling/data/GIAB/bam
</code></pre></div>
<div class="highlight"><span class="filename">Expected outputs:</span><pre><span></span><code><a id="__codelineno-15-1" name="__codelineno-15-1" href="#__codelineno-15-1"></a>AJtrio.download<span class="w">                                        </span>HCtrio.download<span class="w">  </span>
<a id="__codelineno-15-2" name="__codelineno-15-2" href="#__codelineno-15-2"></a>AJtrio_Illumina_2x250bps_novoaligns_GRCh37_GRCh38.txt<span class="w">  </span>HCtrio_Illumina300X100X_wgs_novoalign_GRCh37_GRCh38.txt
<a id="__codelineno-15-3" name="__codelineno-15-3" href="#__codelineno-15-3"></a>HG002_corrected_md5sums.feb19upload.txt
</code></pre></div>
</details>
<details class="success">
<summary>Check | <code>benchmark/</code></summary>
<div class="highlight"><span class="filename">Run the following at the command line:</span><pre><span></span><code><a id="__codelineno-16-1" name="__codelineno-16-1" href="#__codelineno-16-1"></a>ls<span class="w"> </span>triotrain/variant_calling/data/GIAB/benchmark/
</code></pre></div>
<div class="highlight"><span class="filename">Expected outputs:</span><pre><span></span><code><a id="__codelineno-17-1" name="__codelineno-17-1" href="#__codelineno-17-1"></a>HG002_GRCh38_1_22_v4.2.1_benchmark.bed<span class="w">         </span>HG005_GRCh38_1_22_v4.2.1_benchmark.vcf.gz
<a id="__codelineno-17-2" name="__codelineno-17-2" href="#__codelineno-17-2"></a>HG002_GRCh38_1_22_v4.2.1_benchmark.vcf.gz<span class="w">      </span>HG005_GRCh38_1_22_v4.2.1_benchmark.vcf.gz.tbi
<a id="__codelineno-17-3" name="__codelineno-17-3" href="#__codelineno-17-3"></a>HG002_GRCh38_1_22_v4.2.1_benchmark.vcf.gz.tbi<span class="w">  </span>HG005_README_v4.2.1.txt
<a id="__codelineno-17-4" name="__codelineno-17-4" href="#__codelineno-17-4"></a>HG002_README_v4.2.1.txt<span class="w">                      </span>HG006_benchmark.md5
<a id="__codelineno-17-5" name="__codelineno-17-5" href="#__codelineno-17-5"></a>HG003_GRCh38_1_22_v4.2.1_benchmark.bed<span class="w">         </span>HG006_GRCh38_1_22_v4.2.1_benchmark.bed
<a id="__codelineno-17-6" name="__codelineno-17-6" href="#__codelineno-17-6"></a>HG003_GRCh38_1_22_v4.2.1_benchmark.vcf.gz<span class="w">      </span>HG006_GRCh38_1_22_v4.2.1_benchmark.vcf.gz
<a id="__codelineno-17-7" name="__codelineno-17-7" href="#__codelineno-17-7"></a>HG003_GRCh38_1_22_v4.2.1_benchmark.vcf.gz.tbi<span class="w">  </span>HG006_GRCh38_1_22_v4.2.1_benchmark.vcf.gz.tbi
<a id="__codelineno-17-8" name="__codelineno-17-8" href="#__codelineno-17-8"></a>HG003_README_v4.2.1.txt<span class="w">                      </span>HG006_README_v4.2.1.txt
<a id="__codelineno-17-9" name="__codelineno-17-9" href="#__codelineno-17-9"></a>HG004_GRCh38_1_22_v4.2.1_benchmark.bed<span class="w">         </span>HG007_benchmark.md5
<a id="__codelineno-17-10" name="__codelineno-17-10" href="#__codelineno-17-10"></a>HG004_GRCh38_1_22_v4.2.1_benchmark.vcf.gz<span class="w">      </span>HG007_GRCh38_1_22_v4.2.1_benchmark.bed
<a id="__codelineno-17-11" name="__codelineno-17-11" href="#__codelineno-17-11"></a>HG004_GRCh38_1_22_v4.2.1_benchmark.vcf.gz.tbi<span class="w">  </span>HG007_GRCh38_1_22_v4.2.1_benchmark.vcf.gz
<a id="__codelineno-17-12" name="__codelineno-17-12" href="#__codelineno-17-12"></a>HG004_README_v4.2.1.txt<span class="w">                      </span>HG007_GRCh38_1_22_v4.2.1_benchmark.vcf.gz.tbi
<a id="__codelineno-17-13" name="__codelineno-17-13" href="#__codelineno-17-13"></a>HG005_benchmark.md5<span class="w">                          </span>HG007_README_v4.2.1.txt
<a id="__codelineno-17-14" name="__codelineno-17-14" href="#__codelineno-17-14"></a>HG005_GRCh38_1_22_v4.2.1_benchmark.bed
</code></pre></div>
</details>
<details class="success">
<summary>Check | <code>reference/</code></summary>
<div class="highlight"><span class="filename">Run the following at the command line:</span><pre><span></span><code><a id="__codelineno-18-1" name="__codelineno-18-1" href="#__codelineno-18-1"></a>ls<span class="w"> </span>triotrain/variant_calling/data/GIAB/reference/
</code></pre></div>
<div class="highlight"><span class="filename">Expected outputs:</span><pre><span></span><code><a id="__codelineno-19-1" name="__codelineno-19-1" href="#__codelineno-19-1"></a>GRCh38_no_alt_analysis_set.fasta<span class="w">  </span>GRCh38_no_alt_analysis_set.fasta.fai<span class="w">  </span>md5checksums.txt
</code></pre></div>
</details>
<details class="success">
<summary>Check | Processing Raw Data Scripts</summary>
<p>In addition to the raw data, <code>download_GIAB.sh</code> also creates (3) bash scripts to process raw data into the formats expected by the tutorial:</p>
<ol>
<li><code>concat_PopVCFs.sh</code> + input file <code>PopVCF.merge.list</code> &mdash; merges the per-chr VCFs into a single file</li>
<li><code>AJtrio.download</code> &mdash; downloads GIAB Trio1</li>
<li><code>HCtrio.download</code> &mdash; downloads GIAB Trio2</li>
</ol>
</details>
<h2 id="6-process-raw-data">6. Process Raw Data</h2>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>These scripts can either be wrapped with SBATCH, or run interactively at the command line if you have enough memory. However, each script can take awhile to complete, particularly the <code>.download</code> scripts (1hr+).</p>
</div>
<h3 id="a-merge-the-popvcf">a. Merge the PopVCF</h3>
<div class="admonition warning">
<p class="admonition-title">Warning | Download Size</p>
</div>
<p>We need to create a single, genome-wide PopVCF from the raw per-chromosome PopVCF files. The per-chr Population VCFs were produced by the One Thousand Genomes Project (1kGP) and used to train the Human WGS.AF model. <a href="https://console.cloud.google.com/storage/browser/brain-genomics-public/research/cohort/1KGP/cohort_dv_glnexus_opt/v3_missing2ref?pageState=(%22StorageObjectListTable%22:(%22f%22:%22%255B%255D%22))&amp;prefix=&amp;forceOnObjectsSortingFiltering=false">You can view the raw files on Google Cloud Platform.</a></p>
<div class="highlight"><span class="filename">Run the following at the command line:</span><pre><span></span><code><a id="__codelineno-20-1" name="__codelineno-20-1" href="#__codelineno-20-1"></a>bash<span class="w"> </span>triotrain/variant_calling/data/GIAB/allele_freq/concat_PopVCFs.sh
</code></pre></div>
<details class="example">
<summary>Example | <code>concat_PopVCFs.sh</code></summary>
<div class="highlight"><pre><span></span><code><a id="__codelineno-21-1" name="__codelineno-21-1" href="#__codelineno-21-1"></a>source ./scripts/setup/modules.sh
<a id="__codelineno-21-2" name="__codelineno-21-2" href="#__codelineno-21-2"></a>bcftools concat --file-list ./triotrain/variant_calling/data/GIAB/allele_freq/PopVCF.merge.list -Oz -o ./triotrain/variant_calling/data/GIAB/allele_freq/cohort.release_missing2ref.no_calls.vcf.gz
<a id="__codelineno-21-3" name="__codelineno-21-3" href="#__codelineno-21-3"></a>bcftools index ./triotrain/variant_calling/data/GIAB/allele_freq/cohort.release_missing2ref.no_calls.vcf.gz
</code></pre></div>
</details>
<details class="success">
<summary>Check Intermediate Data | <code>allele_freq/</code></summary>
<div class="highlight"><span class="filename">Run the following at the command line:</span><pre><span></span><code><a id="__codelineno-22-1" name="__codelineno-22-1" href="#__codelineno-22-1"></a>ls<span class="w"> </span>triotrain/variant_calling/data/GIAB/allele_freq/<span class="w"> </span><span class="p">|</span><span class="w"> </span>grep<span class="w"> </span>cohort.release
</code></pre></div>
<div class="highlight"><span class="filename">Expected outputs:</span><pre><span></span><code><a id="__codelineno-23-1" name="__codelineno-23-1" href="#__codelineno-23-1"></a>cohort.release_missing2ref.no_calls.vcf.gz
<a id="__codelineno-23-2" name="__codelineno-23-2" href="#__codelineno-23-2"></a>cohort.release_missing2ref.no_calls.vcf.gz.csi
</code></pre></div>
</details>
<h3 id="b-download-sequence-data">b. Download Sequence Data</h3>
<div class="admonition warning">
<p class="admonition-title">Warning | Download Size</p>
</div>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>Given the large file size, the NIST FTP server runs slowly causing <code>curl</code> to timeout. <strong>You may need to run these scripts repeatedly until all data is transfered completely.</strong></p>
</div>
<p>We need to download the large sequence data files, and confirm they are not corrupted by checking the MD5 checksums. These <code>BAM/BAI</code> files orginate from the GIAB FTP site. <a href="https://github.com/genome-in-a-bottle/giab_data_indexes">An index of GIAB data created with these samples can be found on GitHub.</a></p>
<div class="highlight"><span class="filename">1. Run the following at the command line:</span><pre><span></span><code><a id="__codelineno-24-1" name="__codelineno-24-1" href="#__codelineno-24-1"></a>bash<span class="w"> </span>triotrain/variant_calling/data/GIAB/bam/AJtrio.download
</code></pre></div>
<details class="example">
<summary>Example | <code>AJtrio.download</code></summary>
<div class="highlight"><pre><span></span><code><a id="__codelineno-25-1" name="__codelineno-25-1" href="#__codelineno-25-1"></a>echo $(date &quot;+%Y-%m-%d %H:%M:%S&quot;) INFO: downloading [HG002.GRCh38.2x250.bam] now...
<a id="__codelineno-25-2" name="__codelineno-25-2" href="#__codelineno-25-2"></a>curl -o ./triotrain/variant_calling/data/GIAB/bam/HG002.GRCh38.2x250.bam -C - ftp://ftp-trace.ncbi.nlm.nih.gov/ReferenceSamples/giab/data/AshkenazimTrio/HG002_NA24385_son/NIST_Illumina_2x250bps/novoalign_bams/HG002.GRCh38.2x250.bam --keepalive-time 300
<a id="__codelineno-25-3" name="__codelineno-25-3" href="#__codelineno-25-3"></a>echo $(date &quot;+%Y-%m-%d %H:%M:%S&quot;) INFO: downloading [HG002.GRCh38.2x250.bam]... done
<a id="__codelineno-25-4" name="__codelineno-25-4" href="#__codelineno-25-4"></a>echo 56c30eaa4e2f25ff0ac80ef30e09d78e   ./triotrain/variant_calling/data/GIAB/bam/HG002.GRCh38.2x250.bam &gt; ./triotrain/variant_calling/data/GIAB/bam/HG002.GRCh38.2x250.bam.md5
<a id="__codelineno-25-5" name="__codelineno-25-5" href="#__codelineno-25-5"></a>echo $(date &quot;+%Y-%m-%d %H:%M:%S&quot;) INFO: checking [HG002.GRCh38.2x250.bam] for corruption...
<a id="__codelineno-25-6" name="__codelineno-25-6" href="#__codelineno-25-6"></a>md5sum -c ./triotrain/variant_calling/data/GIAB/bam/HG002.GRCh38.2x250.bam.md5
<a id="__codelineno-25-7" name="__codelineno-25-7" href="#__codelineno-25-7"></a>echo $(date &quot;+%Y-%m-%d %H:%M:%S&quot;) INFO: checking [HG002.GRCh38.2x250.bam] for corruption... done
<a id="__codelineno-25-8" name="__codelineno-25-8" href="#__codelineno-25-8"></a>echo ----------------------------------------------------
<a id="__codelineno-25-9" name="__codelineno-25-9" href="#__codelineno-25-9"></a>echo $(date &quot;+%Y-%m-%d %H:%M:%S&quot;) INFO: downloading [HG002.GRCh38.2x250.bam.bai] now...
<a id="__codelineno-25-10" name="__codelineno-25-10" href="#__codelineno-25-10"></a>curl -o ./triotrain/variant_calling/data/GIAB/bam/HG002.GRCh38.2x250.bam.bai -C - ftp://ftp-trace.ncbi.nlm.nih.gov/ReferenceSamples/giab/data/AshkenazimTrio/HG002_NA24385_son/NIST_Illumina_2x250bps/novoalign_bams/HG002.GRCh38.2x250.bam.bai
<a id="__codelineno-25-11" name="__codelineno-25-11" href="#__codelineno-25-11"></a>echo $(date &quot;+%Y-%m-%d %H:%M:%S&quot;) INFO: downloading [HG002.GRCh38.2x250.bam.bai] ... done
<a id="__codelineno-25-12" name="__codelineno-25-12" href="#__codelineno-25-12"></a>echo a3c2b449df6509ca83fbd3fea22b9aee   ./triotrain/variant_calling/data/GIAB/bam/HG002.GRCh38.2x250.bam.bai &gt; ./triotrain/variant_calling/data/GIAB/bam/HG002.GRCh38.2x250.bam.bai.md5
<a id="__codelineno-25-13" name="__codelineno-25-13" href="#__codelineno-25-13"></a>echo $(date &quot;+%Y-%m-%d %H:%M:%S&quot;) INFO: checking [HG002.GRCh38.2x250.bam.bai] for corruption...
<a id="__codelineno-25-14" name="__codelineno-25-14" href="#__codelineno-25-14"></a>md5sum -c ./triotrain/variant_calling/data/GIAB/bam/HG002.GRCh38.2x250.bam.bai.md5
<a id="__codelineno-25-15" name="__codelineno-25-15" href="#__codelineno-25-15"></a>echo $(date &quot;+%Y-%m-%d %H:%M:%S&quot;) INFO: checking [HG002.GRCh38.2x250.bam.bai] for corruption... done
<a id="__codelineno-25-16" name="__codelineno-25-16" href="#__codelineno-25-16"></a>echo ====================================================
<a id="__codelineno-25-17" name="__codelineno-25-17" href="#__codelineno-25-17"></a>echo $(date &quot;+%Y-%m-%d %H:%M:%S&quot;) INFO: downloading [HG003.GRCh38.2x250.bam] now...
<a id="__codelineno-25-18" name="__codelineno-25-18" href="#__codelineno-25-18"></a>curl -o ./triotrain/variant_calling/data/GIAB/bam/HG003.GRCh38.2x250.bam -C - ftp://ftp-trace.ncbi.nlm.nih.gov/ReferenceSamples/giab/data/AshkenazimTrio/HG003_NA24149_father/NIST_Illumina_2x250bps/novoalign_bams/HG003.GRCh38.2x250.bam --keepalive-time 300
<a id="__codelineno-25-19" name="__codelineno-25-19" href="#__codelineno-25-19"></a>echo $(date &quot;+%Y-%m-%d %H:%M:%S&quot;) INFO: downloading [HG003.GRCh38.2x250.bam]... done
<a id="__codelineno-25-20" name="__codelineno-25-20" href="#__codelineno-25-20"></a>echo 1b1ad2e8f5eba062086890cceb9671ab   ./triotrain/variant_calling/data/GIAB/bam/HG003.GRCh38.2x250.bam &gt; ./triotrain/variant_calling/data/GIAB/bam/HG003.GRCh38.2x250.bam.md5
<a id="__codelineno-25-21" name="__codelineno-25-21" href="#__codelineno-25-21"></a>echo $(date &quot;+%Y-%m-%d %H:%M:%S&quot;) INFO: checking [HG003.GRCh38.2x250.bam] for corruption...
<a id="__codelineno-25-22" name="__codelineno-25-22" href="#__codelineno-25-22"></a>md5sum -c ./triotrain/variant_calling/data/GIAB/bam/HG003.GRCh38.2x250.bam.md5
<a id="__codelineno-25-23" name="__codelineno-25-23" href="#__codelineno-25-23"></a>echo $(date &quot;+%Y-%m-%d %H:%M:%S&quot;) INFO: checking [HG003.GRCh38.2x250.bam] for corruption... done
<a id="__codelineno-25-24" name="__codelineno-25-24" href="#__codelineno-25-24"></a>echo ----------------------------------------------------
<a id="__codelineno-25-25" name="__codelineno-25-25" href="#__codelineno-25-25"></a>echo $(date &quot;+%Y-%m-%d %H:%M:%S&quot;) INFO: downloading [HG003.GRCh38.2x250.bam.bai] now...
<a id="__codelineno-25-26" name="__codelineno-25-26" href="#__codelineno-25-26"></a>curl -o ./triotrain/variant_calling/data/GIAB/bam/HG003.GRCh38.2x250.bam.bai -C - ftp://ftp-trace.ncbi.nlm.nih.gov/ReferenceSamples/giab/data/AshkenazimTrio/HG003_NA24149_father/NIST_Illumina_2x250bps/novoalign_bams/HG003.GRCh38.2x250.bam.bai
<a id="__codelineno-25-27" name="__codelineno-25-27" href="#__codelineno-25-27"></a>echo $(date &quot;+%Y-%m-%d %H:%M:%S&quot;) INFO: downloading [HG003.GRCh38.2x250.bam.bai] ... done
<a id="__codelineno-25-28" name="__codelineno-25-28" href="#__codelineno-25-28"></a>echo a2697b3ebb9b5ac6d9a27734406a211c   ./triotrain/variant_calling/data/GIAB/bam/HG003.GRCh38.2x250.bam.bai &gt; ./triotrain/variant_calling/data/GIAB/bam/HG003.GRCh38.2x250.bam.bai.md5
<a id="__codelineno-25-29" name="__codelineno-25-29" href="#__codelineno-25-29"></a>echo $(date &quot;+%Y-%m-%d %H:%M:%S&quot;) INFO: checking [HG003.GRCh38.2x250.bam.bai] for corruption...
<a id="__codelineno-25-30" name="__codelineno-25-30" href="#__codelineno-25-30"></a>md5sum -c ./triotrain/variant_calling/data/GIAB/bam/HG003.GRCh38.2x250.bam.bai.md5
<a id="__codelineno-25-31" name="__codelineno-25-31" href="#__codelineno-25-31"></a>echo $(date &quot;+%Y-%m-%d %H:%M:%S&quot;) INFO: checking [HG003.GRCh38.2x250.bam.bai] for corruption... done
<a id="__codelineno-25-32" name="__codelineno-25-32" href="#__codelineno-25-32"></a>echo ====================================================
<a id="__codelineno-25-33" name="__codelineno-25-33" href="#__codelineno-25-33"></a>echo $(date &quot;+%Y-%m-%d %H:%M:%S&quot;) INFO: downloading [HG004.GRCh38.2x250.bam] now...
<a id="__codelineno-25-34" name="__codelineno-25-34" href="#__codelineno-25-34"></a>curl -o ./triotrain/variant_calling/data/GIAB/bam/HG004.GRCh38.2x250.bam -C - ftp://ftp-trace.ncbi.nlm.nih.gov/ReferenceSamples/giab/data/AshkenazimTrio/HG004_NA24143_mother/NIST_Illumina_2x250bps/novoalign_bams/HG004.GRCh38.2x250.bam --keepalive-time 300
<a id="__codelineno-25-35" name="__codelineno-25-35" href="#__codelineno-25-35"></a>echo $(date &quot;+%Y-%m-%d %H:%M:%S&quot;) INFO: downloading [HG004.GRCh38.2x250.bam]... done
<a id="__codelineno-25-36" name="__codelineno-25-36" href="#__codelineno-25-36"></a>echo e084422a267170b550c9704ad61db9b2   ./triotrain/variant_calling/data/GIAB/bam/HG004.GRCh38.2x250.bam &gt; ./triotrain/variant_calling/data/GIAB/bam/HG004.GRCh38.2x250.bam.md5
<a id="__codelineno-25-37" name="__codelineno-25-37" href="#__codelineno-25-37"></a>echo $(date &quot;+%Y-%m-%d %H:%M:%S&quot;) INFO: checking [HG004.GRCh38.2x250.bam] for corruption...
<a id="__codelineno-25-38" name="__codelineno-25-38" href="#__codelineno-25-38"></a>md5sum -c ./triotrain/variant_calling/data/GIAB/bam/HG004.GRCh38.2x250.bam.md5
<a id="__codelineno-25-39" name="__codelineno-25-39" href="#__codelineno-25-39"></a>echo $(date &quot;+%Y-%m-%d %H:%M:%S&quot;) INFO: checking [HG004.GRCh38.2x250.bam] for corruption... done
<a id="__codelineno-25-40" name="__codelineno-25-40" href="#__codelineno-25-40"></a>echo ----------------------------------------------------
<a id="__codelineno-25-41" name="__codelineno-25-41" href="#__codelineno-25-41"></a>echo $(date &quot;+%Y-%m-%d %H:%M:%S&quot;) INFO: downloading [HG004.GRCh38.2x250.bam.bai] now...
<a id="__codelineno-25-42" name="__codelineno-25-42" href="#__codelineno-25-42"></a>curl -o ./triotrain/variant_calling/data/GIAB/bam/HG004.GRCh38.2x250.bam.bai -C - ftp://ftp-trace.ncbi.nlm.nih.gov/ReferenceSamples/giab/data/AshkenazimTrio/HG004_NA24143_mother/NIST_Illumina_2x250bps/novoalign_bams/HG004.GRCh38.2x250.bam.bai
<a id="__codelineno-25-43" name="__codelineno-25-43" href="#__codelineno-25-43"></a>echo $(date &quot;+%Y-%m-%d %H:%M:%S&quot;) INFO: downloading [HG004.GRCh38.2x250.bam.bai] ... done
<a id="__codelineno-25-44" name="__codelineno-25-44" href="#__codelineno-25-44"></a>echo 4827d6457b202f85482787964a1bc6c7   ./triotrain/variant_calling/data/GIAB/bam/HG004.GRCh38.2x250.bam.bai &gt; ./triotrain/variant_calling/data/GIAB/bam/HG004.GRCh38.2x250.bam.bai.md5
<a id="__codelineno-25-45" name="__codelineno-25-45" href="#__codelineno-25-45"></a>echo $(date &quot;+%Y-%m-%d %H:%M:%S&quot;) INFO: checking [HG004.GRCh38.2x250.bam.bai] for corruption...
<a id="__codelineno-25-46" name="__codelineno-25-46" href="#__codelineno-25-46"></a>md5sum -c ./triotrain/variant_calling/data/GIAB/bam/HG004.GRCh38.2x250.bam.bai.md5
<a id="__codelineno-25-47" name="__codelineno-25-47" href="#__codelineno-25-47"></a>echo $(date &quot;+%Y-%m-%d %H:%M:%S&quot;) INFO: checking [HG004.GRCh38.2x250.bam.bai] for corruption... done
<a id="__codelineno-25-48" name="__codelineno-25-48" href="#__codelineno-25-48"></a>echo ====================================================
</code></pre></div>
</details>
<details class="success">
<summary>Check Intermediate Data | HG002:</summary>
<div class="highlight"><span class="filename">Run the following at the command line:</span><pre><span></span><code><a id="__codelineno-26-1" name="__codelineno-26-1" href="#__codelineno-26-1"></a>ls<span class="w"> </span>triotrain/variant_calling/data/GIAB/bam/<span class="w"> </span><span class="p">|</span><span class="w"> </span>grep<span class="w"> </span>HG002
</code></pre></div>
<div class="highlight"><span class="filename">Expected outputs:</span><pre><span></span><code><a id="__codelineno-27-1" name="__codelineno-27-1" href="#__codelineno-27-1"></a>HG002_corrected_md5sums.feb19upload.txt
<a id="__codelineno-27-2" name="__codelineno-27-2" href="#__codelineno-27-2"></a>HG002.GRCh38.2x250.bam
<a id="__codelineno-27-3" name="__codelineno-27-3" href="#__codelineno-27-3"></a>HG002.GRCh38.2x250.bam.bai
<a id="__codelineno-27-4" name="__codelineno-27-4" href="#__codelineno-27-4"></a>HG002.GRCh38.2x250.bam.bai.md5
<a id="__codelineno-27-5" name="__codelineno-27-5" href="#__codelineno-27-5"></a>HG002.GRCh38.2x250.bam.md5
</code></pre></div>
</details>
<details class="success">
<summary>Check Intermediate Data | HG003:</summary>
<div class="highlight"><span class="filename">Run the following at the command line:</span><pre><span></span><code><a id="__codelineno-28-1" name="__codelineno-28-1" href="#__codelineno-28-1"></a>ls<span class="w"> </span>triotrain/variant_calling/data/GIAB/bam/<span class="w"> </span><span class="p">|</span><span class="w"> </span>grep<span class="w"> </span>HG003
</code></pre></div>
<div class="highlight"><span class="filename">Expected outputs:</span><pre><span></span><code><a id="__codelineno-29-1" name="__codelineno-29-1" href="#__codelineno-29-1"></a>HG003.GRCh38.2x250.bam
<a id="__codelineno-29-2" name="__codelineno-29-2" href="#__codelineno-29-2"></a>HG003.GRCh38.2x250.bam.bai
<a id="__codelineno-29-3" name="__codelineno-29-3" href="#__codelineno-29-3"></a>HG003.GRCh38.2x250.bam.bai.md5
<a id="__codelineno-29-4" name="__codelineno-29-4" href="#__codelineno-29-4"></a>HG003.GRCh38.2x250.bam.md5
</code></pre></div>
</details>
<details class="success">
<summary>Check Intermediate Data | HG004:</summary>
<div class="highlight"><span class="filename">Run the following at the command line:</span><pre><span></span><code><a id="__codelineno-30-1" name="__codelineno-30-1" href="#__codelineno-30-1"></a>ls<span class="w"> </span>triotrain/variant_calling/data/GIAB/bam/<span class="w"> </span><span class="p">|</span><span class="w"> </span>grep<span class="w"> </span>HG004
</code></pre></div>
<div class="highlight"><span class="filename">Expected outputs:</span><pre><span></span><code><a id="__codelineno-31-1" name="__codelineno-31-1" href="#__codelineno-31-1"></a>HG004.GRCh38.2x250.bam
<a id="__codelineno-31-2" name="__codelineno-31-2" href="#__codelineno-31-2"></a>HG004.GRCh38.2x250.bam.bai
<a id="__codelineno-31-3" name="__codelineno-31-3" href="#__codelineno-31-3"></a>HG004.GRCh38.2x250.bam.bai.md5
<a id="__codelineno-31-4" name="__codelineno-31-4" href="#__codelineno-31-4"></a>HG004.GRCh38.2x250.bam.md5
</code></pre></div>
</details>
<div class="highlight"><span class="filename">2. Run the following at the command line:</span><pre><span></span><code><a id="__codelineno-32-1" name="__codelineno-32-1" href="#__codelineno-32-1"></a>bash<span class="w"> </span>triotrain/variant_calling/data/GIAB/bam/HCtrio.download
</code></pre></div>
<details class="example">
<summary>Example | <code>HCtrio.download</code></summary>
<div class="highlight"><pre><span></span><code><a id="__codelineno-33-1" name="__codelineno-33-1" href="#__codelineno-33-1"></a>echo $(date &quot;+%Y-%m-%d %H:%M:%S&quot;) INFO: downloading [HG005.GRCh38_full_plus_hs38d1_analysis_set_minus_alts.300x.bam] now...
<a id="__codelineno-33-2" name="__codelineno-33-2" href="#__codelineno-33-2"></a>curl -o ./triotrain/variant_calling/data/GIAB/bam/HG005.GRCh38_full_plus_hs38d1_analysis_set_minus_alts.300x.bam -C - ftp://ftp-trace.ncbi.nlm.nih.gov/ReferenceSamples/giab/data/ChineseTrio/HG005_NA24631_son/HG005_NA24631_son_HiSeq_300x/NHGRI_Illumina300X_Chinesetrio_novoalign_bams/HG005.GRCh38_full_plus_hs38d1_analysis_set_minus_alts.300x.bam --keepalive-time 300
<a id="__codelineno-33-3" name="__codelineno-33-3" href="#__codelineno-33-3"></a>echo $(date &quot;+%Y-%m-%d %H:%M:%S&quot;) INFO: downloading [HG005.GRCh38_full_plus_hs38d1_analysis_set_minus_alts.300x.bam]... done
<a id="__codelineno-33-4" name="__codelineno-33-4" href="#__codelineno-33-4"></a>echo 50080d14ba49462cfc6348e1fb41b5b3   ./triotrain/variant_calling/data/GIAB/bam/HG005.GRCh38_full_plus_hs38d1_analysis_set_minus_alts.300x.bam &gt; ./triotrain/variant_calling/data/GIAB/bam/HG005.GRCh38_full_plus_hs38d1_analysis_set_minus_alts.300x.bam.md5
<a id="__codelineno-33-5" name="__codelineno-33-5" href="#__codelineno-33-5"></a>echo $(date &quot;+%Y-%m-%d %H:%M:%S&quot;) INFO: checking [HG005.GRCh38_full_plus_hs38d1_analysis_set_minus_alts.300x.bam] for corruption...
<a id="__codelineno-33-6" name="__codelineno-33-6" href="#__codelineno-33-6"></a>md5sum -c ./triotrain/variant_calling/data/GIAB/bam/HG005.GRCh38_full_plus_hs38d1_analysis_set_minus_alts.300x.bam.md5
<a id="__codelineno-33-7" name="__codelineno-33-7" href="#__codelineno-33-7"></a>echo $(date &quot;+%Y-%m-%d %H:%M:%S&quot;) INFO: checking [HG005.GRCh38_full_plus_hs38d1_analysis_set_minus_alts.300x.bam] for corruption... done
<a id="__codelineno-33-8" name="__codelineno-33-8" href="#__codelineno-33-8"></a>echo ----------------------------------------------------
<a id="__codelineno-33-9" name="__codelineno-33-9" href="#__codelineno-33-9"></a>echo $(date &quot;+%Y-%m-%d %H:%M:%S&quot;) INFO: downloading [HG005.GRCh38_full_plus_hs38d1_analysis_set_minus_alts.300x.bam.bai] now...
<a id="__codelineno-33-10" name="__codelineno-33-10" href="#__codelineno-33-10"></a>curl -o ./triotrain/variant_calling/data/GIAB/bam/HG005.GRCh38_full_plus_hs38d1_analysis_set_minus_alts.300x.bam.bai -C - ftp://ftp-trace.ncbi.nlm.nih.gov/ReferenceSamples/giab/data/ChineseTrio/HG005_NA24631_son/HG005_NA24631_son_HiSeq_300x/NHGRI_Illumina300X_Chinesetrio_novoalign_bams/HG005.GRCh38_full_plus_hs38d1_analysis_set_minus_alts.300x.bam.bai
<a id="__codelineno-33-11" name="__codelineno-33-11" href="#__codelineno-33-11"></a>echo $(date &quot;+%Y-%m-%d %H:%M:%S&quot;) INFO: downloading [HG005.GRCh38_full_plus_hs38d1_analysis_set_minus_alts.300x.bam.bai] ... done
<a id="__codelineno-33-12" name="__codelineno-33-12" href="#__codelineno-33-12"></a>echo 2d38162ba61e909d08e4ed438096174e   ./triotrain/variant_calling/data/GIAB/bam/HG005.GRCh38_full_plus_hs38d1_analysis_set_minus_alts.300x.bam.bai &gt; ./triotrain/variant_calling/data/GIAB/bam/HG005.GRCh38_full_plus_hs38d1_analysis_set_minus_alts.300x.bam.bai.md5
<a id="__codelineno-33-13" name="__codelineno-33-13" href="#__codelineno-33-13"></a>echo $(date &quot;+%Y-%m-%d %H:%M:%S&quot;) INFO: checking [HG005.GRCh38_full_plus_hs38d1_analysis_set_minus_alts.300x.bam.bai] for corruption...
<a id="__codelineno-33-14" name="__codelineno-33-14" href="#__codelineno-33-14"></a>md5sum -c ./triotrain/variant_calling/data/GIAB/bam/HG005.GRCh38_full_plus_hs38d1_analysis_set_minus_alts.300x.bam.bai.md5
<a id="__codelineno-33-15" name="__codelineno-33-15" href="#__codelineno-33-15"></a>echo $(date &quot;+%Y-%m-%d %H:%M:%S&quot;) INFO: checking [HG005.GRCh38_full_plus_hs38d1_analysis_set_minus_alts.300x.bam.bai] for corruption... done
<a id="__codelineno-33-16" name="__codelineno-33-16" href="#__codelineno-33-16"></a>echo ====================================================
<a id="__codelineno-33-17" name="__codelineno-33-17" href="#__codelineno-33-17"></a>echo $(date &quot;+%Y-%m-%d %H:%M:%S&quot;) INFO: downloading [HG006.GRCh38_full_plus_hs38d1_analysis_set_minus_alts.100x.bam] now...
<a id="__codelineno-33-18" name="__codelineno-33-18" href="#__codelineno-33-18"></a>curl -o ./triotrain/variant_calling/data/GIAB/bam/HG006.GRCh38_full_plus_hs38d1_analysis_set_minus_alts.100x.bam -C - ftp://ftp-trace.ncbi.nlm.nih.gov/ReferenceSamples/giab/data/ChineseTrio/HG006_NA24694-huCA017E_father/NA24694_Father_HiSeq100x/NHGRI_Illumina100X_Chinesetrio_novoalign_bams/HG006.GRCh38_full_plus_hs38d1_analysis_set_minus_alts.100x.bam --keepalive-time 300
<a id="__codelineno-33-19" name="__codelineno-33-19" href="#__codelineno-33-19"></a>echo $(date &quot;+%Y-%m-%d %H:%M:%S&quot;) INFO: downloading [HG006.GRCh38_full_plus_hs38d1_analysis_set_minus_alts.100x.bam]... done
<a id="__codelineno-33-20" name="__codelineno-33-20" href="#__codelineno-33-20"></a>echo 3a7c239988862e5a4556eb8781058b1f   ./triotrain/variant_calling/data/GIAB/bam/HG006.GRCh38_full_plus_hs38d1_analysis_set_minus_alts.100x.bam &gt; ./triotrain/variant_calling/data/GIAB/bam/HG006.GRCh38_full_plus_hs38d1_analysis_set_minus_alts.100x.bam.md5
<a id="__codelineno-33-21" name="__codelineno-33-21" href="#__codelineno-33-21"></a>echo $(date &quot;+%Y-%m-%d %H:%M:%S&quot;) INFO: checking [HG006.GRCh38_full_plus_hs38d1_analysis_set_minus_alts.100x.bam] for corruption...
<a id="__codelineno-33-22" name="__codelineno-33-22" href="#__codelineno-33-22"></a>md5sum -c ./triotrain/variant_calling/data/GIAB/bam/HG006.GRCh38_full_plus_hs38d1_analysis_set_minus_alts.100x.bam.md5
<a id="__codelineno-33-23" name="__codelineno-33-23" href="#__codelineno-33-23"></a>echo $(date &quot;+%Y-%m-%d %H:%M:%S&quot;) INFO: checking [HG006.GRCh38_full_plus_hs38d1_analysis_set_minus_alts.100x.bam] for corruption... done
<a id="__codelineno-33-24" name="__codelineno-33-24" href="#__codelineno-33-24"></a>echo ----------------------------------------------------
<a id="__codelineno-33-25" name="__codelineno-33-25" href="#__codelineno-33-25"></a>echo $(date &quot;+%Y-%m-%d %H:%M:%S&quot;) INFO: downloading [HG006.GRCh38_full_plus_hs38d1_analysis_set_minus_alts.100x.bam.bai] now...
<a id="__codelineno-33-26" name="__codelineno-33-26" href="#__codelineno-33-26"></a>curl -o ./triotrain/variant_calling/data/GIAB/bam/HG006.GRCh38_full_plus_hs38d1_analysis_set_minus_alts.100x.bam.bai -C - ftp://ftp-trace.ncbi.nlm.nih.gov/ReferenceSamples/giab/data/ChineseTrio/HG006_NA24694-huCA017E_father/NA24694_Father_HiSeq100x/NHGRI_Illumina100X_Chinesetrio_novoalign_bams/HG006.GRCh38_full_plus_hs38d1_analysis_set_minus_alts.100x.bam.bai
<a id="__codelineno-33-27" name="__codelineno-33-27" href="#__codelineno-33-27"></a>echo $(date &quot;+%Y-%m-%d %H:%M:%S&quot;) INFO: downloading [HG006.GRCh38_full_plus_hs38d1_analysis_set_minus_alts.100x.bam.bai] ... done
<a id="__codelineno-33-28" name="__codelineno-33-28" href="#__codelineno-33-28"></a>echo 7bfe1f0988d27898befeaa4673d23a47   ./triotrain/variant_calling/data/GIAB/bam/HG006.GRCh38_full_plus_hs38d1_analysis_set_minus_alts.100x.bam.bai &gt; ./triotrain/variant_calling/data/GIAB/bam/HG006.GRCh38_full_plus_hs38d1_analysis_set_minus_alts.100x.bam.bai.md5
<a id="__codelineno-33-29" name="__codelineno-33-29" href="#__codelineno-33-29"></a>echo $(date &quot;+%Y-%m-%d %H:%M:%S&quot;) INFO: checking [HG006.GRCh38_full_plus_hs38d1_analysis_set_minus_alts.100x.bam.bai] for corruption...
<a id="__codelineno-33-30" name="__codelineno-33-30" href="#__codelineno-33-30"></a>md5sum -c ./triotrain/variant_calling/data/GIAB/bam/HG006.GRCh38_full_plus_hs38d1_analysis_set_minus_alts.100x.bam.bai.md5
<a id="__codelineno-33-31" name="__codelineno-33-31" href="#__codelineno-33-31"></a>echo $(date &quot;+%Y-%m-%d %H:%M:%S&quot;) INFO: checking [HG006.GRCh38_full_plus_hs38d1_analysis_set_minus_alts.100x.bam.bai] for corruption... done
<a id="__codelineno-33-32" name="__codelineno-33-32" href="#__codelineno-33-32"></a>echo ====================================================
<a id="__codelineno-33-33" name="__codelineno-33-33" href="#__codelineno-33-33"></a>echo $(date &quot;+%Y-%m-%d %H:%M:%S&quot;) INFO: downloading [HG007.GRCh38_full_plus_hs38d1_analysis_set_minus_alts.100x.bam] now...
<a id="__codelineno-33-34" name="__codelineno-33-34" href="#__codelineno-33-34"></a>curl -o ./triotrain/variant_calling/data/GIAB/bam/HG007.GRCh38_full_plus_hs38d1_analysis_set_minus_alts.100x.bam -C - ftp://ftp-trace.ncbi.nlm.nih.gov/ReferenceSamples/giab/data/ChineseTrio/HG007_NA24695-hu38168_mother/NA24695_Mother_HiSeq100x/NHGRI_Illumina100X_Chinesetrio_novoalign_bams/HG007.GRCh38_full_plus_hs38d1_analysis_set_minus_alts.100x.bam --keepalive-time 300
<a id="__codelineno-33-35" name="__codelineno-33-35" href="#__codelineno-33-35"></a>echo $(date &quot;+%Y-%m-%d %H:%M:%S&quot;) INFO: downloading [HG007.GRCh38_full_plus_hs38d1_analysis_set_minus_alts.100x.bam]... done
<a id="__codelineno-33-36" name="__codelineno-33-36" href="#__codelineno-33-36"></a>echo 3b8442105a4a45f71c3c1bda8af238dd   ./triotrain/variant_calling/data/GIAB/bam/HG007.GRCh38_full_plus_hs38d1_analysis_set_minus_alts.100x.bam &gt; ./triotrain/variant_calling/data/GIAB/bam/HG007.GRCh38_full_plus_hs38d1_analysis_set_minus_alts.100x.bam.md5
<a id="__codelineno-33-37" name="__codelineno-33-37" href="#__codelineno-33-37"></a>echo $(date &quot;+%Y-%m-%d %H:%M:%S&quot;) INFO: checking [HG007.GRCh38_full_plus_hs38d1_analysis_set_minus_alts.100x.bam] for corruption...
<a id="__codelineno-33-38" name="__codelineno-33-38" href="#__codelineno-33-38"></a>md5sum -c ./triotrain/variant_calling/data/GIAB/bam/HG007.GRCh38_full_plus_hs38d1_analysis_set_minus_alts.100x.bam.md5
<a id="__codelineno-33-39" name="__codelineno-33-39" href="#__codelineno-33-39"></a>echo $(date &quot;+%Y-%m-%d %H:%M:%S&quot;) INFO: checking [HG007.GRCh38_full_plus_hs38d1_analysis_set_minus_alts.100x.bam] for corruption... done
<a id="__codelineno-33-40" name="__codelineno-33-40" href="#__codelineno-33-40"></a>echo ----------------------------------------------------
<a id="__codelineno-33-41" name="__codelineno-33-41" href="#__codelineno-33-41"></a>echo $(date &quot;+%Y-%m-%d %H:%M:%S&quot;) INFO: downloading [HG007.GRCh38_full_plus_hs38d1_analysis_set_minus_alts.100x.bam.bai] now...
<a id="__codelineno-33-42" name="__codelineno-33-42" href="#__codelineno-33-42"></a>curl -o ./triotrain/variant_calling/data/GIAB/bam/HG007.GRCh38_full_plus_hs38d1_analysis_set_minus_alts.100x.bam.bai -C - ftp://ftp-trace.ncbi.nlm.nih.gov/ReferenceSamples/giab/data/ChineseTrio/HG007_NA24695-hu38168_mother/NA24695_Mother_HiSeq100x/NHGRI_Illumina100X_Chinesetrio_novoalign_bams/HG007.GRCh38_full_plus_hs38d1_analysis_set_minus_alts.100x.bam.bai
<a id="__codelineno-33-43" name="__codelineno-33-43" href="#__codelineno-33-43"></a>echo $(date &quot;+%Y-%m-%d %H:%M:%S&quot;) INFO: downloading [HG007.GRCh38_full_plus_hs38d1_analysis_set_minus_alts.100x.bam.bai] ... done
<a id="__codelineno-33-44" name="__codelineno-33-44" href="#__codelineno-33-44"></a>echo 0c218cecd8cde9ad272de6d77bbd110f   ./triotrain/variant_calling/data/GIAB/bam/HG007.GRCh38_full_plus_hs38d1_analysis_set_minus_alts.100x.bam.bai &gt; ./triotrain/variant_calling/data/GIAB/bam/HG007.GRCh38_full_plus_hs38d1_analysis_set_minus_alts.100x.bam.bai.md5
<a id="__codelineno-33-45" name="__codelineno-33-45" href="#__codelineno-33-45"></a>echo $(date &quot;+%Y-%m-%d %H:%M:%S&quot;) INFO: checking [HG007.GRCh38_full_plus_hs38d1_analysis_set_minus_alts.100x.bam.bai] for corruption...
<a id="__codelineno-33-46" name="__codelineno-33-46" href="#__codelineno-33-46"></a>md5sum -c ./triotrain/variant_calling/data/GIAB/bam/HG007.GRCh38_full_plus_hs38d1_analysis_set_minus_alts.100x.bam.bai.md5
<a id="__codelineno-33-47" name="__codelineno-33-47" href="#__codelineno-33-47"></a>echo $(date &quot;+%Y-%m-%d %H:%M:%S&quot;) INFO: checking [HG007.GRCh38_full_plus_hs38d1_analysis_set_minus_alts.100x.bam.bai] for corruption... done
<a id="__codelineno-33-48" name="__codelineno-33-48" href="#__codelineno-33-48"></a>echo ====================================================
</code></pre></div>
</details>
<details class="success">
<summary>Check Intermediate Data | HG005:</summary>
<div class="highlight"><span class="filename">Run the following at the command line:</span><pre><span></span><code><a id="__codelineno-34-1" name="__codelineno-34-1" href="#__codelineno-34-1"></a>ls<span class="w"> </span>triotrain/variant_calling/data/GIAB/bam/<span class="w"> </span><span class="p">|</span><span class="w"> </span>grep<span class="w"> </span>HG005
</code></pre></div>
<div class="highlight"><span class="filename">Expected outputs:</span><pre><span></span><code><a id="__codelineno-35-1" name="__codelineno-35-1" href="#__codelineno-35-1"></a>HG005.GRCh38_full_plus_hs38d1_analysis_set_minus_alts.300x.bam
<a id="__codelineno-35-2" name="__codelineno-35-2" href="#__codelineno-35-2"></a>HG005.GRCh38_full_plus_hs38d1_analysis_set_minus_alts.300x.bam.bai
<a id="__codelineno-35-3" name="__codelineno-35-3" href="#__codelineno-35-3"></a>HG005.GRCh38_full_plus_hs38d1_analysis_set_minus_alts.300x.bam.bai.md5
<a id="__codelineno-35-4" name="__codelineno-35-4" href="#__codelineno-35-4"></a>HG005.GRCh38_full_plus_hs38d1_analysis_set_minus_alts.300x.bam.md5
</code></pre></div>
</details>
<details class="success">
<summary>Check Intermediate Data | HG006:</summary>
<div class="highlight"><span class="filename">Run the following at the command line:</span><pre><span></span><code><a id="__codelineno-36-1" name="__codelineno-36-1" href="#__codelineno-36-1"></a>ls<span class="w"> </span>triotrain/variant_calling/data/GIAB/bam/<span class="w"> </span><span class="p">|</span><span class="w"> </span>grep<span class="w"> </span>HG006
</code></pre></div>
<div class="highlight"><span class="filename">Expected outputs:</span><pre><span></span><code><a id="__codelineno-37-1" name="__codelineno-37-1" href="#__codelineno-37-1"></a>HG006.GRCh38_full_plus_hs38d1_analysis_set_minus_alts.100x.bam
<a id="__codelineno-37-2" name="__codelineno-37-2" href="#__codelineno-37-2"></a>HG006.GRCh38_full_plus_hs38d1_analysis_set_minus_alts.100x.bam.bai
<a id="__codelineno-37-3" name="__codelineno-37-3" href="#__codelineno-37-3"></a>HG006.GRCh38_full_plus_hs38d1_analysis_set_minus_alts.100x.bam.bai.md5
<a id="__codelineno-37-4" name="__codelineno-37-4" href="#__codelineno-37-4"></a>HG006.GRCh38_full_plus_hs38d1_analysis_set_minus_alts.100x.bam.md5
</code></pre></div>
</details>
<details class="success">
<summary>Check Intermediate Data | HG007:</summary>
<div class="highlight"><span class="filename">Run the following at the command line:</span><pre><span></span><code><a id="__codelineno-38-1" name="__codelineno-38-1" href="#__codelineno-38-1"></a>ls<span class="w"> </span>triotrain/variant_calling/data/GIAB/bam/<span class="w"> </span><span class="p">|</span><span class="w"> </span>grep<span class="w"> </span>HG007
</code></pre></div>
<div class="highlight"><span class="filename">Expected outputs:</span><pre><span></span><code><a id="__codelineno-39-1" name="__codelineno-39-1" href="#__codelineno-39-1"></a>HG007.GRCh38_full_plus_hs38d1_analysis_set_minus_alts.100x.bam
<a id="__codelineno-39-2" name="__codelineno-39-2" href="#__codelineno-39-2"></a>HG007.GRCh38_full_plus_hs38d1_analysis_set_minus_alts.100x.bam.bai
<a id="__codelineno-39-3" name="__codelineno-39-3" href="#__codelineno-39-3"></a>HG007.GRCh38_full_plus_hs38d1_analysis_set_minus_alts.100x.bam.bai.md5
<a id="__codelineno-39-4" name="__codelineno-39-4" href="#__codelineno-39-4"></a>HG007.GRCh38_full_plus_hs38d1_analysis_set_minus_alts.100x.bam.md5
</code></pre></div>
</details>
<h2 id="7-create-triotrain-inputs">7. Create TrioTrain Inputs</h2>
<p>There are (3) required input files we must create before we can run TrioTrain. <a href="../../user-guide/usage_guide/#assumptions">Complete details about all required data can be found in the TrioTrain User Guide.</a></p>
<h3 id="a-reference-dictionary-file-dict">a. <a href="../../user-guide/usage_guide/#reference">Reference Dictionary File (<code>.dict</code>)</a></h3>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>This step is specific to the Human reference genome GRCh38 since cattle-specific input files are pre-packaged with TrioTrain. <strong>If you are working with a new species, you will need to create this file for your reference genome.</strong></p>
</div>
<p>We need a reference dictionary file in the same directory as the reference genome. This file defines the valid genomic coordinates for TrioTrain's region shuffling. </p>
<p>By default, region shuffling will only use the autosomes and X chromosome. However, you can expand or contract the shuffling area by providing an alternative region file for a particular trio by providing an existing <code>BED</code> file under the <code>RegionsFile</code> column within the <a href="#b-metadata-file-csv">metadata file (<code>.csv</code>).</a></p>
<div class="highlight"><span class="filename">Run the following at the command line:</span><pre><span></span><code><a id="__codelineno-40-1" name="__codelineno-40-1" href="#__codelineno-40-1"></a>picard<span class="w"> </span>CreateSequenceDictionary<span class="w"> </span><span class="se">\</span>
<a id="__codelineno-40-2" name="__codelineno-40-2" href="#__codelineno-40-2"></a><span class="w">    </span>--REFERENCE<span class="w"> </span>./triotrain/variant_calling/data/GIAB/reference/GRCh38_no_alt_analysis_set.fasta<span class="w"> </span><span class="se">\</span>
<a id="__codelineno-40-3" name="__codelineno-40-3" href="#__codelineno-40-3"></a><span class="w">    </span>--OUTPUT<span class="w"> </span>./triotrain/variant_calling/data/GIAB/reference/GRCh38_no_alt_analysis_set.dict<span class="w"> </span><span class="se">\</span>
<a id="__codelineno-40-4" name="__codelineno-40-4" href="#__codelineno-40-4"></a><span class="w">    </span>--SPECIES<span class="w"> </span>human
</code></pre></div>
<details class="success">
<summary>Check | Dictionary File </summary>
<div class="highlight"><span class="filename">Run the following at the command line:</span><pre><span></span><code><a id="__codelineno-41-1" name="__codelineno-41-1" href="#__codelineno-41-1"></a>ls<span class="w"> </span>triotrain/variant_calling/data/GIAB/reference/<span class="w"> </span><span class="p">|</span><span class="w"> </span>grep<span class="w"> </span>.dict
</code></pre></div>
<div class="highlight"><span class="filename">Expected output:</span><pre><span></span><code><a id="__codelineno-42-1" name="__codelineno-42-1" href="#__codelineno-42-1"></a>GRCh38_no_alt_analysis_set.dict
</code></pre></div>
</details>
<h3 id="b-metadata-file-csv">b. <a href="../../user-guide/usage_guide/#providing-required-data-to-triotrain">Metadata file (<code>.csv</code>)</a></h3>
<p>We also need a metadata file to tell TrioTrain where to find all of previously downloaded Human GIAB data. This file contains pedigree information, and the absolute paths for file inputs. Absolute paths are required to help the Apptainer/Singularity containers identify local files. <a href="../../user-guide/usage_guide/#metadata-format">Formatting specifications for this required input can be found in the TrioTrain User Guide.</a></p>
<p>For the tutorial, we've created a helper script to automatically create an example of this file. This script uses expectations of where the tutorial data are stored to add local paths. <strong>However, outside of the tutorial this file is a user-created input.</strong></p>
<div class="highlight"><span class="filename">Run the following at the command line:</span><pre><span></span><code><a id="__codelineno-43-1" name="__codelineno-43-1" href="#__codelineno-43-1"></a><span class="nb">source</span><span class="w"> </span>./scripts/setup/modules.sh
<a id="__codelineno-43-2" name="__codelineno-43-2" href="#__codelineno-43-2"></a><span class="nb">source</span><span class="w"> </span>./scripts/start_conda.sh<span class="w">                 </span><span class="c1"># Ensure the previously built conda env is active</span>
<a id="__codelineno-43-3" name="__codelineno-43-3" href="#__codelineno-43-3"></a>python3<span class="w"> </span>./triotrain/model_training/tutorial/create_metadata.py
</code></pre></div>
<details class="example">
<summary>Example | <code>create_metadata.py</code></summary>
<div class="highlight"><pre><span></span><code><a id="__codelineno-44-1" name="__codelineno-44-1" href="#__codelineno-44-1"></a>#!/usr/bin/python3
<a id="__codelineno-44-2" name="__codelineno-44-2" href="#__codelineno-44-2"></a>&quot;&quot;&quot;
<a id="__codelineno-44-3" name="__codelineno-44-3" href="#__codelineno-44-3"></a>description: creates a tutorial metadata file from the Human GIAB data downloaded during the walk-through
<a id="__codelineno-44-4" name="__codelineno-44-4" href="#__codelineno-44-4"></a>example:
<a id="__codelineno-44-5" name="__codelineno-44-5" href="#__codelineno-44-5"></a>    python3 triotrain/model_training/tutorial/create_metadata.py
<a id="__codelineno-44-6" name="__codelineno-44-6" href="#__codelineno-44-6"></a>&quot;&quot;&quot;
<a id="__codelineno-44-7" name="__codelineno-44-7" href="#__codelineno-44-7"></a>
<a id="__codelineno-44-8" name="__codelineno-44-8" href="#__codelineno-44-8"></a>import os
<a id="__codelineno-44-9" name="__codelineno-44-9" href="#__codelineno-44-9"></a>from pathlib import Path
<a id="__codelineno-44-10" name="__codelineno-44-10" href="#__codelineno-44-10"></a>from sys import path
<a id="__codelineno-44-11" name="__codelineno-44-11" href="#__codelineno-44-11"></a>
<a id="__codelineno-44-12" name="__codelineno-44-12" href="#__codelineno-44-12"></a>import regex
<a id="__codelineno-44-13" name="__codelineno-44-13" href="#__codelineno-44-13"></a>
<a id="__codelineno-44-14" name="__codelineno-44-14" href="#__codelineno-44-14"></a># get the absolute path to the triotrain/ dir
<a id="__codelineno-44-15" name="__codelineno-44-15" href="#__codelineno-44-15"></a>abs_path = Path(__file__).resolve()
<a id="__codelineno-44-16" name="__codelineno-44-16" href="#__codelineno-44-16"></a>module_path = str(abs_path.parent.parent.parent)
<a id="__codelineno-44-17" name="__codelineno-44-17" href="#__codelineno-44-17"></a>path.append(module_path)
<a id="__codelineno-44-18" name="__codelineno-44-18" href="#__codelineno-44-18"></a>
<a id="__codelineno-44-19" name="__codelineno-44-19" href="#__codelineno-44-19"></a>from helpers.dictionary import add_to_dict
<a id="__codelineno-44-20" name="__codelineno-44-20" href="#__codelineno-44-20"></a>from helpers.files import WriteFiles
<a id="__codelineno-44-21" name="__codelineno-44-21" href="#__codelineno-44-21"></a>from helpers.utils import get_logger
<a id="__codelineno-44-22" name="__codelineno-44-22" href="#__codelineno-44-22"></a>from helpers.wrapper import Wrapper, timestamp
<a id="__codelineno-44-23" name="__codelineno-44-23" href="#__codelineno-44-23"></a>
<a id="__codelineno-44-24" name="__codelineno-44-24" href="#__codelineno-44-24"></a># Collect start time
<a id="__codelineno-44-25" name="__codelineno-44-25" href="#__codelineno-44-25"></a>Wrapper(__file__, &quot;start&quot;).wrap_script(timestamp())
<a id="__codelineno-44-26" name="__codelineno-44-26" href="#__codelineno-44-26"></a>
<a id="__codelineno-44-27" name="__codelineno-44-27" href="#__codelineno-44-27"></a># Create error log
<a id="__codelineno-44-28" name="__codelineno-44-28" href="#__codelineno-44-28"></a>current_file = os.path.basename(__file__)
<a id="__codelineno-44-29" name="__codelineno-44-29" href="#__codelineno-44-29"></a>module_name = os.path.splitext(current_file)[0]
<a id="__codelineno-44-30" name="__codelineno-44-30" href="#__codelineno-44-30"></a>logger = get_logger(module_name)
<a id="__codelineno-44-31" name="__codelineno-44-31" href="#__codelineno-44-31"></a>
<a id="__codelineno-44-32" name="__codelineno-44-32" href="#__codelineno-44-32"></a>defaults = {
<a id="__codelineno-44-33" name="__codelineno-44-33" href="#__codelineno-44-33"></a>    &quot;RunOrder&quot;: 1,
<a id="__codelineno-44-34" name="__codelineno-44-34" href="#__codelineno-44-34"></a>    &quot;RunName&quot;: &quot;Human_tutorial&quot;,
<a id="__codelineno-44-35" name="__codelineno-44-35" href="#__codelineno-44-35"></a>    &quot;ChildSampleID&quot;: &quot;NA24385&quot;,
<a id="__codelineno-44-36" name="__codelineno-44-36" href="#__codelineno-44-36"></a>    &quot;ChildLabID&quot;: &quot;HG002&quot;,
<a id="__codelineno-44-37" name="__codelineno-44-37" href="#__codelineno-44-37"></a>    &quot;FatherSampleID&quot;: &quot;NA24149&quot;,
<a id="__codelineno-44-38" name="__codelineno-44-38" href="#__codelineno-44-38"></a>    &quot;FatherLabID&quot;: &quot;HG003&quot;,
<a id="__codelineno-44-39" name="__codelineno-44-39" href="#__codelineno-44-39"></a>    &quot;MotherSampleID&quot;: &quot;NA24695&quot;,
<a id="__codelineno-44-40" name="__codelineno-44-40" href="#__codelineno-44-40"></a>    &quot;MotherLabID&quot;: &quot;HG004&quot;,
<a id="__codelineno-44-41" name="__codelineno-44-41" href="#__codelineno-44-41"></a>    &quot;ChildSex&quot;: &quot;M&quot;,
<a id="__codelineno-44-42" name="__codelineno-44-42" href="#__codelineno-44-42"></a>    &quot;RefFASTA&quot;: &quot;/triotrain/variant_calling/data/GIAB/reference/GRCh38_no_alt_analysis_set.fasta&quot;,
<a id="__codelineno-44-43" name="__codelineno-44-43" href="#__codelineno-44-43"></a>    &quot;PopVCF&quot;: &quot;/triotrain/variant_calling/data/GIAB/allele_freq/cohort.release_missing2ref.no_calls.vcf.gz&quot;,
<a id="__codelineno-44-44" name="__codelineno-44-44" href="#__codelineno-44-44"></a>    &quot;RegionsFile&quot;: &quot;NA&quot;,
<a id="__codelineno-44-45" name="__codelineno-44-45" href="#__codelineno-44-45"></a>    &quot;ChildReadsBAM&quot;: &quot;/triotrain/variant_calling/data/GIAB/bam/HG002.GRCh38.2x250.bam&quot;,
<a id="__codelineno-44-46" name="__codelineno-44-46" href="#__codelineno-44-46"></a>    &quot;ChildTruthVCF&quot;: &quot;/triotrain/variant_calling/data/GIAB/benchmark/HG002_GRCh38_1_22_v4.2.1_benchmark.vcf.gz&quot;,
<a id="__codelineno-44-47" name="__codelineno-44-47" href="#__codelineno-44-47"></a>    &quot;ChildCallableBED&quot;: &quot;/triotrain/variant_calling/data/GIAB/benchmark/HG002_GRCh38_1_22_v4.2.1_benchmark_noinconsistent.bed&quot;,
<a id="__codelineno-44-48" name="__codelineno-44-48" href="#__codelineno-44-48"></a>    &quot;FatherReadsBAM&quot;: &quot;/triotrain/variant_calling/data/GIAB/bam/HG003.GRCh38.2x250.bam&quot;,
<a id="__codelineno-44-49" name="__codelineno-44-49" href="#__codelineno-44-49"></a>    &quot;FatherTruthVCF&quot;: &quot;/triotrain/variant_calling/data/GIAB/benchmark/HG003_GRCh38_1_22_v4.2.1_benchmark.vcf.gz&quot;,
<a id="__codelineno-44-50" name="__codelineno-44-50" href="#__codelineno-44-50"></a>    &quot;FatherCallableBED&quot;: &quot;/triotrain/variant_calling/data/GIAB/benchmark/HG003_GRCh38_1_22_v4.2.1_benchmark_noinconsistent.bed&quot;,
<a id="__codelineno-44-51" name="__codelineno-44-51" href="#__codelineno-44-51"></a>    &quot;MotherReadsBAM&quot;: &quot;/triotrain/variant_calling/data/GIAB/bam/HG004.GRCh38.2x250.bam&quot;,
<a id="__codelineno-44-52" name="__codelineno-44-52" href="#__codelineno-44-52"></a>    &quot;MotherTruthVCF&quot;: &quot;/triotrain/variant_calling/data/GIAB/benchmark/HG004_GRCh38_1_22_v4.2.1_benchmark.vcf.gz&quot;,
<a id="__codelineno-44-53" name="__codelineno-44-53" href="#__codelineno-44-53"></a>    &quot;MotherCallableBED&quot;: &quot;/triotrain/variant_calling/data/GIAB/benchmark/HG004_GRCh38_1_22_v4.2.1_benchmark_noinconsistent.bed&quot;,
<a id="__codelineno-44-54" name="__codelineno-44-54" href="#__codelineno-44-54"></a>    &quot;Test1ReadsBAM&quot;: &quot;/triotrain/variant_calling/data/GIAB/bam/HG005.GRCh38_full_plus_hs38d1_analysis_set_minus_alts.300x.bam&quot;,
<a id="__codelineno-44-55" name="__codelineno-44-55" href="#__codelineno-44-55"></a>    &quot;Test1TruthVCF&quot;: &quot;/triotrain/variant_calling/data/GIAB/benchmark/HG005_GRCh38_1_22_v4.2.1_benchmark.vcf.gz&quot;,
<a id="__codelineno-44-56" name="__codelineno-44-56" href="#__codelineno-44-56"></a>    &quot;Test1CallableBED&quot;: &quot;/triotrain/variant_calling/data/GIAB/benchmark/HG005_GRCh38_1_22_v4.2.1_benchmark.bed&quot;,
<a id="__codelineno-44-57" name="__codelineno-44-57" href="#__codelineno-44-57"></a>    &quot;Test2ReadsBAM&quot;: &quot;/triotrain/variant_calling/data/GIAB/bam/HG006.GRCh38_full_plus_hs38d1_analysis_set_minus_alts.100x.bam&quot;,
<a id="__codelineno-44-58" name="__codelineno-44-58" href="#__codelineno-44-58"></a>    &quot;Test2TruthVCF&quot;: &quot;/triotrain/variant_calling/data/GIAB/benchmark/HG006_GRCh38_1_22_v4.2.1_benchmark.vcf.gz&quot;,
<a id="__codelineno-44-59" name="__codelineno-44-59" href="#__codelineno-44-59"></a>    &quot;Test2CallableBED&quot;: &quot;/triotrain/variant_calling/data/GIAB/benchmark/HG006_GRCh38_1_22_v4.2.1_benchmark.bed&quot;,
<a id="__codelineno-44-60" name="__codelineno-44-60" href="#__codelineno-44-60"></a>    &quot;Test3ReadsBAM&quot;: &quot;/triotrain/variant_calling/data/GIAB/bam/HG007.GRCh38_full_plus_hs38d1_analysis_set_minus_alts.100x.bam&quot;,
<a id="__codelineno-44-61" name="__codelineno-44-61" href="#__codelineno-44-61"></a>    &quot;Test3TruthVCF&quot;: &quot;/triotrain/variant_calling/data/GIAB/benchmark/HG007_GRCh38_1_22_v4.2.1_benchmark.vcf.gz&quot;,
<a id="__codelineno-44-62" name="__codelineno-44-62" href="#__codelineno-44-62"></a>    &quot;Test3CallableBED&quot;: &quot;/triotrain/variant_calling/data/GIAB/benchmark/HG007_GRCh38_1_22_v4.2.1_benchmark.bed&quot;,
<a id="__codelineno-44-63" name="__codelineno-44-63" href="#__codelineno-44-63"></a>}
<a id="__codelineno-44-64" name="__codelineno-44-64" href="#__codelineno-44-64"></a>
<a id="__codelineno-44-65" name="__codelineno-44-65" href="#__codelineno-44-65"></a>cwd = Path.cwd()
<a id="__codelineno-44-66" name="__codelineno-44-66" href="#__codelineno-44-66"></a>output_dict = dict()
<a id="__codelineno-44-67" name="__codelineno-44-67" href="#__codelineno-44-67"></a>
<a id="__codelineno-44-68" name="__codelineno-44-68" href="#__codelineno-44-68"></a>for k, v in defaults.items():
<a id="__codelineno-44-69" name="__codelineno-44-69" href="#__codelineno-44-69"></a>    match = regex.search(r&quot;\/triotrain\/variant_calling\/.*&quot;, str(v))
<a id="__codelineno-44-70" name="__codelineno-44-70" href="#__codelineno-44-70"></a>    if match:
<a id="__codelineno-44-71" name="__codelineno-44-71" href="#__codelineno-44-71"></a>        add_to_dict(
<a id="__codelineno-44-72" name="__codelineno-44-72" href="#__codelineno-44-72"></a>            update_dict=output_dict,
<a id="__codelineno-44-73" name="__codelineno-44-73" href="#__codelineno-44-73"></a>            new_key=k,
<a id="__codelineno-44-74" name="__codelineno-44-74" href="#__codelineno-44-74"></a>            new_val=f&quot;{cwd}{v}&quot;,
<a id="__codelineno-44-75" name="__codelineno-44-75" href="#__codelineno-44-75"></a>            logger=logger,
<a id="__codelineno-44-76" name="__codelineno-44-76" href="#__codelineno-44-76"></a>            logger_msg=&quot;[tutorial]&quot;,
<a id="__codelineno-44-77" name="__codelineno-44-77" href="#__codelineno-44-77"></a>        )
<a id="__codelineno-44-78" name="__codelineno-44-78" href="#__codelineno-44-78"></a>    else:
<a id="__codelineno-44-79" name="__codelineno-44-79" href="#__codelineno-44-79"></a>        add_to_dict(
<a id="__codelineno-44-80" name="__codelineno-44-80" href="#__codelineno-44-80"></a>            update_dict=output_dict,
<a id="__codelineno-44-81" name="__codelineno-44-81" href="#__codelineno-44-81"></a>            new_key=k,
<a id="__codelineno-44-82" name="__codelineno-44-82" href="#__codelineno-44-82"></a>            new_val=v,
<a id="__codelineno-44-83" name="__codelineno-44-83" href="#__codelineno-44-83"></a>            logger=logger,
<a id="__codelineno-44-84" name="__codelineno-44-84" href="#__codelineno-44-84"></a>            logger_msg=&quot;[tutorial]&quot;,
<a id="__codelineno-44-85" name="__codelineno-44-85" href="#__codelineno-44-85"></a>        )
<a id="__codelineno-44-86" name="__codelineno-44-86" href="#__codelineno-44-86"></a>
<a id="__codelineno-44-87" name="__codelineno-44-87" href="#__codelineno-44-87"></a>output_file = WriteFiles(
<a id="__codelineno-44-88" name="__codelineno-44-88" href="#__codelineno-44-88"></a>    path_to_file=str(cwd / &quot;triotrain&quot; / &quot;model_training&quot; / &quot;tutorial&quot;),
<a id="__codelineno-44-89" name="__codelineno-44-89" href="#__codelineno-44-89"></a>    file=f&quot;GIAB.Human_tutorial_metadata.csv&quot;,
<a id="__codelineno-44-90" name="__codelineno-44-90" href="#__codelineno-44-90"></a>    logger=logger,
<a id="__codelineno-44-91" name="__codelineno-44-91" href="#__codelineno-44-91"></a>    logger_msg=&quot;[tutorial]&quot;,
<a id="__codelineno-44-92" name="__codelineno-44-92" href="#__codelineno-44-92"></a>)
<a id="__codelineno-44-93" name="__codelineno-44-93" href="#__codelineno-44-93"></a>
<a id="__codelineno-44-94" name="__codelineno-44-94" href="#__codelineno-44-94"></a>output_file.check_missing()
<a id="__codelineno-44-95" name="__codelineno-44-95" href="#__codelineno-44-95"></a>
<a id="__codelineno-44-96" name="__codelineno-44-96" href="#__codelineno-44-96"></a>if output_file.file_exists:
<a id="__codelineno-44-97" name="__codelineno-44-97" href="#__codelineno-44-97"></a>    logger.info(&quot;[tutorial]: tutorial metadata file already exists... SKIPPING AHEAD&quot;)
<a id="__codelineno-44-98" name="__codelineno-44-98" href="#__codelineno-44-98"></a>else:
<a id="__codelineno-44-99" name="__codelineno-44-99" href="#__codelineno-44-99"></a>    logger.info(&quot;[tutorial]: creating a tutorial metadata file now...&quot;)
<a id="__codelineno-44-100" name="__codelineno-44-100" href="#__codelineno-44-100"></a>    output_file.add_rows(col_names=output_dict.keys(), data_dict=output_dict)
<a id="__codelineno-44-101" name="__codelineno-44-101" href="#__codelineno-44-101"></a>    output_file.check_missing()
<a id="__codelineno-44-102" name="__codelineno-44-102" href="#__codelineno-44-102"></a>    if output_file.file_exists:
<a id="__codelineno-44-103" name="__codelineno-44-103" href="#__codelineno-44-103"></a>        logger.info(&quot;[tutorial]: successfully created the tutorial metadata file&quot;)
<a id="__codelineno-44-104" name="__codelineno-44-104" href="#__codelineno-44-104"></a>
<a id="__codelineno-44-105" name="__codelineno-44-105" href="#__codelineno-44-105"></a># Collect end time
<a id="__codelineno-44-106" name="__codelineno-44-106" href="#__codelineno-44-106"></a>Wrapper(__file__, &quot;start&quot;).wrap_script(timestamp())
</code></pre></div>
</details>
<details class="success">
<summary>Check | Tutorial Metadata File</summary>
<div class="highlight"><span class="filename">Run the following at the command line:</span><pre><span></span><code><a id="__codelineno-45-1" name="__codelineno-45-1" href="#__codelineno-45-1"></a>ls<span class="w"> </span>triotrain/model_training/tutorial
</code></pre></div>
<div class="highlight"><span class="filename">Expected output:</span><pre><span></span><code><a id="__codelineno-46-1" name="__codelineno-46-1" href="#__codelineno-46-1"></a>create_metadata.py<span class="w">  </span>estimate.py<span class="w">  </span>GIAB.Human_tutorial_metadata.csv<span class="w">  </span>__init__.py<span class="w">  </span>resources_used.json
</code></pre></div>
</details>
<h3 id="c-slurm-resource-config-file-json">c. <a href="../../user-guide/usage_guide/#resource-config-format">SLURM Resource Config File (<code>.json</code>)</a></h3>
<p>The last required input we need for TrioTrain is the SLURM resource config file (<code>.json</code>). This file tells TrioTrain what resources to request from your HPC cluster when submitting SLURM jobs. [</p>
<details class="example">
<summary>Example | Resource Config File</summary>
<div class="highlight"><span class="filename">triotrain/model_training/tutorial/resources_used.json</span><pre><span></span><code><a id="__codelineno-47-1" name="__codelineno-47-1" href="#__codelineno-47-1"></a>{
<a id="__codelineno-47-2" name="__codelineno-47-2" href="#__codelineno-47-2"></a>    &quot;make_examples&quot;: {
<a id="__codelineno-47-3" name="__codelineno-47-3" href="#__codelineno-47-3"></a>        &quot;partition&quot;: &quot;hpc5,hpc6,BioCompute&quot;,
<a id="__codelineno-47-4" name="__codelineno-47-4" href="#__codelineno-47-4"></a>        &quot;nodes&quot;: 1,
<a id="__codelineno-47-5" name="__codelineno-47-5" href="#__codelineno-47-5"></a>        &quot;ntasks&quot;: 40,
<a id="__codelineno-47-6" name="__codelineno-47-6" href="#__codelineno-47-6"></a>        &quot;mem&quot;: 379067,
<a id="__codelineno-47-7" name="__codelineno-47-7" href="#__codelineno-47-7"></a>        &quot;CPUmem&quot;: 9000,
<a id="__codelineno-47-8" name="__codelineno-47-8" href="#__codelineno-47-8"></a>        &quot;time&quot;: &quot;0-2:00:00&quot;,
<a id="__codelineno-47-9" name="__codelineno-47-9" href="#__codelineno-47-9"></a>        &quot;account&quot;: &quot;animalsci&quot;,
<a id="__codelineno-47-10" name="__codelineno-47-10" href="#__codelineno-47-10"></a>        &quot;email&quot;: &quot;jakth2@mail.missouri.edu&quot;
<a id="__codelineno-47-11" name="__codelineno-47-11" href="#__codelineno-47-11"></a>    },
<a id="__codelineno-47-12" name="__codelineno-47-12" href="#__codelineno-47-12"></a>    &quot;beam_shuffle&quot;: {
<a id="__codelineno-47-13" name="__codelineno-47-13" href="#__codelineno-47-13"></a>        &quot;partition&quot;: &quot;hpc5,BioCompute&quot;,
<a id="__codelineno-47-14" name="__codelineno-47-14" href="#__codelineno-47-14"></a>        &quot;nodes&quot;: 1,
<a id="__codelineno-47-15" name="__codelineno-47-15" href="#__codelineno-47-15"></a>        &quot;ntasks&quot;: 40,
<a id="__codelineno-47-16" name="__codelineno-47-16" href="#__codelineno-47-16"></a>        &quot;mem&quot;: 379067,
<a id="__codelineno-47-17" name="__codelineno-47-17" href="#__codelineno-47-17"></a>        &quot;time&quot;: &quot;0-2:00:00&quot;,
<a id="__codelineno-47-18" name="__codelineno-47-18" href="#__codelineno-47-18"></a>        &quot;account&quot;: &quot;animalsci&quot;,
<a id="__codelineno-47-19" name="__codelineno-47-19" href="#__codelineno-47-19"></a>        &quot;email&quot;: &quot;jakth2@mail.missouri.edu&quot;
<a id="__codelineno-47-20" name="__codelineno-47-20" href="#__codelineno-47-20"></a>    },
<a id="__codelineno-47-21" name="__codelineno-47-21" href="#__codelineno-47-21"></a>    &quot;re_shuffle&quot;: {
<a id="__codelineno-47-22" name="__codelineno-47-22" href="#__codelineno-47-22"></a>        &quot;partition&quot;: &quot;BioCompute,Lewis&quot;,
<a id="__codelineno-47-23" name="__codelineno-47-23" href="#__codelineno-47-23"></a>        &quot;nodes&quot;: 1,
<a id="__codelineno-47-24" name="__codelineno-47-24" href="#__codelineno-47-24"></a>        &quot;ntasks&quot;: 1,
<a id="__codelineno-47-25" name="__codelineno-47-25" href="#__codelineno-47-25"></a>        &quot;mem&quot;: &quot;200G&quot;,
<a id="__codelineno-47-26" name="__codelineno-47-26" href="#__codelineno-47-26"></a>        &quot;time&quot;: &quot;0-10:00:00&quot;,
<a id="__codelineno-47-27" name="__codelineno-47-27" href="#__codelineno-47-27"></a>        &quot;account&quot;: &quot;animalsci&quot;,
<a id="__codelineno-47-28" name="__codelineno-47-28" href="#__codelineno-47-28"></a>        &quot;email&quot;: &quot;jakth2@mail.missouri.edu&quot;
<a id="__codelineno-47-29" name="__codelineno-47-29" href="#__codelineno-47-29"></a>    },
<a id="__codelineno-47-30" name="__codelineno-47-30" href="#__codelineno-47-30"></a>    &quot;train_eval&quot;: {
<a id="__codelineno-47-31" name="__codelineno-47-31" href="#__codelineno-47-31"></a>        &quot;partition&quot;: &quot;gpu3&quot;,
<a id="__codelineno-47-32" name="__codelineno-47-32" href="#__codelineno-47-32"></a>        &quot;gres&quot;: &quot;gpu:2&quot;,
<a id="__codelineno-47-33" name="__codelineno-47-33" href="#__codelineno-47-33"></a>        &quot;nodes&quot;: 1,
<a id="__codelineno-47-34" name="__codelineno-47-34" href="#__codelineno-47-34"></a>        &quot;ntasks&quot;: 16,
<a id="__codelineno-47-35" name="__codelineno-47-35" href="#__codelineno-47-35"></a>        &quot;mem&quot;: &quot;0&quot;,
<a id="__codelineno-47-36" name="__codelineno-47-36" href="#__codelineno-47-36"></a>        &quot;time&quot;: &quot;2-00:00:00&quot;,
<a id="__codelineno-47-37" name="__codelineno-47-37" href="#__codelineno-47-37"></a>        &quot;account&quot;: &quot;animalsci&quot;,
<a id="__codelineno-47-38" name="__codelineno-47-38" href="#__codelineno-47-38"></a>        &quot;email&quot;: &quot;jakth2@mail.missouri.edu&quot;
<a id="__codelineno-47-39" name="__codelineno-47-39" href="#__codelineno-47-39"></a>    },
<a id="__codelineno-47-40" name="__codelineno-47-40" href="#__codelineno-47-40"></a>    &quot;select_ckpt&quot;: {
<a id="__codelineno-47-41" name="__codelineno-47-41" href="#__codelineno-47-41"></a>        &quot;partition&quot;: &quot;BioCompute,Lewis&quot;,
<a id="__codelineno-47-42" name="__codelineno-47-42" href="#__codelineno-47-42"></a>        &quot;nodes&quot;: 1,
<a id="__codelineno-47-43" name="__codelineno-47-43" href="#__codelineno-47-43"></a>        &quot;ntasks&quot;: 1,
<a id="__codelineno-47-44" name="__codelineno-47-44" href="#__codelineno-47-44"></a>        &quot;mem&quot;: 500,
<a id="__codelineno-47-45" name="__codelineno-47-45" href="#__codelineno-47-45"></a>        &quot;time&quot;: &quot;0-00:30:00&quot;,
<a id="__codelineno-47-46" name="__codelineno-47-46" href="#__codelineno-47-46"></a>        &quot;account&quot;: &quot;animalsci&quot;,
<a id="__codelineno-47-47" name="__codelineno-47-47" href="#__codelineno-47-47"></a>        &quot;email&quot;: &quot;jakth2@mail.missouri.edu&quot;
<a id="__codelineno-47-48" name="__codelineno-47-48" href="#__codelineno-47-48"></a>    },
<a id="__codelineno-47-49" name="__codelineno-47-49" href="#__codelineno-47-49"></a>    &quot;call_variants&quot;: {
<a id="__codelineno-47-50" name="__codelineno-47-50" href="#__codelineno-47-50"></a>        &quot;partition&quot;: &quot;hpc5,BioCompute&quot;,
<a id="__codelineno-47-51" name="__codelineno-47-51" href="#__codelineno-47-51"></a>        &quot;nodes&quot;: 1,
<a id="__codelineno-47-52" name="__codelineno-47-52" href="#__codelineno-47-52"></a>        &quot;ntasks&quot;: 40,
<a id="__codelineno-47-53" name="__codelineno-47-53" href="#__codelineno-47-53"></a>        &quot;mem&quot;: 379067,
<a id="__codelineno-47-54" name="__codelineno-47-54" href="#__codelineno-47-54"></a>        &quot;time&quot;: &quot;2-00:00:00&quot;,
<a id="__codelineno-47-55" name="__codelineno-47-55" href="#__codelineno-47-55"></a>        &quot;account&quot;: &quot;animalsci&quot;,
<a id="__codelineno-47-56" name="__codelineno-47-56" href="#__codelineno-47-56"></a>        &quot;email&quot;: &quot;jakth2@mail.missouri.edu&quot;
<a id="__codelineno-47-57" name="__codelineno-47-57" href="#__codelineno-47-57"></a>    },
<a id="__codelineno-47-58" name="__codelineno-47-58" href="#__codelineno-47-58"></a>    &quot;compare_happy&quot;: {
<a id="__codelineno-47-59" name="__codelineno-47-59" href="#__codelineno-47-59"></a>        &quot;partition&quot;: &quot;BioCompute,hpc5&quot;,
<a id="__codelineno-47-60" name="__codelineno-47-60" href="#__codelineno-47-60"></a>        &quot;nodes&quot;: 1,
<a id="__codelineno-47-61" name="__codelineno-47-61" href="#__codelineno-47-61"></a>        &quot;ntasks&quot;: 40,
<a id="__codelineno-47-62" name="__codelineno-47-62" href="#__codelineno-47-62"></a>        &quot;mem&quot;: &quot;300G&quot;,
<a id="__codelineno-47-63" name="__codelineno-47-63" href="#__codelineno-47-63"></a>        &quot;time&quot;: &quot;0-12:00:00&quot;,
<a id="__codelineno-47-64" name="__codelineno-47-64" href="#__codelineno-47-64"></a>        &quot;account&quot;: &quot;biocommunity&quot;,
<a id="__codelineno-47-65" name="__codelineno-47-65" href="#__codelineno-47-65"></a>        &quot;email&quot;: &quot;jakth2@mail.missouri.edu&quot;
<a id="__codelineno-47-66" name="__codelineno-47-66" href="#__codelineno-47-66"></a>    },
<a id="__codelineno-47-67" name="__codelineno-47-67" href="#__codelineno-47-67"></a>    &quot;convert_happy&quot;: {
<a id="__codelineno-47-68" name="__codelineno-47-68" href="#__codelineno-47-68"></a>        &quot;partition&quot;: &quot;BioCompute,hpc3,hpc5,Lewis&quot;,
<a id="__codelineno-47-69" name="__codelineno-47-69" href="#__codelineno-47-69"></a>        &quot;nodes&quot;: 1,
<a id="__codelineno-47-70" name="__codelineno-47-70" href="#__codelineno-47-70"></a>        &quot;ntasks&quot;: 24,
<a id="__codelineno-47-71" name="__codelineno-47-71" href="#__codelineno-47-71"></a>        &quot;mem&quot;: &quot;120G&quot;,
<a id="__codelineno-47-72" name="__codelineno-47-72" href="#__codelineno-47-72"></a>        &quot;time&quot;: &quot;0-05:00:00&quot;,
<a id="__codelineno-47-73" name="__codelineno-47-73" href="#__codelineno-47-73"></a>        &quot;account&quot;: &quot;biocommunity&quot;,
<a id="__codelineno-47-74" name="__codelineno-47-74" href="#__codelineno-47-74"></a>        &quot;email&quot;: &quot;jakth2@mail.missouri.edu&quot;
<a id="__codelineno-47-75" name="__codelineno-47-75" href="#__codelineno-47-75"></a>    },
<a id="__codelineno-47-76" name="__codelineno-47-76" href="#__codelineno-47-76"></a>    &quot;show_examples&quot;: {
<a id="__codelineno-47-77" name="__codelineno-47-77" href="#__codelineno-47-77"></a>        &quot;partition&quot;: &quot;BioCompute,Lewis&quot;,
<a id="__codelineno-47-78" name="__codelineno-47-78" href="#__codelineno-47-78"></a>        &quot;nodes&quot;: 1,
<a id="__codelineno-47-79" name="__codelineno-47-79" href="#__codelineno-47-79"></a>        &quot;mem&quot;: &quot;1G&quot;,
<a id="__codelineno-47-80" name="__codelineno-47-80" href="#__codelineno-47-80"></a>        &quot;time&quot;: &quot;0-02:00:00&quot;,
<a id="__codelineno-47-81" name="__codelineno-47-81" href="#__codelineno-47-81"></a>        &quot;account&quot;: &quot;animalsci&quot;,
<a id="__codelineno-47-82" name="__codelineno-47-82" href="#__codelineno-47-82"></a>        &quot;email&quot;: &quot;jakth2@mail.missouri.edu&quot;
<a id="__codelineno-47-83" name="__codelineno-47-83" href="#__codelineno-47-83"></a>    },
<a id="__codelineno-47-84" name="__codelineno-47-84" href="#__codelineno-47-84"></a>    &quot;summary&quot;: {
<a id="__codelineno-47-85" name="__codelineno-47-85" href="#__codelineno-47-85"></a>        &quot;partition&quot;: &quot;hpc5,hpc6,Lewis&quot;,
<a id="__codelineno-47-86" name="__codelineno-47-86" href="#__codelineno-47-86"></a>        &quot;nodes&quot;: 1,
<a id="__codelineno-47-87" name="__codelineno-47-87" href="#__codelineno-47-87"></a>        &quot;ntasks&quot;: 4,
<a id="__codelineno-47-88" name="__codelineno-47-88" href="#__codelineno-47-88"></a>        &quot;mem&quot;: &quot;40G&quot;,
<a id="__codelineno-47-89" name="__codelineno-47-89" href="#__codelineno-47-89"></a>        &quot;time&quot;: &quot;0-10:00:00&quot;,
<a id="__codelineno-47-90" name="__codelineno-47-90" href="#__codelineno-47-90"></a>        &quot;account&quot;: &quot;schnabellab&quot;,
<a id="__codelineno-47-91" name="__codelineno-47-91" href="#__codelineno-47-91"></a>        &quot;email&quot;: &quot;jakth2@mail.missouri.edu&quot;
<a id="__codelineno-47-92" name="__codelineno-47-92" href="#__codelineno-47-92"></a>    }
<a id="__codelineno-47-93" name="__codelineno-47-93" href="#__codelineno-47-93"></a>}
</code></pre></div>
</details>
<p>The hardware listed above for each phase (e.g. <code>mem</code>, <code>ntasks</code> , <code>gpus</code>, etc.) vary, as some phases are memory intensive. These values should not be interpreted as the minimum or the optimum resources required for each phase. The MU Lewis Research Computing Cluster is heterogenous, the example config file requests resources to maximize the number of compute nodes when running memory-intensive phases.</p>
<p>For the tutorial, copy the above example into a new file, and manually edit the SBATCH parameters to match your HPC cluster (i.e. changing the partition list, account, and email address). This new resource config file is passed to TrioTrain via <code>-r &lt;/path/to/new_file.json&gt;</code></p>
<hr />
<h2 id="8-run-shuffling-demo">8. Run Shuffling Demo</h2>
<p><strong>Shuffling the labeled examples is a critical step to re-training DeepVariant because the model assumes successive training images are independent of one another.</strong> DeepVariant includes an Apache Beam pipeline that puts training examples in random genomic order. However, in our experience, getting the Python SDK for Beam, Apache Spark and SLURM to cooperate is a dependency nightmare.</p>
<p>Our alternative approach, referred to as "Region Shuffling", splits the autosomes and the X chromosome into subset regions before making the labeled examples. Using the previously created <code>.dict</code> file created for the reference, TrioTrain will determine how many regions are required automatically.</p>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>For our analyses in bovine genomes, we exclude parts of the genome from our region shuffling, including:</p>
<ol>
<li>
<p>the Y chromosome and the Mitochondrial genome &mdash; due to limitations with the <code>ARS-UCD1.2_Btau5.0.1Y</code> reference genome</p>
</li>
<li>
<p>the unmapped reads &mdash;  due to a large volume of contigs containing a small amount of variants that can not be easily distributed across 60-150+ shuffling regions</p>
</li>
</ol>
<p>TrioTrain will automatically search the reference genome <code>.dict</code> file for contigs labeled with <code>Y</code> and <code>M/MT</code> and will ignore them prior to defining shuffling regions.</p>
<p>Use the <code>--unmapped-reads</code> flag to provide a contig prefix pattern to exclude. The default value (NKLS) is approprate for the <code>ARS-UCD1.2_Btau5.0.1Y</code> reference genome.</p>
</div>
<details class="note">
<summary>Note | Identifying Contig Labels</summary>
<div class="highlight"><span class="filename">Run the following at the command line:</span><pre><span></span><code><a id="__codelineno-48-1" name="__codelineno-48-1" href="#__codelineno-48-1"></a>cat<span class="w"> </span>ARS-UCD1.2_Btau5.0.1Y.dict<span class="w"> </span><span class="p">|</span><span class="w"> </span>awk<span class="w"> </span><span class="s1">&#39;{if($1 ~ &quot;@SQ&quot;) print $2}&#39;</span><span class="w"> </span><span class="p">|</span><span class="w"> </span>cut<span class="w"> </span>-c4-7<span class="w"> </span><span class="p">|</span><span class="w"> </span>sort<span class="w"> </span><span class="p">|</span><span class="w"> </span>uniq<span class="w"> </span>-c<span class="w"> </span><span class="p">|</span><span class="w"> </span>sort<span class="w"> </span>-k1
</code></pre></div>
<div class="highlight"><span class="filename">Check output</span><pre><span></span><code><a id="__codelineno-49-1" name="__codelineno-49-1" href="#__codelineno-49-1"></a><span class="w">    </span><span class="m">1</span><span class="w"> </span><span class="m">1</span>
<a id="__codelineno-49-2" name="__codelineno-49-2" href="#__codelineno-49-2"></a><span class="w">    </span><span class="m">1</span><span class="w"> </span><span class="m">10</span>
<a id="__codelineno-49-3" name="__codelineno-49-3" href="#__codelineno-49-3"></a><span class="w">    </span><span class="m">1</span><span class="w"> </span><span class="m">11</span>
<a id="__codelineno-49-4" name="__codelineno-49-4" href="#__codelineno-49-4"></a><span class="w">    </span><span class="m">1</span><span class="w"> </span><span class="m">12</span>
<a id="__codelineno-49-5" name="__codelineno-49-5" href="#__codelineno-49-5"></a><span class="w">    </span><span class="m">1</span><span class="w"> </span><span class="m">13</span>
<a id="__codelineno-49-6" name="__codelineno-49-6" href="#__codelineno-49-6"></a><span class="w">    </span><span class="m">1</span><span class="w"> </span><span class="m">14</span>
<a id="__codelineno-49-7" name="__codelineno-49-7" href="#__codelineno-49-7"></a><span class="w">    </span><span class="m">1</span><span class="w"> </span><span class="m">15</span>
<a id="__codelineno-49-8" name="__codelineno-49-8" href="#__codelineno-49-8"></a><span class="w">    </span><span class="m">1</span><span class="w"> </span><span class="m">16</span>
<a id="__codelineno-49-9" name="__codelineno-49-9" href="#__codelineno-49-9"></a><span class="w">    </span><span class="m">1</span><span class="w"> </span><span class="m">17</span>
<a id="__codelineno-49-10" name="__codelineno-49-10" href="#__codelineno-49-10"></a><span class="w">    </span><span class="m">1</span><span class="w"> </span><span class="m">18</span>
<a id="__codelineno-49-11" name="__codelineno-49-11" href="#__codelineno-49-11"></a><span class="w">    </span><span class="m">1</span><span class="w"> </span><span class="m">19</span>
<a id="__codelineno-49-12" name="__codelineno-49-12" href="#__codelineno-49-12"></a><span class="w">    </span><span class="m">1</span><span class="w"> </span><span class="m">2</span>
<a id="__codelineno-49-13" name="__codelineno-49-13" href="#__codelineno-49-13"></a><span class="w">    </span><span class="m">1</span><span class="w"> </span><span class="m">20</span>
<a id="__codelineno-49-14" name="__codelineno-49-14" href="#__codelineno-49-14"></a><span class="w">    </span><span class="m">1</span><span class="w"> </span><span class="m">21</span>
<a id="__codelineno-49-15" name="__codelineno-49-15" href="#__codelineno-49-15"></a><span class="w">    </span><span class="m">1</span><span class="w"> </span><span class="m">22</span>
<a id="__codelineno-49-16" name="__codelineno-49-16" href="#__codelineno-49-16"></a><span class="w">    </span><span class="m">1</span><span class="w"> </span><span class="m">23</span>
<a id="__codelineno-49-17" name="__codelineno-49-17" href="#__codelineno-49-17"></a><span class="w">    </span><span class="m">1</span><span class="w"> </span><span class="m">24</span>
<a id="__codelineno-49-18" name="__codelineno-49-18" href="#__codelineno-49-18"></a><span class="w">    </span><span class="m">1</span><span class="w"> </span><span class="m">25</span>
<a id="__codelineno-49-19" name="__codelineno-49-19" href="#__codelineno-49-19"></a><span class="w">    </span><span class="m">1</span><span class="w"> </span><span class="m">26</span>
<a id="__codelineno-49-20" name="__codelineno-49-20" href="#__codelineno-49-20"></a><span class="w">    </span><span class="m">1</span><span class="w"> </span><span class="m">27</span>
<a id="__codelineno-49-21" name="__codelineno-49-21" href="#__codelineno-49-21"></a><span class="w">    </span><span class="m">1</span><span class="w"> </span><span class="m">28</span>
<a id="__codelineno-49-22" name="__codelineno-49-22" href="#__codelineno-49-22"></a><span class="w">    </span><span class="m">1</span><span class="w"> </span><span class="m">29</span>
<a id="__codelineno-49-23" name="__codelineno-49-23" href="#__codelineno-49-23"></a><span class="w">    </span><span class="m">1</span><span class="w"> </span><span class="m">3</span>
<a id="__codelineno-49-24" name="__codelineno-49-24" href="#__codelineno-49-24"></a><span class="w">    </span><span class="m">1</span><span class="w"> </span><span class="m">4</span>
<a id="__codelineno-49-25" name="__codelineno-49-25" href="#__codelineno-49-25"></a><span class="w">    </span><span class="m">1</span><span class="w"> </span><span class="m">5</span>
<a id="__codelineno-49-26" name="__codelineno-49-26" href="#__codelineno-49-26"></a><span class="w">    </span><span class="m">1</span><span class="w"> </span><span class="m">6</span>
<a id="__codelineno-49-27" name="__codelineno-49-27" href="#__codelineno-49-27"></a><span class="w">    </span><span class="m">1</span><span class="w"> </span><span class="m">7</span>
<a id="__codelineno-49-28" name="__codelineno-49-28" href="#__codelineno-49-28"></a><span class="w">    </span><span class="m">1</span><span class="w"> </span><span class="m">8</span>
<a id="__codelineno-49-29" name="__codelineno-49-29" href="#__codelineno-49-29"></a><span class="w">    </span><span class="m">1</span><span class="w"> </span><span class="m">9</span>
<a id="__codelineno-49-30" name="__codelineno-49-30" href="#__codelineno-49-30"></a><span class="w">    </span><span class="m">1</span><span class="w"> </span>MT
<a id="__codelineno-49-31" name="__codelineno-49-31" href="#__codelineno-49-31"></a><span class="w">    </span><span class="m">1</span><span class="w"> </span>X
<a id="__codelineno-49-32" name="__codelineno-49-32" href="#__codelineno-49-32"></a><span class="w">    </span><span class="m">1</span><span class="w"> </span>Y
<a id="__codelineno-49-33" name="__codelineno-49-33" href="#__codelineno-49-33"></a><span class="w">    </span><span class="m">2180</span><span class="w"> </span>NKLS
</code></pre></div>
</details>
<hr />
<p>A shuffling region is defined in a non-overlapping, 0-based <code>BED</code> file. The goal is to create a subset of labeled examples which are representative of the whole genome, yet small enough to be shuffled within the memory of a single compute node. After defining the shuffling regions, labeled and shuffled examples are created by running parallel SLURM jobs. In our testing with bovine genomes, constraining each region to produce ~200,000 labeled examples works well.</p>
<p>In our experience, numerous small jobs running quickly minimizes wall time between phases. We also priorize a large number of regions, as this further increases the randomness during the final <code>re_shuffle</code> processes that randomizes the order regions are given to the model.</p>
<details class="example">
<summary>Example | Computing Power</summary>
<p>Given ~370G mem across 40 cores, DeepVariant will shuffle each region's examples in around 6 minutes, which can be submitted in parallel to reduce wall time.</p>
</details>
<details class="note">
<summary>Note | Output Volume</summary>
<p>A typical cattle genome with ~8 million variants results in around ~65 regions, resulting in (65 regions x 40 cores) tfrecord output files. TrioTrain repeats this process for shuffling, resulting in another (65 regions x 40 cores) shuffled tfrecord output files. This is repeated for all 3 genomes within a trio resulting in a large volume of individual files (65 regions x 40 cores x 2 phases per genome).</p>
<p>The final step of our approach, referrred to as <code>re_shuffle</code>, condenses the 40 tfrecord files for a region, reducing the number of files fed to DeepVariant per genome back down to 65. To further improve shuffling, we randomize the order regions are provided to DeepVariant.</p>
<p><strong>If your species of interest produces a large volume of variants (&gt; 8 million / genome), TrioTrain can easily overload your cluster's I/O capabilities, or write more than 10,000 files to a directory, or overwhelm the SLURM scheduler by submitting thousands of jobs simultaneously.</strong> Future versions of TrioTrain will address these challenges, but proceed with caution; your cluster's SysAdmin will thank you!</p>
</details>
<p><strong>This shuffling demo enables you to adjust TrioTrain's parameters for estimating how many shuffling regions to use for each genome.</strong> The two primary flags which control Region Shuffling are:</p>
<ol>
<li>
<p><code>--max-examples</code> &mdash; defines the upper limit of total number of examples to produce for each region; values greater than the default (200,000) will increase the wall time.</p>
</li>
<li>
<p><code>--est-examples</code> &mdash; defines expectations of the number of examples DeepVariant produces per genomic variant; the default value (1.5) is appropriate for bovine genomes, but with the human GIAB samples, we found a value of 1 to be more appropriate.</p>
</li>
</ol>
<p>Additionally, TrioTrain bases the Region Shuffling calculations on the total number of variants found in each genome. This value can vary drastically across training genomes, so TrioTrain uses the number of PASS variants within the corresponding TruthVCF. A larger number of truth variants results in more regions made per region.</p>
<p><strong>Before running TrioTrain in a new mamalian genome or on a new HPC cluster, we strongly recommend completing the region shuffling demo using the smallest chromosome possible using the <code>--demo-chr=&lt;region_literal&gt;</code> flag.</strong></p>
<details class="note">
<summary>Note | Working in Other Species</summary>
<p>The appropriate values for <code>max-examples</code> and <code>est-examples</code> may vary widely across species, as the number of examples DeepVariant produces depends on several factors including:</p>
<ul>
<li>
<p>genome size</p>
</li>
<li>
<p>the reference genome used</p>
</li>
<li>
<p>how many variants are identified in an individual genome</p>
</li>
</ul>
<p>For example, the F1-hybrid offspring in cattle are crosses between diverent lineages, resulting in an abnormal amount heterzygous genotypes compared to typical cattle genomes. Instead of typical 8 million variants per genome, these samples produced 20+ million variants per genome.</p>
</details>
<hr />
<p>For this tutorial, complete the demo for Human genomes by <strong>editing the following at the command to include your new SLURM config file</strong>:</p>
<div class="highlight"><span class="filename">Run the following at the command line:</span><pre><span></span><code><a id="__codelineno-50-1" name="__codelineno-50-1" href="#__codelineno-50-1"></a><span class="c1"># You can add the --dry-run flag to this command to confirm the TrioTrain pipeline runs smoothly before submitting jobs to the queue</span>
<a id="__codelineno-50-2" name="__codelineno-50-2" href="#__codelineno-50-2"></a><span class="nb">source</span><span class="w"> </span>./scripts/setup/modules.sh
<a id="__codelineno-50-3" name="__codelineno-50-3" href="#__codelineno-50-3"></a><span class="nb">source</span><span class="w"> </span>./scripts/start_conda.sh<span class="w">                 </span><span class="c1"># Ensure the previously built conda env is active</span>
<a id="__codelineno-50-4" name="__codelineno-50-4" href="#__codelineno-50-4"></a>python3<span class="w"> </span>triotrain/run_trio_train.py<span class="w">                                                                         </span><span class="se">\</span>
<a id="__codelineno-50-5" name="__codelineno-50-5" href="#__codelineno-50-5"></a><span class="w">    </span>-g<span class="w"> </span>Father<span class="w">                                                                                               </span><span class="se">\</span>
<a id="__codelineno-50-6" name="__codelineno-50-6" href="#__codelineno-50-6"></a><span class="w">    </span>--unmapped-reads<span class="w"> </span>chrUn<span class="w">                                                                                  </span><span class="se">\</span>
<a id="__codelineno-50-7" name="__codelineno-50-7" href="#__codelineno-50-7"></a><span class="w">    </span>-m<span class="w"> </span>triotrain/model_training/tutorial/GIAB.Human_tutorial_metadata.csv<span class="w">                                   </span><span class="se">\</span>
<a id="__codelineno-50-8" name="__codelineno-50-8" href="#__codelineno-50-8"></a><span class="w">    </span>-n<span class="w"> </span>demo<span class="w">                                                                                                 </span><span class="se">\</span>
<a id="__codelineno-50-9" name="__codelineno-50-9" href="#__codelineno-50-9"></a><span class="w">    </span>--demo-mode<span class="w">                                                                                             </span><span class="se">\</span>
<a id="__codelineno-50-10" name="__codelineno-50-10" href="#__codelineno-50-10"></a><span class="w">    </span>--demo-chr<span class="w"> </span>chr21<span class="w">                                                                                        </span><span class="se">\</span>
<a id="__codelineno-50-11" name="__codelineno-50-11" href="#__codelineno-50-11"></a><span class="w">    </span>--num-tests<span class="w"> </span><span class="m">3</span><span class="w">                                                                                           </span><span class="se">\</span>
<a id="__codelineno-50-12" name="__codelineno-50-12" href="#__codelineno-50-12"></a><span class="w">    </span>--custom-checkpoint<span class="w"> </span>triotrain/model_training/pretrained_models/v1.4.0_withIS_withAF/wgs_af.model.ckpt<span class="w">   </span><span class="se">\</span>
<a id="__codelineno-50-13" name="__codelineno-50-13" href="#__codelineno-50-13"></a><span class="w">    </span>--output<span class="w"> </span>../TUTORIAL<span class="w">                                                                                    </span><span class="se">\</span>
<a id="__codelineno-50-14" name="__codelineno-50-14" href="#__codelineno-50-14"></a><span class="w">    </span>-r<span class="w"> </span>&lt;/path/to/new_file.json&gt;
</code></pre></div>
<hr />
<p>This demo will produce and submit the following SLURM jobs:</p>
<ol>
<li><code>make_examples</code> &mdash; for both Father &amp; Child</li>
<li><code>beam_shuffle</code> &mdash; for both Father &amp; Child</li>
<li><code>re_shuffle</code> &mdash; for both Father &amp; Child</li>
<li><code>call_variants</code> &mdash; for just the Father</li>
</ol>
<hr />
<p>There are (6) types of output files from running the demo:</p>
<table>
<thead>
<tr>
<th>File Extension</th>
<th>Number of Files (Shards)</th>
<th>Genome Used</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>labeled.tfrecords*.gz</code></td>
<td>N</td>
<td>Father, Child</td>
</tr>
<tr>
<td><code>labeled.tfrecords*.gz.example_info.json</code></td>
<td>N</td>
<td>Father, Child</td>
</tr>
<tr>
<td><code>labeled.shuffled*.tfrecord.gz</code></td>
<td>N</td>
<td>Father, Child</td>
</tr>
<tr>
<td><code>labeled.shuffled.dataset_config.pbtxt</code></td>
<td>1</td>
<td>Father, Child</td>
</tr>
<tr>
<td><code>labeled.shuffled.merged.dataset_config.pbtxt</code></td>
<td>1</td>
<td>Father, Child</td>
</tr>
<tr>
<td><code>labeled.shuffled.merged.tfrecord.gz</code></td>
<td>1</td>
<td>Father, Child</td>
</tr>
<tr>
<td><em>N = number of CPUs requested in your <code>resources_used.json</code> file, shards are numbered 0 &mdash; (N - 1).</em></td>
<td></td>
<td></td>
</tr>
</tbody>
</table>
<details class="success">
<summary>Expected Output | Father Shuffling:</summary>
<div class="highlight"><span class="filename">Run the following at the command line:</span><pre><span></span><code><a id="__codelineno-51-1" name="__codelineno-51-1" href="#__codelineno-51-1"></a>ls<span class="w"> </span>../TUTORIAL/demo/Human_tutorial/examples/<span class="w"> </span><span class="p">|</span><span class="w"> </span>grep<span class="w"> </span>Father
</code></pre></div>
<div class="highlight"><span class="filename">Check output</span><pre><span></span><code><a id="__codelineno-52-1" name="__codelineno-52-1" href="#__codelineno-52-1"></a>Father.chr21.labeled.shuffled-00000-of-00040.tfrecord.gz
<a id="__codelineno-52-2" name="__codelineno-52-2" href="#__codelineno-52-2"></a>Father.chr21.labeled.shuffled-00001-of-00040.tfrecord.gz
<a id="__codelineno-52-3" name="__codelineno-52-3" href="#__codelineno-52-3"></a>Father.chr21.labeled.shuffled-00002-of-00040.tfrecord.gz
<a id="__codelineno-52-4" name="__codelineno-52-4" href="#__codelineno-52-4"></a>Father.chr21.labeled.shuffled-00003-of-00040.tfrecord.gz
<a id="__codelineno-52-5" name="__codelineno-52-5" href="#__codelineno-52-5"></a>Father.chr21.labeled.shuffled-00004-of-00040.tfrecord.gz
<a id="__codelineno-52-6" name="__codelineno-52-6" href="#__codelineno-52-6"></a>Father.chr21.labeled.shuffled-00005-of-00040.tfrecord.gz
<a id="__codelineno-52-7" name="__codelineno-52-7" href="#__codelineno-52-7"></a>Father.chr21.labeled.shuffled-00006-of-00040.tfrecord.gz
<a id="__codelineno-52-8" name="__codelineno-52-8" href="#__codelineno-52-8"></a>Father.chr21.labeled.shuffled-00007-of-00040.tfrecord.gz
<a id="__codelineno-52-9" name="__codelineno-52-9" href="#__codelineno-52-9"></a>Father.chr21.labeled.shuffled-00008-of-00040.tfrecord.gz
<a id="__codelineno-52-10" name="__codelineno-52-10" href="#__codelineno-52-10"></a>Father.chr21.labeled.shuffled-00009-of-00040.tfrecord.gz
<a id="__codelineno-52-11" name="__codelineno-52-11" href="#__codelineno-52-11"></a>Father.chr21.labeled.shuffled-00010-of-00040.tfrecord.gz
<a id="__codelineno-52-12" name="__codelineno-52-12" href="#__codelineno-52-12"></a>Father.chr21.labeled.shuffled-00011-of-00040.tfrecord.gz
<a id="__codelineno-52-13" name="__codelineno-52-13" href="#__codelineno-52-13"></a>Father.chr21.labeled.shuffled-00012-of-00040.tfrecord.gz
<a id="__codelineno-52-14" name="__codelineno-52-14" href="#__codelineno-52-14"></a>Father.chr21.labeled.shuffled-00013-of-00040.tfrecord.gz
<a id="__codelineno-52-15" name="__codelineno-52-15" href="#__codelineno-52-15"></a>Father.chr21.labeled.shuffled-00014-of-00040.tfrecord.gz
<a id="__codelineno-52-16" name="__codelineno-52-16" href="#__codelineno-52-16"></a>Father.chr21.labeled.shuffled-00015-of-00040.tfrecord.gz
<a id="__codelineno-52-17" name="__codelineno-52-17" href="#__codelineno-52-17"></a>Father.chr21.labeled.shuffled-00016-of-00040.tfrecord.gz
<a id="__codelineno-52-18" name="__codelineno-52-18" href="#__codelineno-52-18"></a>Father.chr21.labeled.shuffled-00017-of-00040.tfrecord.gz
<a id="__codelineno-52-19" name="__codelineno-52-19" href="#__codelineno-52-19"></a>Father.chr21.labeled.shuffled-00018-of-00040.tfrecord.gz
<a id="__codelineno-52-20" name="__codelineno-52-20" href="#__codelineno-52-20"></a>Father.chr21.labeled.shuffled-00019-of-00040.tfrecord.gz
<a id="__codelineno-52-21" name="__codelineno-52-21" href="#__codelineno-52-21"></a>Father.chr21.labeled.shuffled-00020-of-00040.tfrecord.gz
<a id="__codelineno-52-22" name="__codelineno-52-22" href="#__codelineno-52-22"></a>Father.chr21.labeled.shuffled-00021-of-00040.tfrecord.gz
<a id="__codelineno-52-23" name="__codelineno-52-23" href="#__codelineno-52-23"></a>Father.chr21.labeled.shuffled-00022-of-00040.tfrecord.gz
<a id="__codelineno-52-24" name="__codelineno-52-24" href="#__codelineno-52-24"></a>Father.chr21.labeled.shuffled-00023-of-00040.tfrecord.gz
<a id="__codelineno-52-25" name="__codelineno-52-25" href="#__codelineno-52-25"></a>Father.chr21.labeled.shuffled-00024-of-00040.tfrecord.gz
<a id="__codelineno-52-26" name="__codelineno-52-26" href="#__codelineno-52-26"></a>Father.chr21.labeled.shuffled-00025-of-00040.tfrecord.gz
<a id="__codelineno-52-27" name="__codelineno-52-27" href="#__codelineno-52-27"></a>Father.chr21.labeled.shuffled-00026-of-00040.tfrecord.gz
<a id="__codelineno-52-28" name="__codelineno-52-28" href="#__codelineno-52-28"></a>Father.chr21.labeled.shuffled-00027-of-00040.tfrecord.gz
<a id="__codelineno-52-29" name="__codelineno-52-29" href="#__codelineno-52-29"></a>Father.chr21.labeled.shuffled-00028-of-00040.tfrecord.gz
<a id="__codelineno-52-30" name="__codelineno-52-30" href="#__codelineno-52-30"></a>Father.chr21.labeled.shuffled-00029-of-00040.tfrecord.gz
<a id="__codelineno-52-31" name="__codelineno-52-31" href="#__codelineno-52-31"></a>Father.chr21.labeled.shuffled-00030-of-00040.tfrecord.gz
<a id="__codelineno-52-32" name="__codelineno-52-32" href="#__codelineno-52-32"></a>Father.chr21.labeled.shuffled-00031-of-00040.tfrecord.gz
<a id="__codelineno-52-33" name="__codelineno-52-33" href="#__codelineno-52-33"></a>Father.chr21.labeled.shuffled-00032-of-00040.tfrecord.gz
<a id="__codelineno-52-34" name="__codelineno-52-34" href="#__codelineno-52-34"></a>Father.chr21.labeled.shuffled-00033-of-00040.tfrecord.gz
<a id="__codelineno-52-35" name="__codelineno-52-35" href="#__codelineno-52-35"></a>Father.chr21.labeled.shuffled-00034-of-00040.tfrecord.gz
<a id="__codelineno-52-36" name="__codelineno-52-36" href="#__codelineno-52-36"></a>Father.chr21.labeled.shuffled-00035-of-00040.tfrecord.gz
<a id="__codelineno-52-37" name="__codelineno-52-37" href="#__codelineno-52-37"></a>Father.chr21.labeled.shuffled-00036-of-00040.tfrecord.gz
<a id="__codelineno-52-38" name="__codelineno-52-38" href="#__codelineno-52-38"></a>Father.chr21.labeled.shuffled-00037-of-00040.tfrecord.gz
<a id="__codelineno-52-39" name="__codelineno-52-39" href="#__codelineno-52-39"></a>Father.chr21.labeled.shuffled-00038-of-00040.tfrecord.gz
<a id="__codelineno-52-40" name="__codelineno-52-40" href="#__codelineno-52-40"></a>Father.chr21.labeled.shuffled-00039-of-00040.tfrecord.gz
<a id="__codelineno-52-41" name="__codelineno-52-41" href="#__codelineno-52-41"></a>Father.chr21.labeled.shuffled.dataset_config.pbtxt
<a id="__codelineno-52-42" name="__codelineno-52-42" href="#__codelineno-52-42"></a>Father.chr21.labeled.tfrecords-00000-of-00040.gz
<a id="__codelineno-52-43" name="__codelineno-52-43" href="#__codelineno-52-43"></a>Father.chr21.labeled.tfrecords-00000-of-00040.gz.example_info.json
<a id="__codelineno-52-44" name="__codelineno-52-44" href="#__codelineno-52-44"></a>Father.chr21.labeled.tfrecords-00001-of-00040.gz
<a id="__codelineno-52-45" name="__codelineno-52-45" href="#__codelineno-52-45"></a>Father.chr21.labeled.tfrecords-00001-of-00040.gz.example_info.json
<a id="__codelineno-52-46" name="__codelineno-52-46" href="#__codelineno-52-46"></a>Father.chr21.labeled.tfrecords-00002-of-00040.gz
<a id="__codelineno-52-47" name="__codelineno-52-47" href="#__codelineno-52-47"></a>Father.chr21.labeled.tfrecords-00002-of-00040.gz.example_info.json
<a id="__codelineno-52-48" name="__codelineno-52-48" href="#__codelineno-52-48"></a>Father.chr21.labeled.tfrecords-00003-of-00040.gz
<a id="__codelineno-52-49" name="__codelineno-52-49" href="#__codelineno-52-49"></a>Father.chr21.labeled.tfrecords-00003-of-00040.gz.example_info.json
<a id="__codelineno-52-50" name="__codelineno-52-50" href="#__codelineno-52-50"></a>Father.chr21.labeled.tfrecords-00004-of-00040.gz
<a id="__codelineno-52-51" name="__codelineno-52-51" href="#__codelineno-52-51"></a>Father.chr21.labeled.tfrecords-00004-of-00040.gz.example_info.json
<a id="__codelineno-52-52" name="__codelineno-52-52" href="#__codelineno-52-52"></a>Father.chr21.labeled.tfrecords-00005-of-00040.gz
<a id="__codelineno-52-53" name="__codelineno-52-53" href="#__codelineno-52-53"></a>Father.chr21.labeled.tfrecords-00005-of-00040.gz.example_info.json
<a id="__codelineno-52-54" name="__codelineno-52-54" href="#__codelineno-52-54"></a>Father.chr21.labeled.tfrecords-00006-of-00040.gz
<a id="__codelineno-52-55" name="__codelineno-52-55" href="#__codelineno-52-55"></a>Father.chr21.labeled.tfrecords-00006-of-00040.gz.example_info.json
<a id="__codelineno-52-56" name="__codelineno-52-56" href="#__codelineno-52-56"></a>Father.chr21.labeled.tfrecords-00007-of-00040.gz
<a id="__codelineno-52-57" name="__codelineno-52-57" href="#__codelineno-52-57"></a>Father.chr21.labeled.tfrecords-00007-of-00040.gz.example_info.json
<a id="__codelineno-52-58" name="__codelineno-52-58" href="#__codelineno-52-58"></a>Father.chr21.labeled.tfrecords-00008-of-00040.gz
<a id="__codelineno-52-59" name="__codelineno-52-59" href="#__codelineno-52-59"></a>Father.chr21.labeled.tfrecords-00008-of-00040.gz.example_info.json
<a id="__codelineno-52-60" name="__codelineno-52-60" href="#__codelineno-52-60"></a>Father.chr21.labeled.tfrecords-00009-of-00040.gz
<a id="__codelineno-52-61" name="__codelineno-52-61" href="#__codelineno-52-61"></a>Father.chr21.labeled.tfrecords-00009-of-00040.gz.example_info.json
<a id="__codelineno-52-62" name="__codelineno-52-62" href="#__codelineno-52-62"></a>Father.chr21.labeled.tfrecords-00010-of-00040.gz
<a id="__codelineno-52-63" name="__codelineno-52-63" href="#__codelineno-52-63"></a>Father.chr21.labeled.tfrecords-00010-of-00040.gz.example_info.json
<a id="__codelineno-52-64" name="__codelineno-52-64" href="#__codelineno-52-64"></a>Father.chr21.labeled.tfrecords-00011-of-00040.gz
<a id="__codelineno-52-65" name="__codelineno-52-65" href="#__codelineno-52-65"></a>Father.chr21.labeled.tfrecords-00011-of-00040.gz.example_info.json
<a id="__codelineno-52-66" name="__codelineno-52-66" href="#__codelineno-52-66"></a>Father.chr21.labeled.tfrecords-00012-of-00040.gz
<a id="__codelineno-52-67" name="__codelineno-52-67" href="#__codelineno-52-67"></a>Father.chr21.labeled.tfrecords-00012-of-00040.gz.example_info.json
<a id="__codelineno-52-68" name="__codelineno-52-68" href="#__codelineno-52-68"></a>Father.chr21.labeled.tfrecords-00013-of-00040.gz
<a id="__codelineno-52-69" name="__codelineno-52-69" href="#__codelineno-52-69"></a>Father.chr21.labeled.tfrecords-00013-of-00040.gz.example_info.json
<a id="__codelineno-52-70" name="__codelineno-52-70" href="#__codelineno-52-70"></a>Father.chr21.labeled.tfrecords-00014-of-00040.gz
<a id="__codelineno-52-71" name="__codelineno-52-71" href="#__codelineno-52-71"></a>Father.chr21.labeled.tfrecords-00014-of-00040.gz.example_info.json
<a id="__codelineno-52-72" name="__codelineno-52-72" href="#__codelineno-52-72"></a>Father.chr21.labeled.tfrecords-00015-of-00040.gz
<a id="__codelineno-52-73" name="__codelineno-52-73" href="#__codelineno-52-73"></a>Father.chr21.labeled.tfrecords-00015-of-00040.gz.example_info.json
<a id="__codelineno-52-74" name="__codelineno-52-74" href="#__codelineno-52-74"></a>Father.chr21.labeled.tfrecords-00016-of-00040.gz
<a id="__codelineno-52-75" name="__codelineno-52-75" href="#__codelineno-52-75"></a>Father.chr21.labeled.tfrecords-00016-of-00040.gz.example_info.json
<a id="__codelineno-52-76" name="__codelineno-52-76" href="#__codelineno-52-76"></a>Father.chr21.labeled.tfrecords-00017-of-00040.gz
<a id="__codelineno-52-77" name="__codelineno-52-77" href="#__codelineno-52-77"></a>Father.chr21.labeled.tfrecords-00017-of-00040.gz.example_info.json
<a id="__codelineno-52-78" name="__codelineno-52-78" href="#__codelineno-52-78"></a>Father.chr21.labeled.tfrecords-00018-of-00040.gz
<a id="__codelineno-52-79" name="__codelineno-52-79" href="#__codelineno-52-79"></a>Father.chr21.labeled.tfrecords-00018-of-00040.gz.example_info.json
<a id="__codelineno-52-80" name="__codelineno-52-80" href="#__codelineno-52-80"></a>Father.chr21.labeled.tfrecords-00019-of-00040.gz
<a id="__codelineno-52-81" name="__codelineno-52-81" href="#__codelineno-52-81"></a>Father.chr21.labeled.tfrecords-00019-of-00040.gz.example_info.json
<a id="__codelineno-52-82" name="__codelineno-52-82" href="#__codelineno-52-82"></a>Father.chr21.labeled.tfrecords-00020-of-00040.gz
<a id="__codelineno-52-83" name="__codelineno-52-83" href="#__codelineno-52-83"></a>Father.chr21.labeled.tfrecords-00020-of-00040.gz.example_info.json
<a id="__codelineno-52-84" name="__codelineno-52-84" href="#__codelineno-52-84"></a>Father.chr21.labeled.tfrecords-00021-of-00040.gz
<a id="__codelineno-52-85" name="__codelineno-52-85" href="#__codelineno-52-85"></a>Father.chr21.labeled.tfrecords-00021-of-00040.gz.example_info.json
<a id="__codelineno-52-86" name="__codelineno-52-86" href="#__codelineno-52-86"></a>Father.chr21.labeled.tfrecords-00022-of-00040.gz
<a id="__codelineno-52-87" name="__codelineno-52-87" href="#__codelineno-52-87"></a>Father.chr21.labeled.tfrecords-00022-of-00040.gz.example_info.json
<a id="__codelineno-52-88" name="__codelineno-52-88" href="#__codelineno-52-88"></a>Father.chr21.labeled.tfrecords-00023-of-00040.gz
<a id="__codelineno-52-89" name="__codelineno-52-89" href="#__codelineno-52-89"></a>Father.chr21.labeled.tfrecords-00023-of-00040.gz.example_info.json
<a id="__codelineno-52-90" name="__codelineno-52-90" href="#__codelineno-52-90"></a>Father.chr21.labeled.tfrecords-00024-of-00040.gz
<a id="__codelineno-52-91" name="__codelineno-52-91" href="#__codelineno-52-91"></a>Father.chr21.labeled.tfrecords-00024-of-00040.gz.example_info.json
<a id="__codelineno-52-92" name="__codelineno-52-92" href="#__codelineno-52-92"></a>Father.chr21.labeled.tfrecords-00025-of-00040.gz
<a id="__codelineno-52-93" name="__codelineno-52-93" href="#__codelineno-52-93"></a>Father.chr21.labeled.tfrecords-00025-of-00040.gz.example_info.json
<a id="__codelineno-52-94" name="__codelineno-52-94" href="#__codelineno-52-94"></a>Father.chr21.labeled.tfrecords-00026-of-00040.gz
<a id="__codelineno-52-95" name="__codelineno-52-95" href="#__codelineno-52-95"></a>Father.chr21.labeled.tfrecords-00026-of-00040.gz.example_info.json
<a id="__codelineno-52-96" name="__codelineno-52-96" href="#__codelineno-52-96"></a>Father.chr21.labeled.tfrecords-00027-of-00040.gz
<a id="__codelineno-52-97" name="__codelineno-52-97" href="#__codelineno-52-97"></a>Father.chr21.labeled.tfrecords-00027-of-00040.gz.example_info.json
<a id="__codelineno-52-98" name="__codelineno-52-98" href="#__codelineno-52-98"></a>Father.chr21.labeled.tfrecords-00028-of-00040.gz
<a id="__codelineno-52-99" name="__codelineno-52-99" href="#__codelineno-52-99"></a>Father.chr21.labeled.tfrecords-00028-of-00040.gz.example_info.json
<a id="__codelineno-52-100" name="__codelineno-52-100" href="#__codelineno-52-100"></a>Father.chr21.labeled.tfrecords-00029-of-00040.gz
<a id="__codelineno-52-101" name="__codelineno-52-101" href="#__codelineno-52-101"></a>Father.chr21.labeled.tfrecords-00029-of-00040.gz.example_info.json
<a id="__codelineno-52-102" name="__codelineno-52-102" href="#__codelineno-52-102"></a>Father.chr21.labeled.tfrecords-00030-of-00040.gz
<a id="__codelineno-52-103" name="__codelineno-52-103" href="#__codelineno-52-103"></a>Father.chr21.labeled.tfrecords-00030-of-00040.gz.example_info.json
<a id="__codelineno-52-104" name="__codelineno-52-104" href="#__codelineno-52-104"></a>Father.chr21.labeled.tfrecords-00031-of-00040.gz
<a id="__codelineno-52-105" name="__codelineno-52-105" href="#__codelineno-52-105"></a>Father.chr21.labeled.tfrecords-00031-of-00040.gz.example_info.json
<a id="__codelineno-52-106" name="__codelineno-52-106" href="#__codelineno-52-106"></a>Father.chr21.labeled.tfrecords-00032-of-00040.gz
<a id="__codelineno-52-107" name="__codelineno-52-107" href="#__codelineno-52-107"></a>Father.chr21.labeled.tfrecords-00032-of-00040.gz.example_info.json
<a id="__codelineno-52-108" name="__codelineno-52-108" href="#__codelineno-52-108"></a>Father.chr21.labeled.tfrecords-00033-of-00040.gz
<a id="__codelineno-52-109" name="__codelineno-52-109" href="#__codelineno-52-109"></a>Father.chr21.labeled.tfrecords-00033-of-00040.gz.example_info.json
<a id="__codelineno-52-110" name="__codelineno-52-110" href="#__codelineno-52-110"></a>Father.chr21.labeled.tfrecords-00034-of-00040.gz
<a id="__codelineno-52-111" name="__codelineno-52-111" href="#__codelineno-52-111"></a>Father.chr21.labeled.tfrecords-00034-of-00040.gz.example_info.json
<a id="__codelineno-52-112" name="__codelineno-52-112" href="#__codelineno-52-112"></a>Father.chr21.labeled.tfrecords-00035-of-00040.gz
<a id="__codelineno-52-113" name="__codelineno-52-113" href="#__codelineno-52-113"></a>Father.chr21.labeled.tfrecords-00036-of-00040.gz.example_info.json
<a id="__codelineno-52-114" name="__codelineno-52-114" href="#__codelineno-52-114"></a>Father.chr21.labeled.tfrecords-00037-of-00040.gz
<a id="__codelineno-52-115" name="__codelineno-52-115" href="#__codelineno-52-115"></a>Father.chr21.labeled.tfrecords-00037-of-00040.gz.example_info.json
<a id="__codelineno-52-116" name="__codelineno-52-116" href="#__codelineno-52-116"></a>Father.chr21.labeled.tfrecords-00038-of-00040.gz
<a id="__codelineno-52-117" name="__codelineno-52-117" href="#__codelineno-52-117"></a>Father.chr21.labeled.tfrecords-00038-of-00040.gz.example_info.json
<a id="__codelineno-52-118" name="__codelineno-52-118" href="#__codelineno-52-118"></a>Father.chr21.labeled.tfrecords-00039-of-00040.gz
<a id="__codelineno-52-119" name="__codelineno-52-119" href="#__codelineno-52-119"></a>Father.chr21.labeled.tfrecords-00039-of-00040.gz.example_info.json
</code></pre></div>
<p><strong>Confirm that SLURM jobs completed successfully</strong>
<div class="highlight"><span class="filename">Run the following at the command line:</span><pre><span></span><code><a id="__codelineno-53-1" name="__codelineno-53-1" href="#__codelineno-53-1"></a>cat<span class="w"> </span>../TUTORIAL/demo/Human_tutorial/logs/tracking-Baseline-v1.4.0.log<span class="w"> </span><span class="p">|</span><span class="w"> </span>grep<span class="w"> </span>SUCCESS<span class="w"> </span><span class="p">|</span><span class="w"> </span>grep<span class="w"> </span>Father<span class="w"> </span><span class="p">|</span><span class="w"> </span>wc<span class="w"> </span>-l
</code></pre></div></p>
<div class="highlight"><span class="filename">Check output</span><pre><span></span><code><a id="__codelineno-54-1" name="__codelineno-54-1" href="#__codelineno-54-1"></a><span class="m">41</span>
</code></pre></div>
</details>
<details class="success">
<summary>Expected Output | Child Shuffling:</summary>
<div class="highlight"><span class="filename">Run the following at the command line:</span><pre><span></span><code><a id="__codelineno-55-1" name="__codelineno-55-1" href="#__codelineno-55-1"></a>ls<span class="w"> </span>../TUTORIAL/demo/Human_tutorial/examples/<span class="w"> </span><span class="p">|</span><span class="w"> </span>grep<span class="w"> </span>Child
</code></pre></div>
<div class="highlight"><span class="filename">Check output</span><pre><span></span><code><a id="__codelineno-56-1" name="__codelineno-56-1" href="#__codelineno-56-1"></a>Child.chr21.labeled.shuffled-00000-of-00040.tfrecord.gz
<a id="__codelineno-56-2" name="__codelineno-56-2" href="#__codelineno-56-2"></a>Child.chr21.labeled.shuffled-00001-of-00040.tfrecord.gz
<a id="__codelineno-56-3" name="__codelineno-56-3" href="#__codelineno-56-3"></a>Child.chr21.labeled.shuffled-00002-of-00040.tfrecord.gz
<a id="__codelineno-56-4" name="__codelineno-56-4" href="#__codelineno-56-4"></a>Child.chr21.labeled.shuffled-00003-of-00040.tfrecord.gz
<a id="__codelineno-56-5" name="__codelineno-56-5" href="#__codelineno-56-5"></a>Child.chr21.labeled.shuffled-00004-of-00040.tfrecord.gz
<a id="__codelineno-56-6" name="__codelineno-56-6" href="#__codelineno-56-6"></a>Child.chr21.labeled.shuffled-00005-of-00040.tfrecord.gz
<a id="__codelineno-56-7" name="__codelineno-56-7" href="#__codelineno-56-7"></a>Child.chr21.labeled.shuffled-00006-of-00040.tfrecord.gz
<a id="__codelineno-56-8" name="__codelineno-56-8" href="#__codelineno-56-8"></a>Child.chr21.labeled.shuffled-00007-of-00040.tfrecord.gz
<a id="__codelineno-56-9" name="__codelineno-56-9" href="#__codelineno-56-9"></a>Child.chr21.labeled.shuffled-00008-of-00040.tfrecord.gz
<a id="__codelineno-56-10" name="__codelineno-56-10" href="#__codelineno-56-10"></a>Child.chr21.labeled.shuffled-00009-of-00040.tfrecord.gz
<a id="__codelineno-56-11" name="__codelineno-56-11" href="#__codelineno-56-11"></a>Child.chr21.labeled.shuffled-00010-of-00040.tfrecord.gz
<a id="__codelineno-56-12" name="__codelineno-56-12" href="#__codelineno-56-12"></a>Child.chr21.labeled.shuffled-00011-of-00040.tfrecord.gz
<a id="__codelineno-56-13" name="__codelineno-56-13" href="#__codelineno-56-13"></a>Child.chr21.labeled.shuffled-00012-of-00040.tfrecord.gz
<a id="__codelineno-56-14" name="__codelineno-56-14" href="#__codelineno-56-14"></a>Child.chr21.labeled.shuffled-00013-of-00040.tfrecord.gz
<a id="__codelineno-56-15" name="__codelineno-56-15" href="#__codelineno-56-15"></a>Child.chr21.labeled.shuffled-00014-of-00040.tfrecord.gz
<a id="__codelineno-56-16" name="__codelineno-56-16" href="#__codelineno-56-16"></a>Child.chr21.labeled.shuffled-00015-of-00040.tfrecord.gz
<a id="__codelineno-56-17" name="__codelineno-56-17" href="#__codelineno-56-17"></a>Child.chr21.labeled.shuffled-00016-of-00040.tfrecord.gz
<a id="__codelineno-56-18" name="__codelineno-56-18" href="#__codelineno-56-18"></a>Child.chr21.labeled.shuffled-00017-of-00040.tfrecord.gz
<a id="__codelineno-56-19" name="__codelineno-56-19" href="#__codelineno-56-19"></a>Child.chr21.labeled.shuffled-00018-of-00040.tfrecord.gz
<a id="__codelineno-56-20" name="__codelineno-56-20" href="#__codelineno-56-20"></a>Child.chr21.labeled.shuffled-00019-of-00040.tfrecord.gz
<a id="__codelineno-56-21" name="__codelineno-56-21" href="#__codelineno-56-21"></a>Child.chr21.labeled.shuffled-00020-of-00040.tfrecord.gz
<a id="__codelineno-56-22" name="__codelineno-56-22" href="#__codelineno-56-22"></a>Child.chr21.labeled.shuffled-00021-of-00040.tfrecord.gz
<a id="__codelineno-56-23" name="__codelineno-56-23" href="#__codelineno-56-23"></a>Child.chr21.labeled.shuffled-00022-of-00040.tfrecord.gz
<a id="__codelineno-56-24" name="__codelineno-56-24" href="#__codelineno-56-24"></a>Child.chr21.labeled.shuffled-00023-of-00040.tfrecord.gz
<a id="__codelineno-56-25" name="__codelineno-56-25" href="#__codelineno-56-25"></a>Child.chr21.labeled.shuffled-00024-of-00040.tfrecord.gz
<a id="__codelineno-56-26" name="__codelineno-56-26" href="#__codelineno-56-26"></a>Child.chr21.labeled.shuffled-00025-of-00040.tfrecord.gz
<a id="__codelineno-56-27" name="__codelineno-56-27" href="#__codelineno-56-27"></a>Child.chr21.labeled.shuffled-00026-of-00040.tfrecord.gz
<a id="__codelineno-56-28" name="__codelineno-56-28" href="#__codelineno-56-28"></a>Child.chr21.labeled.shuffled-00027-of-00040.tfrecord.gz
<a id="__codelineno-56-29" name="__codelineno-56-29" href="#__codelineno-56-29"></a>Child.chr21.labeled.shuffled-00028-of-00040.tfrecord.gz
<a id="__codelineno-56-30" name="__codelineno-56-30" href="#__codelineno-56-30"></a>Child.chr21.labeled.shuffled-00029-of-00040.tfrecord.gz
<a id="__codelineno-56-31" name="__codelineno-56-31" href="#__codelineno-56-31"></a>Child.chr21.labeled.shuffled-00030-of-00040.tfrecord.gz
<a id="__codelineno-56-32" name="__codelineno-56-32" href="#__codelineno-56-32"></a>Child.chr21.labeled.shuffled-00031-of-00040.tfrecord.gz
<a id="__codelineno-56-33" name="__codelineno-56-33" href="#__codelineno-56-33"></a>Child.chr21.labeled.shuffled-00032-of-00040.tfrecord.gz
<a id="__codelineno-56-34" name="__codelineno-56-34" href="#__codelineno-56-34"></a>Child.chr21.labeled.shuffled-00033-of-00040.tfrecord.gz
<a id="__codelineno-56-35" name="__codelineno-56-35" href="#__codelineno-56-35"></a>Child.chr21.labeled.shuffled-00034-of-00040.tfrecord.gz
<a id="__codelineno-56-36" name="__codelineno-56-36" href="#__codelineno-56-36"></a>Child.chr21.labeled.shuffled-00035-of-00040.tfrecord.gz
<a id="__codelineno-56-37" name="__codelineno-56-37" href="#__codelineno-56-37"></a>Child.chr21.labeled.shuffled-00036-of-00040.tfrecord.gz
<a id="__codelineno-56-38" name="__codelineno-56-38" href="#__codelineno-56-38"></a>Child.chr21.labeled.shuffled-00037-of-00040.tfrecord.gz
<a id="__codelineno-56-39" name="__codelineno-56-39" href="#__codelineno-56-39"></a>Child.chr21.labeled.shuffled-00038-of-00040.tfrecord.gz
<a id="__codelineno-56-40" name="__codelineno-56-40" href="#__codelineno-56-40"></a>Child.chr21.labeled.shuffled-00039-of-00040.tfrecord.gz
<a id="__codelineno-56-41" name="__codelineno-56-41" href="#__codelineno-56-41"></a>Child.chr21.labeled.shuffled.dataset_config.pbtxt
<a id="__codelineno-56-42" name="__codelineno-56-42" href="#__codelineno-56-42"></a>Child.chr21.labeled.tfrecords-00000-of-00040.gz
<a id="__codelineno-56-43" name="__codelineno-56-43" href="#__codelineno-56-43"></a>Child.chr21.labeled.tfrecords-00000-of-00040.gz.example_info.json
<a id="__codelineno-56-44" name="__codelineno-56-44" href="#__codelineno-56-44"></a>Child.chr21.labeled.tfrecords-00001-of-00040.gz
<a id="__codelineno-56-45" name="__codelineno-56-45" href="#__codelineno-56-45"></a>Child.chr21.labeled.tfrecords-00001-of-00040.gz.example_info.json
<a id="__codelineno-56-46" name="__codelineno-56-46" href="#__codelineno-56-46"></a>Child.chr21.labeled.tfrecords-00002-of-00040.gz
<a id="__codelineno-56-47" name="__codelineno-56-47" href="#__codelineno-56-47"></a>Child.chr21.labeled.tfrecords-00002-of-00040.gz.example_info.json
<a id="__codelineno-56-48" name="__codelineno-56-48" href="#__codelineno-56-48"></a>Child.chr21.labeled.tfrecords-00003-of-00040.gz
<a id="__codelineno-56-49" name="__codelineno-56-49" href="#__codelineno-56-49"></a>Child.chr21.labeled.tfrecords-00003-of-00040.gz.example_info.json
<a id="__codelineno-56-50" name="__codelineno-56-50" href="#__codelineno-56-50"></a>Child.chr21.labeled.tfrecords-00004-of-00040.gz
<a id="__codelineno-56-51" name="__codelineno-56-51" href="#__codelineno-56-51"></a>Child.chr21.labeled.tfrecords-00004-of-00040.gz.example_info.json
<a id="__codelineno-56-52" name="__codelineno-56-52" href="#__codelineno-56-52"></a>Child.chr21.labeled.tfrecords-00005-of-00040.gz
<a id="__codelineno-56-53" name="__codelineno-56-53" href="#__codelineno-56-53"></a>Child.chr21.labeled.tfrecords-00005-of-00040.gz.example_info.json
<a id="__codelineno-56-54" name="__codelineno-56-54" href="#__codelineno-56-54"></a>Child.chr21.labeled.tfrecords-00006-of-00040.gz
<a id="__codelineno-56-55" name="__codelineno-56-55" href="#__codelineno-56-55"></a>Child.chr21.labeled.tfrecords-00006-of-00040.gz.example_info.json
<a id="__codelineno-56-56" name="__codelineno-56-56" href="#__codelineno-56-56"></a>Child.chr21.labeled.tfrecords-00007-of-00040.gz
<a id="__codelineno-56-57" name="__codelineno-56-57" href="#__codelineno-56-57"></a>Child.chr21.labeled.tfrecords-00007-of-00040.gz.example_info.json
<a id="__codelineno-56-58" name="__codelineno-56-58" href="#__codelineno-56-58"></a>Child.chr21.labeled.tfrecords-00008-of-00040.gz
<a id="__codelineno-56-59" name="__codelineno-56-59" href="#__codelineno-56-59"></a>Child.chr21.labeled.tfrecords-00008-of-00040.gz.example_info.json
<a id="__codelineno-56-60" name="__codelineno-56-60" href="#__codelineno-56-60"></a>Child.chr21.labeled.tfrecords-00009-of-00040.gz
<a id="__codelineno-56-61" name="__codelineno-56-61" href="#__codelineno-56-61"></a>Child.chr21.labeled.tfrecords-00009-of-00040.gz.example_info.json
<a id="__codelineno-56-62" name="__codelineno-56-62" href="#__codelineno-56-62"></a>Child.chr21.labeled.tfrecords-00010-of-00040.gz
<a id="__codelineno-56-63" name="__codelineno-56-63" href="#__codelineno-56-63"></a>Child.chr21.labeled.tfrecords-00010-of-00040.gz.example_info.json
<a id="__codelineno-56-64" name="__codelineno-56-64" href="#__codelineno-56-64"></a>Child.chr21.labeled.tfrecords-00011-of-00040.gz
<a id="__codelineno-56-65" name="__codelineno-56-65" href="#__codelineno-56-65"></a>Child.chr21.labeled.tfrecords-00011-of-00040.gz.example_info.json
<a id="__codelineno-56-66" name="__codelineno-56-66" href="#__codelineno-56-66"></a>Child.chr21.labeled.tfrecords-00012-of-00040.gz
<a id="__codelineno-56-67" name="__codelineno-56-67" href="#__codelineno-56-67"></a>Child.chr21.labeled.tfrecords-00012-of-00040.gz.example_info.json
<a id="__codelineno-56-68" name="__codelineno-56-68" href="#__codelineno-56-68"></a>Child.chr21.labeled.tfrecords-00013-of-00040.gz
<a id="__codelineno-56-69" name="__codelineno-56-69" href="#__codelineno-56-69"></a>Child.chr21.labeled.tfrecords-00013-of-00040.gz.example_info.json
<a id="__codelineno-56-70" name="__codelineno-56-70" href="#__codelineno-56-70"></a>Child.chr21.labeled.tfrecords-00014-of-00040.gz
<a id="__codelineno-56-71" name="__codelineno-56-71" href="#__codelineno-56-71"></a>Child.chr21.labeled.tfrecords-00014-of-00040.gz.example_info.json
<a id="__codelineno-56-72" name="__codelineno-56-72" href="#__codelineno-56-72"></a>Child.chr21.labeled.tfrecords-00015-of-00040.gz
<a id="__codelineno-56-73" name="__codelineno-56-73" href="#__codelineno-56-73"></a>Child.chr21.labeled.tfrecords-00015-of-00040.gz.example_info.json
<a id="__codelineno-56-74" name="__codelineno-56-74" href="#__codelineno-56-74"></a>Child.chr21.labeled.tfrecords-00016-of-00040.gz
<a id="__codelineno-56-75" name="__codelineno-56-75" href="#__codelineno-56-75"></a>Child.chr21.labeled.tfrecords-00016-of-00040.gz.example_info.json
<a id="__codelineno-56-76" name="__codelineno-56-76" href="#__codelineno-56-76"></a>Child.chr21.labeled.tfrecords-00017-of-00040.gz
<a id="__codelineno-56-77" name="__codelineno-56-77" href="#__codelineno-56-77"></a>Child.chr21.labeled.tfrecords-00017-of-00040.gz.example_info.json
<a id="__codelineno-56-78" name="__codelineno-56-78" href="#__codelineno-56-78"></a>Child.chr21.labeled.tfrecords-00018-of-00040.gz
<a id="__codelineno-56-79" name="__codelineno-56-79" href="#__codelineno-56-79"></a>Child.chr21.labeled.tfrecords-00018-of-00040.gz.example_info.json
<a id="__codelineno-56-80" name="__codelineno-56-80" href="#__codelineno-56-80"></a>Child.chr21.labeled.tfrecords-00019-of-00040.gz
<a id="__codelineno-56-81" name="__codelineno-56-81" href="#__codelineno-56-81"></a>Child.chr21.labeled.tfrecords-00019-of-00040.gz.example_info.json
<a id="__codelineno-56-82" name="__codelineno-56-82" href="#__codelineno-56-82"></a>Child.chr21.labeled.tfrecords-00020-of-00040.gz
<a id="__codelineno-56-83" name="__codelineno-56-83" href="#__codelineno-56-83"></a>Child.chr21.labeled.tfrecords-00020-of-00040.gz.example_info.json
<a id="__codelineno-56-84" name="__codelineno-56-84" href="#__codelineno-56-84"></a>Child.chr21.labeled.tfrecords-00021-of-00040.gz
<a id="__codelineno-56-85" name="__codelineno-56-85" href="#__codelineno-56-85"></a>Child.chr21.labeled.tfrecords-00021-of-00040.gz.example_info.json
<a id="__codelineno-56-86" name="__codelineno-56-86" href="#__codelineno-56-86"></a>Child.chr21.labeled.tfrecords-00022-of-00040.gz
<a id="__codelineno-56-87" name="__codelineno-56-87" href="#__codelineno-56-87"></a>Child.chr21.labeled.tfrecords-00022-of-00040.gz.example_info.json
<a id="__codelineno-56-88" name="__codelineno-56-88" href="#__codelineno-56-88"></a>Child.chr21.labeled.tfrecords-00023-of-00040.gz
<a id="__codelineno-56-89" name="__codelineno-56-89" href="#__codelineno-56-89"></a>Child.chr21.labeled.tfrecords-00023-of-00040.gz.example_info.json
<a id="__codelineno-56-90" name="__codelineno-56-90" href="#__codelineno-56-90"></a>Child.chr21.labeled.tfrecords-00024-of-00040.gz
<a id="__codelineno-56-91" name="__codelineno-56-91" href="#__codelineno-56-91"></a>Child.chr21.labeled.tfrecords-00024-of-00040.gz.example_info.json
<a id="__codelineno-56-92" name="__codelineno-56-92" href="#__codelineno-56-92"></a>Child.chr21.labeled.tfrecords-00025-of-00040.gz
<a id="__codelineno-56-93" name="__codelineno-56-93" href="#__codelineno-56-93"></a>Child.chr21.labeled.tfrecords-00025-of-00040.gz.example_info.json
<a id="__codelineno-56-94" name="__codelineno-56-94" href="#__codelineno-56-94"></a>Child.chr21.labeled.tfrecords-00026-of-00040.gz
<a id="__codelineno-56-95" name="__codelineno-56-95" href="#__codelineno-56-95"></a>Child.chr21.labeled.tfrecords-00026-of-00040.gz.example_info.json
<a id="__codelineno-56-96" name="__codelineno-56-96" href="#__codelineno-56-96"></a>Child.chr21.labeled.tfrecords-00027-of-00040.gz
<a id="__codelineno-56-97" name="__codelineno-56-97" href="#__codelineno-56-97"></a>Child.chr21.labeled.tfrecords-00027-of-00040.gz.example_info.json
<a id="__codelineno-56-98" name="__codelineno-56-98" href="#__codelineno-56-98"></a>Child.chr21.labeled.tfrecords-00028-of-00040.gz
<a id="__codelineno-56-99" name="__codelineno-56-99" href="#__codelineno-56-99"></a>Child.chr21.labeled.tfrecords-00028-of-00040.gz.example_info.json
<a id="__codelineno-56-100" name="__codelineno-56-100" href="#__codelineno-56-100"></a>Child.chr21.labeled.tfrecords-00029-of-00040.gz
<a id="__codelineno-56-101" name="__codelineno-56-101" href="#__codelineno-56-101"></a>Child.chr21.labeled.tfrecords-00029-of-00040.gz.example_info.json
<a id="__codelineno-56-102" name="__codelineno-56-102" href="#__codelineno-56-102"></a>Child.chr21.labeled.tfrecords-00030-of-00040.gz
<a id="__codelineno-56-103" name="__codelineno-56-103" href="#__codelineno-56-103"></a>Child.chr21.labeled.tfrecords-00030-of-00040.gz.example_info.json
<a id="__codelineno-56-104" name="__codelineno-56-104" href="#__codelineno-56-104"></a>Child.chr21.labeled.tfrecords-00031-of-00040.gz
<a id="__codelineno-56-105" name="__codelineno-56-105" href="#__codelineno-56-105"></a>Child.chr21.labeled.tfrecords-00031-of-00040.gz.example_info.json
<a id="__codelineno-56-106" name="__codelineno-56-106" href="#__codelineno-56-106"></a>Child.chr21.labeled.tfrecords-00032-of-00040.gz
<a id="__codelineno-56-107" name="__codelineno-56-107" href="#__codelineno-56-107"></a>Child.chr21.labeled.tfrecords-00032-of-00040.gz.example_info.json
<a id="__codelineno-56-108" name="__codelineno-56-108" href="#__codelineno-56-108"></a>Child.chr21.labeled.tfrecords-00033-of-00040.gz
<a id="__codelineno-56-109" name="__codelineno-56-109" href="#__codelineno-56-109"></a>Child.chr21.labeled.tfrecords-00033-of-00040.gz.example_info.json
<a id="__codelineno-56-110" name="__codelineno-56-110" href="#__codelineno-56-110"></a>Child.chr21.labeled.tfrecords-00034-of-00040.gz
<a id="__codelineno-56-111" name="__codelineno-56-111" href="#__codelineno-56-111"></a>Child.chr21.labeled.tfrecords-00034-of-00040.gz.example_info.json
<a id="__codelineno-56-112" name="__codelineno-56-112" href="#__codelineno-56-112"></a>Child.chr21.labeled.tfrecords-00035-of-00040.gz
<a id="__codelineno-56-113" name="__codelineno-56-113" href="#__codelineno-56-113"></a>Child.chr21.labeled.tfrecords-00035-of-00040.gz.example_info.json
<a id="__codelineno-56-114" name="__codelineno-56-114" href="#__codelineno-56-114"></a>Child.chr21.labeled.tfrecords-00036-of-00040.gz
<a id="__codelineno-56-115" name="__codelineno-56-115" href="#__codelineno-56-115"></a>Child.chr21.labeled.tfrecords-00036-of-00040.gz.example_info.json
<a id="__codelineno-56-116" name="__codelineno-56-116" href="#__codelineno-56-116"></a>Child.chr21.labeled.tfrecords-00037-of-00040.gz
<a id="__codelineno-56-117" name="__codelineno-56-117" href="#__codelineno-56-117"></a>Child.chr21.labeled.tfrecords-00037-of-00040.gz.example_info.json
<a id="__codelineno-56-118" name="__codelineno-56-118" href="#__codelineno-56-118"></a>Child.chr21.labeled.tfrecords-00038-of-00040.gz
<a id="__codelineno-56-119" name="__codelineno-56-119" href="#__codelineno-56-119"></a>Child.chr21.labeled.tfrecords-00038-of-00040.gz.example_info.json
<a id="__codelineno-56-120" name="__codelineno-56-120" href="#__codelineno-56-120"></a>Child.chr21.labeled.tfrecords-00039-of-00040.gz
<a id="__codelineno-56-121" name="__codelineno-56-121" href="#__codelineno-56-121"></a>Child.chr21.labeled.tfrecords-00039-of-00040.gz.example_info.json
</code></pre></div>
<p><strong>Confirm that SLURM jobs completed successfully</strong>
<div class="highlight"><span class="filename">Run the following at the command line:</span><pre><span></span><code><a id="__codelineno-57-1" name="__codelineno-57-1" href="#__codelineno-57-1"></a>cat<span class="w"> </span>../TUTORIAL/demo/Human_tutorial/logs/tracking-Baseline-v1.4.0.log<span class="w"> </span><span class="p">|</span><span class="w"> </span>grep<span class="w"> </span>SUCCESS<span class="w"> </span><span class="p">|</span><span class="w"> </span>grep<span class="w"> </span>Child<span class="w"> </span><span class="p">|</span><span class="w"> </span>wc<span class="w"> </span>-l
</code></pre></div></p>
<div class="highlight"><span class="filename">Check output</span><pre><span></span><code><a id="__codelineno-58-1" name="__codelineno-58-1" href="#__codelineno-58-1"></a><span class="m">41</span>
</code></pre></div>
</details>
<details class="success">
<summary>Expected Output | Benchmarking:</summary>
<div class="highlight"><span class="filename">Run the following at the command line:</span><pre><span></span><code><a id="__codelineno-59-1" name="__codelineno-59-1" href="#__codelineno-59-1"></a>less<span class="w"> </span>../TUTORIAL/demo/summary/Human_tutorial.SLURM.job_numbers.csv
</code></pre></div>
<div class="highlight"><span class="filename">Check output</span><pre><span></span><code><a id="__codelineno-60-1" name="__codelineno-60-1" href="#__codelineno-60-1"></a>AnalysisName,RunName,Parent,Phase,JobList
<a id="__codelineno-60-2" name="__codelineno-60-2" href="#__codelineno-60-2"></a>Baseline-v1.4.0,Human_tutorial,Father,make_examples,27669522
<a id="__codelineno-60-3" name="__codelineno-60-3" href="#__codelineno-60-3"></a>Baseline-v1.4.0,Human_tutorial,Father,beam_shuffle,27669523
<a id="__codelineno-60-4" name="__codelineno-60-4" href="#__codelineno-60-4"></a>Baseline-v1.4.0,Human_tutorial,Father,make_examples,27669524
<a id="__codelineno-60-5" name="__codelineno-60-5" href="#__codelineno-60-5"></a>Baseline-v1.4.0,Human_tutorial,Father,beam_shuffle,27669525
<a id="__codelineno-60-6" name="__codelineno-60-6" href="#__codelineno-60-6"></a><span class="c1"># The JobList column will differ based on SLURM job numbers</span>
</code></pre></div>
</details>
<p><strong>The following will provide a conservative estimate for the <code>--max-examples</code> and <code>--est-examples</code> parameters to ensure shuffling easily fits within your available memory.</strong></p>
<div class="highlight"><span class="filename">Run the following at the command line:</span><pre><span></span><code><a id="__codelineno-61-1" name="__codelineno-61-1" href="#__codelineno-61-1"></a><span class="nb">source</span><span class="w"> </span>./scripts/setup/modules.sh
<a id="__codelineno-61-2" name="__codelineno-61-2" href="#__codelineno-61-2"></a><span class="nb">source</span><span class="w"> </span>./scripts/start_conda.sh<span class="w">                 </span><span class="c1"># Ensure the previously built conda env is active</span>
<a id="__codelineno-61-3" name="__codelineno-61-3" href="#__codelineno-61-3"></a>python3<span class="w"> </span>triotrain/model_training/tutorial/estimate.py<span class="w">                           </span><span class="se">\</span>
<a id="__codelineno-61-4" name="__codelineno-61-4" href="#__codelineno-61-4"></a><span class="w">    </span>--vcf-file<span class="w"> </span>../TUTORIAL/demo/Human_tutorial/test_Father/test1_chr21.vcf.gz<span class="w">   </span><span class="se">\</span>
<a id="__codelineno-61-5" name="__codelineno-61-5" href="#__codelineno-61-5"></a><span class="w">    </span>-g<span class="w"> </span>Father<span class="w">                                                                   </span><span class="se">\</span>
<a id="__codelineno-61-6" name="__codelineno-61-6" href="#__codelineno-61-6"></a><span class="w">    </span>--demo-mode<span class="w">                                                                 </span><span class="se">\</span>
<a id="__codelineno-61-7" name="__codelineno-61-7" href="#__codelineno-61-7"></a><span class="w">    </span>--env-file<span class="w"> </span>../TUTORIAL/demo/envs/run1.env
</code></pre></div>
<details class="success">
<summary>Expected Output | Estimating TrioTrain Parameters:</summary>
<div class="highlight"><pre><span></span><code><a id="__codelineno-62-1" name="__codelineno-62-1" href="#__codelineno-62-1"></a><span class="o">=====</span><span class="w"> </span>start<span class="w"> </span>of<span class="w"> </span>triotrain/model_training/tutorial/estimate.py<span class="w"> </span>@<span class="w"> </span><span class="m">2023</span>-06-29<span class="w">  </span><span class="m">11</span>:25:27<span class="w"> </span><span class="o">=====</span>
<a id="__codelineno-62-2" name="__codelineno-62-2" href="#__codelineno-62-2"></a><span class="m">2023</span>-06-29<span class="w"> </span><span class="m">11</span>:25:27<span class="w"> </span>AM<span class="w"> </span>-<span class="w"> </span><span class="o">[</span>INFO<span class="o">]</span><span class="w"> </span>-<span class="w"> </span><span class="o">[</span>DEMO<span class="o">]</span><span class="w"> </span>-<span class="w"> </span><span class="o">[</span>TRIO1<span class="o">]</span><span class="w"> </span>-<span class="w"> </span><span class="o">[</span>count_variants<span class="o">]</span><span class="w"> </span>-<span class="w"> </span><span class="o">[</span>Father<span class="o">]</span>:<span class="w"> </span>number<span class="w"> </span>of<span class="w"> </span>REF/REF<span class="w"> </span>variants<span class="w"> </span>found<span class="w"> </span><span class="p">|</span><span class="w"> </span><span class="m">44</span>,452
<a id="__codelineno-62-3" name="__codelineno-62-3" href="#__codelineno-62-3"></a><span class="m">2023</span>-06-29<span class="w"> </span><span class="m">11</span>:25:27<span class="w"> </span>AM<span class="w"> </span>-<span class="w"> </span><span class="o">[</span>INFO<span class="o">]</span><span class="w"> </span>-<span class="w"> </span><span class="o">[</span>DEMO<span class="o">]</span><span class="w"> </span>-<span class="w"> </span><span class="o">[</span>TRIO1<span class="o">]</span><span class="w"> </span>-<span class="w"> </span><span class="o">[</span>count_variants<span class="o">]</span><span class="w"> </span>-<span class="w"> </span><span class="o">[</span>Father<span class="o">]</span>:<span class="w"> </span>number<span class="w"> </span>of<span class="w"> </span>PASS<span class="w"> </span>variants<span class="w"> </span>found<span class="w"> </span><span class="p">|</span><span class="w"> </span><span class="m">114</span>,332
<a id="__codelineno-62-4" name="__codelineno-62-4" href="#__codelineno-62-4"></a><span class="m">2023</span>-06-29<span class="w"> </span><span class="m">11</span>:25:28<span class="w"> </span>AM<span class="w"> </span>-<span class="w"> </span><span class="o">[</span>INFO<span class="o">]</span><span class="w"> </span>-<span class="w"> </span><span class="o">[</span>DEMO<span class="o">]</span><span class="w"> </span>-<span class="w"> </span><span class="o">[</span>TRIO1<span class="o">]</span><span class="w"> </span>-<span class="w"> </span><span class="o">[</span>count_variants<span class="o">]</span><span class="w"> </span>-<span class="w"> </span><span class="o">[</span>Father<span class="o">]</span>:<span class="w"> </span>default<span class="w"> </span>maximum<span class="w"> </span>examples<span class="w"> </span>per<span class="w"> </span>region<span class="w"> </span><span class="p">|</span><span class="w"> </span><span class="m">200</span>,000
<a id="__codelineno-62-5" name="__codelineno-62-5" href="#__codelineno-62-5"></a><span class="m">2023</span>-06-29<span class="w"> </span><span class="m">11</span>:25:28<span class="w"> </span>AM<span class="w"> </span>-<span class="w"> </span><span class="o">[</span>INFO<span class="o">]</span><span class="w"> </span>-<span class="w"> </span><span class="o">[</span>DEMO<span class="o">]</span><span class="w"> </span>-<span class="w"> </span><span class="o">[</span>TRIO1<span class="o">]</span><span class="w"> </span>-<span class="w"> </span><span class="o">[</span>count_variants<span class="o">]</span><span class="w"> </span>-<span class="w"> </span><span class="o">[</span>Father<span class="o">]</span>:<span class="w"> </span>default<span class="w"> </span>value<span class="w"> </span><span class="k">for</span><span class="w"> </span>--max-examples<span class="w"> </span>is<span class="w"> </span>appropriate
<a id="__codelineno-62-6" name="__codelineno-62-6" href="#__codelineno-62-6"></a><span class="m">2023</span>-06-29<span class="w"> </span><span class="m">11</span>:25:28<span class="w"> </span>AM<span class="w"> </span>-<span class="w"> </span><span class="o">[</span>INFO<span class="o">]</span><span class="w"> </span>-<span class="w"> </span>adding<span class="w"> </span><span class="nv">Demo_TotalVariants</span><span class="o">=</span><span class="s1">&#39;114332&#39;</span>
<a id="__codelineno-62-7" name="__codelineno-62-7" href="#__codelineno-62-7"></a><span class="m">2023</span>-06-29<span class="w"> </span><span class="m">11</span>:25:28<span class="w"> </span>AM<span class="w"> </span>-<span class="w"> </span><span class="o">[</span>INFO<span class="o">]</span><span class="w"> </span>-<span class="w"> </span><span class="o">[</span>DEMO<span class="o">]</span><span class="w"> </span>-<span class="w"> </span><span class="o">[</span>TRIO1<span class="o">]</span><span class="w"> </span>-<span class="w"> </span><span class="o">[</span>count_variants<span class="o">]</span><span class="w"> </span>-<span class="w"> </span><span class="o">[</span>Father<span class="o">]</span>:<span class="w"> </span>added<span class="w"> </span><span class="s1">&#39;Demo_TotalVariants=114332&#39;</span><span class="w"> </span>to<span class="w"> </span>env<span class="w"> </span>file
<a id="__codelineno-62-8" name="__codelineno-62-8" href="#__codelineno-62-8"></a><span class="m">2023</span>-06-29<span class="w"> </span><span class="m">11</span>:25:28<span class="w"> </span>AM<span class="w"> </span>-<span class="w"> </span><span class="o">[</span>INFO<span class="o">]</span><span class="w"> </span>-<span class="w"> </span><span class="o">[</span>DEMO<span class="o">]</span><span class="w"> </span>-<span class="w"> </span><span class="o">[</span>TRIO1<span class="o">]</span><span class="w"> </span>-<span class="w"> </span><span class="o">[</span>count_variants<span class="o">]</span><span class="w"> </span>-<span class="w"> </span><span class="o">[</span>Father<span class="o">]</span>:<span class="w"> </span>number<span class="w"> </span>of<span class="w"> </span>examples<span class="w"> </span>made<span class="w"> </span><span class="p">|</span><span class="w"> </span><span class="m">75</span>,902
<a id="__codelineno-62-9" name="__codelineno-62-9" href="#__codelineno-62-9"></a><span class="m">2023</span>-06-29<span class="w"> </span><span class="m">11</span>:25:28<span class="w"> </span>AM<span class="w"> </span>-<span class="w"> </span><span class="o">[</span>INFO<span class="o">]</span><span class="w"> </span>-<span class="w"> </span><span class="o">[</span>DEMO<span class="o">]</span><span class="w"> </span>-<span class="w"> </span><span class="o">[</span>TRIO1<span class="o">]</span><span class="w"> </span>-<span class="w"> </span><span class="o">[</span>count_variants<span class="o">]</span><span class="w"> </span>-<span class="w"> </span><span class="o">[</span>Father<span class="o">]</span>:<span class="w"> </span>calculated<span class="w"> </span>examples<span class="w"> </span>per<span class="w"> </span>variant<span class="w"> </span><span class="p">|</span><span class="w"> </span><span class="m">0</span>.664
<a id="__codelineno-62-10" name="__codelineno-62-10" href="#__codelineno-62-10"></a><span class="m">2023</span>-06-29<span class="w"> </span><span class="m">11</span>:25:28<span class="w"> </span>AM<span class="w"> </span>-<span class="w"> </span><span class="o">[</span>INFO<span class="o">]</span><span class="w"> </span>-<span class="w"> </span><span class="o">[</span>DEMO<span class="o">]</span><span class="w"> </span>-<span class="w"> </span><span class="o">[</span>TRIO1<span class="o">]</span><span class="w"> </span>-<span class="w"> </span><span class="o">[</span>count_variants<span class="o">]</span><span class="w"> </span>-<span class="w"> </span><span class="o">[</span>Father<span class="o">]</span>:<span class="w"> </span>prevent<span class="w"> </span>underestimating<span class="w"> </span>which<span class="w"> </span>creates<span class="w"> </span>too<span class="w"> </span>many<span class="w"> </span>examples<span class="w"> </span>per<span class="w"> </span>region<span class="w"> </span>by<span class="w"> </span>rounding<span class="w"> </span>up<span class="w"> </span>to<span class="w"> </span>the<span class="w"> </span>nearest<span class="w"> </span><span class="m">0</span>.5<span class="w"> </span><span class="p">|</span><span class="w"> </span><span class="m">1</span>.0
<a id="__codelineno-62-11" name="__codelineno-62-11" href="#__codelineno-62-11"></a><span class="m">2023</span>-06-29<span class="w"> </span><span class="m">11</span>:25:28<span class="w"> </span>AM<span class="w"> </span>-<span class="w"> </span><span class="o">[</span>INFO<span class="o">]</span><span class="w"> </span>-<span class="w"> </span><span class="o">[</span>DEMO<span class="o">]</span><span class="w"> </span>-<span class="w"> </span><span class="o">[</span>TRIO1<span class="o">]</span><span class="w"> </span>-<span class="w"> </span><span class="o">[</span>count_variants<span class="o">]</span><span class="w"> </span>-<span class="w"> </span><span class="o">[</span>Father<span class="o">]</span>:<span class="w"> </span>default<span class="w"> </span>examples<span class="w"> </span>per<span class="w"> </span>variant<span class="w"> </span><span class="p">|</span><span class="w"> </span><span class="m">1</span>.5
<a id="__codelineno-62-12" name="__codelineno-62-12" href="#__codelineno-62-12"></a><span class="m">2023</span>-06-29<span class="w"> </span><span class="m">11</span>:25:28<span class="w"> </span>AM<span class="w"> </span>-<span class="w"> </span><span class="o">[</span>INFO<span class="o">]</span><span class="w"> </span>-<span class="w"> </span><span class="o">[</span>DEMO<span class="o">]</span><span class="w"> </span>-<span class="w"> </span><span class="o">[</span>TRIO1<span class="o">]</span><span class="w"> </span>-<span class="w"> </span><span class="o">[</span>count_variants<span class="o">]</span><span class="w"> </span>-<span class="w"> </span><span class="o">[</span>Father<span class="o">]</span>:<span class="w"> </span>difference<span class="w"> </span>between<span class="w"> </span>default<span class="w"> </span>and<span class="w"> </span>calculated<span class="w"> </span>examples<span class="w"> </span>per<span class="w"> </span>variant<span class="w"> </span><span class="p">|</span><span class="w"> </span><span class="m">0</span>.836
<a id="__codelineno-62-13" name="__codelineno-62-13" href="#__codelineno-62-13"></a><span class="m">2023</span>-06-29<span class="w"> </span><span class="m">11</span>:25:28<span class="w"> </span>AM<span class="w"> </span>-<span class="w"> </span><span class="o">[</span>INFO<span class="o">]</span><span class="w"> </span>-<span class="w"> </span><span class="o">[</span>DEMO<span class="o">]</span><span class="w"> </span>-<span class="w"> </span><span class="o">[</span>TRIO1<span class="o">]</span><span class="w"> </span>-<span class="w"> </span><span class="o">[</span>count_variants<span class="o">]</span><span class="w"> </span>-<span class="w"> </span><span class="o">[</span>Father<span class="o">]</span>:<span class="w"> </span>when<span class="w"> </span>running<span class="w"> </span>TrioTrain<span class="w"> </span>outside<span class="w"> </span>of<span class="w"> </span>this<span class="w"> </span>tutorial,<span class="w"> </span>please<span class="w"> </span>use<span class="w"> </span><span class="s1">&#39;--est-examples=1.0&#39;</span>
<a id="__codelineno-62-14" name="__codelineno-62-14" href="#__codelineno-62-14"></a><span class="m">2023</span>-06-29<span class="w"> </span><span class="m">11</span>:25:28<span class="w"> </span>AM<span class="w"> </span>-<span class="w"> </span><span class="o">[</span>INFO<span class="o">]</span><span class="w"> </span>-<span class="w"> </span>adding<span class="w"> </span><span class="nv">Est_Examples</span><span class="o">=</span><span class="s1">&#39;1.0&#39;</span>
<a id="__codelineno-62-15" name="__codelineno-62-15" href="#__codelineno-62-15"></a><span class="m">2023</span>-06-29<span class="w"> </span><span class="m">11</span>:25:28<span class="w"> </span>AM<span class="w"> </span>-<span class="w"> </span><span class="o">[</span>INFO<span class="o">]</span><span class="w"> </span>-<span class="w"> </span><span class="o">[</span>DEMO<span class="o">]</span><span class="w"> </span>-<span class="w"> </span><span class="o">[</span>TRIO1<span class="o">]</span><span class="w"> </span>-<span class="w"> </span><span class="o">[</span>count_variants<span class="o">]</span><span class="w"> </span>-<span class="w"> </span><span class="o">[</span>Father<span class="o">]</span>:<span class="w"> </span>added<span class="w"> </span><span class="s1">&#39;Est_Examples=1.0&#39;</span><span class="w"> </span>to<span class="w"> </span>env<span class="w"> </span><span class="nv">file</span>
<a id="__codelineno-62-16" name="__codelineno-62-16" href="#__codelineno-62-16"></a><span class="o">=====</span><span class="w"> </span>end<span class="w"> </span>of<span class="w"> </span>triotrain/model_training/tutorial/estimate.py<span class="w"> </span>@<span class="w"> </span><span class="m">2023</span>-06-29<span class="w">  </span><span class="m">11</span>:25:28<span class="w"> </span><span class="o">=====</span>
</code></pre></div>
</details>
<h2 id="9-run-triotrain-with-a-human-trio">9. Run TrioTrain with a Human Trio</h2>
<p>Now that we know how to tailor TrioTrain for our non-bovine species (human), we can move forward with starting the re-training pipeline.</p>
<p>Complete the GIAB TrioTrain by <strong>editing the following at the command to include your new SLURM config file</strong>:</p>
<div class="highlight"><span class="filename">Run the following at the command line:</span><pre><span></span><code><a id="__codelineno-63-1" name="__codelineno-63-1" href="#__codelineno-63-1"></a><span class="c1"># You can add the --dry-run flag to this command to confirm the TrioTrain pipeline runs smoothly prior to submitting jobs to the queue</span>
<a id="__codelineno-63-2" name="__codelineno-63-2" href="#__codelineno-63-2"></a><span class="nb">source</span><span class="w"> </span>./scripts/setup/modules.sh
<a id="__codelineno-63-3" name="__codelineno-63-3" href="#__codelineno-63-3"></a><span class="nb">source</span><span class="w"> </span>./scripts/start_conda.sh<span class="w">                 </span><span class="c1"># Ensure the previously built conda env is active</span>
<a id="__codelineno-63-4" name="__codelineno-63-4" href="#__codelineno-63-4"></a>python3<span class="w"> </span>triotrain/run_trio_train.py<span class="w">                                                                         </span><span class="se">\</span>
<a id="__codelineno-63-5" name="__codelineno-63-5" href="#__codelineno-63-5"></a><span class="w">    </span>-g<span class="w"> </span>Father<span class="w">                                                                                               </span><span class="se">\</span>
<a id="__codelineno-63-6" name="__codelineno-63-6" href="#__codelineno-63-6"></a><span class="w">    </span>--unmapped-reads<span class="w"> </span>chrUn<span class="w">                                                                                  </span><span class="se">\</span>
<a id="__codelineno-63-7" name="__codelineno-63-7" href="#__codelineno-63-7"></a><span class="w">    </span>--est-examples<span class="w"> </span><span class="m">1</span><span class="w">                                                                                        </span><span class="se">\</span>
<a id="__codelineno-63-8" name="__codelineno-63-8" href="#__codelineno-63-8"></a><span class="w">    </span>-m<span class="w"> </span>triotrain/model_training/tutorial/GIAB.Human_tutorial_metadata.csv<span class="w">                                   </span><span class="se">\</span>
<a id="__codelineno-63-9" name="__codelineno-63-9" href="#__codelineno-63-9"></a><span class="w">    </span>-n<span class="w"> </span>GIAB_Trio<span class="w">                                                                                            </span><span class="se">\</span>
<a id="__codelineno-63-10" name="__codelineno-63-10" href="#__codelineno-63-10"></a><span class="w">    </span>--num-tests<span class="w"> </span><span class="m">3</span><span class="w">                                                                                           </span><span class="se">\</span>
<a id="__codelineno-63-11" name="__codelineno-63-11" href="#__codelineno-63-11"></a><span class="w">    </span>--custom-checkpoint<span class="w"> </span>triotrain/model_training/pretrained_models/v1.4.0_withIS_withAF/wgs_af.model.ckpt<span class="w">   </span><span class="se">\</span>
<a id="__codelineno-63-12" name="__codelineno-63-12" href="#__codelineno-63-12"></a><span class="w">    </span>--output<span class="w"> </span>../TUTORIAL<span class="w">                                                                                    </span><span class="se">\</span>
<a id="__codelineno-63-13" name="__codelineno-63-13" href="#__codelineno-63-13"></a><span class="w">    </span>--benchmark<span class="w">                                                                                             </span><span class="se">\</span>
<a id="__codelineno-63-14" name="__codelineno-63-14" href="#__codelineno-63-14"></a><span class="w">    </span>-r<span class="w"> </span>&lt;/path/to/new_file.json&gt;
</code></pre></div>
<p><a class="md-button" href="../../user-guide/resubmit/">Need to Re-Submit a SLURM job <span class="twemoji"><svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 16 16"><path d="M0 8a8 8 0 1 1 16 0A8 8 0 0 1 0 8Zm8-6.5a6.5 6.5 0 1 0 0 13 6.5 6.5 0 0 0 0-13ZM6.92 6.085h.001a.749.749 0 1 1-1.342-.67c.169-.339.436-.701.849-.977C6.845 4.16 7.369 4 8 4a2.756 2.756 0 0 1 1.637.525c.503.377.863.965.863 1.725 0 .448-.115.83-.329 1.15-.205.307-.47.513-.692.662-.109.072-.22.138-.313.195l-.006.004a6.24 6.24 0 0 0-.26.16.952.952 0 0 0-.276.245.75.75 0 0 1-1.248-.832c.184-.264.42-.489.692-.661.103-.067.207-.132.313-.195l.007-.004c.1-.061.182-.11.258-.161a.969.969 0 0 0 .277-.245C8.96 6.514 9 6.427 9 6.25a.612.612 0 0 0-.262-.525A1.27 1.27 0 0 0 8 5.5c-.369 0-.595.09-.74.187a1.01 1.01 0 0 0-.34.398ZM9 11a1 1 0 1 1-2 0 1 1 0 0 1 2 0Z"/></svg></span></a></p>
<details class="success">
<summary>Check | Baseline WGS.AF:</summary>
<div class="highlight"><span class="filename">Run the following at the command line:</span><pre><span></span><code><a id="__codelineno-64-1" name="__codelineno-64-1" href="#__codelineno-64-1"></a>ls<span class="w"> </span>../TUTORIAL/baseline_v1.4.0_withIS_withAF/<span class="w"> </span><span class="p">|</span><span class="w"> </span>grep<span class="w"> </span>total<span class="w"> </span><span class="p">|</span><span class="w"> </span>grep<span class="w"> </span>total
</code></pre></div>
<div class="highlight"><span class="filename">Expected outputs:</span><pre><span></span><code><a id="__codelineno-65-1" name="__codelineno-65-1" href="#__codelineno-65-1"></a>Test1.total.metrics.csv
<a id="__codelineno-65-2" name="__codelineno-65-2" href="#__codelineno-65-2"></a>Test2.total.metrics.csv
<a id="__codelineno-65-3" name="__codelineno-65-3" href="#__codelineno-65-3"></a>Test3.total.metrics.csv
</code></pre></div>
</details>
<details class="success">
<summary>Check | Father:</summary>
<div class="highlight"><span class="filename">Run the following at the command line:</span><pre><span></span><code><a id="__codelineno-66-1" name="__codelineno-66-1" href="#__codelineno-66-1"></a>ls<span class="w"> </span>../TUTORIAL/GIAB_Trio/Human_tutorial/compare_Father/<span class="w"> </span><span class="p">|</span><span class="w"> </span>grep<span class="w"> </span>total
</code></pre></div>
<div class="highlight"><span class="filename">Expected outputs:</span><pre><span></span><code><a id="__codelineno-67-1" name="__codelineno-67-1" href="#__codelineno-67-1"></a>Test1.total.metrics.csv
<a id="__codelineno-67-2" name="__codelineno-67-2" href="#__codelineno-67-2"></a>Test2.total.metrics.csv
<a id="__codelineno-67-3" name="__codelineno-67-3" href="#__codelineno-67-3"></a>Test3.total.metrics.csv
</code></pre></div>
</details>
<details class="success">
<summary>Check | Mother:</summary>
<div class="highlight"><span class="filename">Run the following at the command line:</span><pre><span></span><code><a id="__codelineno-68-1" name="__codelineno-68-1" href="#__codelineno-68-1"></a>ls<span class="w"> </span>../TUTORIAL/GIAB_Trio/Human_tutorial/compare_Mother/<span class="w"> </span><span class="p">|</span><span class="w"> </span>grep<span class="w"> </span>total
</code></pre></div>
<div class="highlight"><span class="filename">Expected outputs:</span><pre><span></span><code><a id="__codelineno-69-1" name="__codelineno-69-1" href="#__codelineno-69-1"></a>Test1.total.metrics.csv
<a id="__codelineno-69-2" name="__codelineno-69-2" href="#__codelineno-69-2"></a>Test2.total.metrics.csv
<a id="__codelineno-69-3" name="__codelineno-69-3" href="#__codelineno-69-3"></a>Test3.total.metrics.csv
</code></pre></div>
</details>
<h2 id="merge-results-per-iteration">Merge Results | Per-Iteration</h2>
<p>Each <code>Test#.total.metrics.csv</code> output file should contain 57 rows and 2 columns. The metrics within are the raw and proprotional performance metrics from hap.py. After all <code>convert_happy</code> jobs complete, we will separately merge the results from running the baseline iteration, and both training iterations completed during the GIAB tutorial:</p>
<div class="highlight"><span class="filename">Run the following at the command line:</span><pre><span></span><code><a id="__codelineno-70-1" name="__codelineno-70-1" href="#__codelineno-70-1"></a><span class="nb">source</span><span class="w"> </span>./scripts/setup/modules.sh
<a id="__codelineno-70-2" name="__codelineno-70-2" href="#__codelineno-70-2"></a><span class="nb">source</span><span class="w"> </span>./scripts/start_conda.sh<span class="w">                 </span><span class="c1"># Ensure the previously built conda env is active</span>
<a id="__codelineno-70-3" name="__codelineno-70-3" href="#__codelineno-70-3"></a><span class="k">for</span><span class="w"> </span>start_i<span class="w"> </span><span class="k">in</span><span class="w"> </span><span class="k">$(</span>seq<span class="w"> </span><span class="m">0</span><span class="w"> </span><span class="m">1</span><span class="k">)</span><span class="p">;</span><span class="w"> </span><span class="k">do</span>
<a id="__codelineno-70-4" name="__codelineno-70-4" href="#__codelineno-70-4"></a><span class="w">    </span><span class="nb">echo</span><span class="w"> </span><span class="s2">&quot;</span><span class="k">$(</span>date<span class="w"> </span><span class="s1">&#39;+%Y-%m-%d %H:%M:%S&#39;</span><span class="k">)</span><span class="s2"> INFO: merging processed results from hap.py for GIAB run#</span><span class="si">${</span><span class="nv">start_i</span><span class="si">}</span><span class="s2">&quot;</span>
<a id="__codelineno-70-5" name="__codelineno-70-5" href="#__codelineno-70-5"></a><span class="w">    </span>python3<span class="w"> </span>triotrain/summarize/merge_results.py<span class="w"> </span>--env<span class="w"> </span>../TUTORIAL/GIAB_Trio/envs/run<span class="si">${</span><span class="nv">start_i</span><span class="si">}</span>.env<span class="w"> </span>-g<span class="w"> </span>Father<span class="w"> </span>-m<span class="w"> </span>triotrain/summarize/data/tutorial_metadata.csv
<a id="__codelineno-70-6" name="__codelineno-70-6" href="#__codelineno-70-6"></a><span class="w">    </span><span class="nb">echo</span><span class="w"> </span><span class="s2">&quot;</span><span class="k">$(</span>date<span class="w"> </span><span class="s1">&#39;+%Y-%m-%d %H:%M:%S&#39;</span><span class="k">)</span><span class="s2"> INFO: finished merging processed results from hap.py for GIAB run#</span><span class="si">${</span><span class="nv">start_i</span><span class="si">}</span><span class="s2">&quot;</span>
<a id="__codelineno-70-7" name="__codelineno-70-7" href="#__codelineno-70-7"></a><span class="k">done</span>
</code></pre></div>
<details class="success">
<summary>Check | Baseline WGS.AF</summary>
<div class="highlight"><span class="filename">Run the following at the command line:</span><pre><span></span><code><a id="__codelineno-71-1" name="__codelineno-71-1" href="#__codelineno-71-1"></a>ls<span class="w"> </span>../TUTORIAL/baseline_v1.4.0_withIS_withAF/<span class="w"> </span><span class="p">|</span><span class="w"> </span>grep<span class="w"> </span>All
</code></pre></div>
<div class="highlight"><span class="filename">Expected outputs:</span><pre><span></span><code><a id="__codelineno-72-1" name="__codelineno-72-1" href="#__codelineno-72-1"></a>DV1.4_WGS.AF_human.AllTests.total.metrics.csv
</code></pre></div>
</details>
<details class="success">
<summary>Check | GIAB Trio1</summary>
<div class="highlight"><span class="filename">Run the following at the command line:</span><pre><span></span><code><a id="__codelineno-73-1" name="__codelineno-73-1" href="#__codelineno-73-1"></a>ls<span class="w"> </span>../TUTORIAL/GIAB_Trio/summary<span class="w"> </span><span class="p">|</span><span class="w"> </span>grep<span class="w"> </span>All
</code></pre></div>
<div class="highlight"><span class="filename">Expected outputs:</span><pre><span></span><code><a id="__codelineno-74-1" name="__codelineno-74-1" href="#__codelineno-74-1"></a>Trio1.AllTests.total.metrics.csv
</code></pre></div>
</details>
<h2 id="clean-up-directories-per-iteration">Clean Up Directories | Per-Iteration</h2>
<p>Running TrioTrain produces a large volume of temporary files. Run the following at the command line to free up space:</p>
<div class="highlight"><span class="filename">Run the following at the command line:</span><pre><span></span><code><a id="__codelineno-75-1" name="__codelineno-75-1" href="#__codelineno-75-1"></a><span class="nb">source</span><span class="w"> </span>./scripts/setup/modules.sh
<a id="__codelineno-75-2" name="__codelineno-75-2" href="#__codelineno-75-2"></a><span class="nb">source</span><span class="w"> </span>./scripts/start_conda.sh<span class="w">                 </span><span class="c1"># Ensure the previously built conda env is active</span>
<a id="__codelineno-75-3" name="__codelineno-75-3" href="#__codelineno-75-3"></a><span class="k">for</span><span class="w"> </span>start_i<span class="w"> </span><span class="k">in</span><span class="w"> </span><span class="k">$(</span>seq<span class="w"> </span><span class="m">0</span><span class="w"> </span><span class="m">1</span><span class="k">)</span><span class="p">;</span><span class="w"> </span><span class="k">do</span>
<a id="__codelineno-75-4" name="__codelineno-75-4" href="#__codelineno-75-4"></a><span class="w">    </span><span class="nb">echo</span><span class="w"> </span><span class="s2">&quot;</span><span class="k">$(</span>date<span class="w"> </span><span class="s1">&#39;+%Y-%m-%d %H:%M:%S&#39;</span><span class="k">)</span><span class="s2"> INFO: removing temp files for GIAB run#</span><span class="si">${</span><span class="nv">start_i</span><span class="si">}</span><span class="s2">&quot;</span>
<a id="__codelineno-75-5" name="__codelineno-75-5" href="#__codelineno-75-5"></a><span class="w">    </span>python3<span class="w"> </span>triotrain/model_training/slurm/clean_tmp.py<span class="w"> </span>--env<span class="w"> </span>../TUTORIAL/GIAB_Trio/envs/run<span class="si">${</span><span class="nv">start_i</span><span class="si">}</span>.env
<a id="__codelineno-75-6" name="__codelineno-75-6" href="#__codelineno-75-6"></a><span class="w">    </span><span class="nb">echo</span><span class="w"> </span><span class="s2">&quot;</span><span class="k">$(</span>date<span class="w"> </span><span class="s1">&#39;+%Y-%m-%d %H:%M:%S&#39;</span><span class="k">)</span><span class="s2"> INFO: finished removing temp files for GIAB run#</span><span class="si">${</span><span class="nv">start_i</span><span class="si">}</span><span class="s2">&quot;</span>
<a id="__codelineno-75-7" name="__codelineno-75-7" href="#__codelineno-75-7"></a><span class="k">done</span>
</code></pre></div>
<details class="success">
<summary>Check | Baseline WGS.AF</summary>
<div class="highlight"><span class="filename">Run the following at the command line:</span><pre><span></span><code><a id="__codelineno-76-1" name="__codelineno-76-1" href="#__codelineno-76-1"></a>du<span class="w"> </span>-h<span class="w"> </span>../TUTORIAL/baseline_v1.4.0_withIS_withAF/
</code></pre></div>
<div class="highlight"><span class="filename">Expected outputs:</span><pre><span></span><code><a id="__codelineno-77-1" name="__codelineno-77-1" href="#__codelineno-77-1"></a>700M<span class="w">    </span>../TUTORIAL/baseline_v1.4.0_withIS_withAF/
</code></pre></div>
</details>
<details class="success">
<summary>Check | GIAB Trio1</summary>
<div class="highlight"><span class="filename">Run the following at the command line:</span><pre><span></span><code><a id="__codelineno-78-1" name="__codelineno-78-1" href="#__codelineno-78-1"></a>du<span class="w"> </span>-h<span class="w"> </span>../TUTORIAL/GIAB_Trio
<a id="__codelineno-78-2" name="__codelineno-78-2" href="#__codelineno-78-2"></a><span class="sb">``</span>
<a id="__codelineno-78-3" name="__codelineno-78-3" href="#__codelineno-78-3"></a>
<a id="__codelineno-78-4" name="__codelineno-78-4" href="#__codelineno-78-4"></a><span class="sb">```</span>bash<span class="w"> </span><span class="nv">title</span><span class="o">=</span><span class="s2">&quot;Expected outputs:&quot;</span>
<a id="__codelineno-78-5" name="__codelineno-78-5" href="#__codelineno-78-5"></a>141M<span class="w">    </span>../TUTORIAL/GIAB_Trio/Human_tutorial/logs
<a id="__codelineno-78-6" name="__codelineno-78-6" href="#__codelineno-78-6"></a>402M<span class="w">    </span>../TUTORIAL/GIAB_Trio/Human_tutorial/compare_Mother
<a id="__codelineno-78-7" name="__codelineno-78-7" href="#__codelineno-78-7"></a><span class="m">1</span>.1M<span class="w">    </span>../TUTORIAL/GIAB_Trio/Human_tutorial/train_Mother/eval_Child
<a id="__codelineno-78-8" name="__codelineno-78-8" href="#__codelineno-78-8"></a>23G<span class="w">     </span>../TUTORIAL/GIAB_Trio/Human_tutorial/train_Mother
<a id="__codelineno-78-9" name="__codelineno-78-9" href="#__codelineno-78-9"></a>394M<span class="w">    </span>../TUTORIAL/GIAB_Trio/Human_tutorial/test_Father
<a id="__codelineno-78-10" name="__codelineno-78-10" href="#__codelineno-78-10"></a>298M<span class="w">    </span>../TUTORIAL/GIAB_Trio/Human_tutorial/compare_Father
<a id="__codelineno-78-11" name="__codelineno-78-11" href="#__codelineno-78-11"></a><span class="m">1</span>.1M<span class="w">    </span>../TUTORIAL/GIAB_Trio/Human_tutorial/train_Father/eval_Child
<a id="__codelineno-78-12" name="__codelineno-78-12" href="#__codelineno-78-12"></a>23G<span class="w">     </span>../TUTORIAL/GIAB_Trio/Human_tutorial/train_Father
<a id="__codelineno-78-13" name="__codelineno-78-13" href="#__codelineno-78-13"></a>243G<span class="w">    </span>../TUTORIAL/GIAB_Trio/Human_tutorial/examples
<a id="__codelineno-78-14" name="__codelineno-78-14" href="#__codelineno-78-14"></a>391M<span class="w">    </span>../TUTORIAL/GIAB_Trio/Human_tutorial/test_Mother
<a id="__codelineno-78-15" name="__codelineno-78-15" href="#__codelineno-78-15"></a><span class="m">3</span>.1M<span class="w">    </span>../TUTORIAL/GIAB_Trio/Human_tutorial/jobs
<a id="__codelineno-78-16" name="__codelineno-78-16" href="#__codelineno-78-16"></a>290G<span class="w">    </span>../TUTORIAL/GIAB_Trio/Human_tutorial
<a id="__codelineno-78-17" name="__codelineno-78-17" href="#__codelineno-78-17"></a>100K<span class="w">    </span>../TUTORIAL/GIAB_Trio/summary
<a id="__codelineno-78-18" name="__codelineno-78-18" href="#__codelineno-78-18"></a>50K<span class="w">     </span>../TUTORIAL/GIAB_Trio/envs
<a id="__codelineno-78-19" name="__codelineno-78-19" href="#__codelineno-78-19"></a>290G<span class="w">    </span>../TUTORIAL/GIAB_Trio/
</code></pre></div>
</details>
<p><a class="md-button" href="../../user-guide/mie/">Next - Comparing Models <span class="twemoji"><svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M5 21a2 2 0 0 1-2-2V5a2 2 0 0 1 2-2h14a2 2 0 0 1 2 2v14a2 2 0 0 1-2 2H5m1-8h8.5L11 16.5l1.42 1.42L18.34 12l-5.92-5.92L11 7.5l3.5 3.5H6v2Z"/></svg></span></a></p>

  <hr>
<div class="md-source-file">
  <small>
    
      Last update:
      <span class="git-revision-date-localized-plugin git-revision-date-localized-plugin-date">March 8, 2024</span>
      
    
  </small>
</div>





                
              </article>
            </div>
          
          
        </div>
        
          <button type="button" class="md-top md-icon" data-md-component="top" hidden>
            <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M13 20h-2V8l-5.5 5.5-1.42-1.42L12 4.16l7.92 7.92-1.42 1.42L13 8v12Z"/></svg>
            Back to top
          </button>
        
      </main>
      
        <footer class="md-footer">
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    <script id="__config" type="application/json">{"base": "../..", "features": ["navigation.tabs", "navigation.tabs.sticky", "navigation.top", "toc.integrate", "content.code.copy"], "search": "../../assets/javascripts/workers/search.208ed371.min.js", "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}}</script>
    
    
      <script src="../../assets/javascripts/bundle.b4d07000.min.js"></script>
      
        <script src="../../js/open_in_new_tab.js"></script>
      
    
  </body>
</html>