# DV-TrioTrain v0.8

Support Documentation for DV-TrioTrain.

---

DeepVariant-TrioTrain is an automated pipeline for extending DeepVariant (DV), a deep-learning-based germline variant caller. Visit the [DV-TrioTrain github page](https://https://github.com/jkalleberg/DV-TrioTrain) for the source code.

<div class="text-center">
<a href="user-guide/getting-started/" class="btn btn-primary" role="button">Getting Started</a>
<a href="user-guide/" class="btn btn-primary" role="button">User Guide</a>
</div>

---

## Feedback and technical support

For questions, suggestions, or technical assistance, feel free to [open an issue](https://github.com/jkalleberg/DV-TrioTrain/issues) page or
reach out to Jenna Kalleberg at <jakth2@mail.missouri.org>

### Contributing to TrioTrain

Please [open a pull request (fix this link)]() if you wish to contribute to TrioTrain.

<a name="citation"></a>

## How to cite

> Citation to go here

Please also cite:

> [A universal SNP and small-indel variant caller using deep neural networks. *Nature Biotechnology* 36, 983â€“987 (2018).](https://rdcu.be/7Dhl) <br/>
Ryan Poplin, Pi-Chuan Chang, David Alexander, Scott Schwartz, Thomas Colthurst, Alexander Ku, Dan Newburger, Jojo Dijamco, Nam Nguyen, Pegah T. Afshar, Sam S. Gross, Lizzie Dorfman, Cory Y. McLean, and Mark A. DePristo.<br/>
doi: <https://doi.org/10.1038/nbt.4235>

## License

[GPL-3.0 license](LICENSE)

### Authors

Jenna Kalleberg (<jakth2@mail.missouri.org>)

## Acknowledgements

Many thanks to the developers and contributors of the many open source packages used by TrioTrain:
